
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
      
        <meta name="author" content="shenwei356">
      
      
      <link rel="shortcut icon" href="../files/favicon.ico">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.4">
    
    
      
        <title>Tutorial - TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.15aa0b43.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.75751829.min.css">
        
          
          
          <meta name="theme-color" content="#00bdd6">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
      <link rel="manifest" href="../manifest.webmanifest" crossorigin="use-credentials">
    
    
    
      
        
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-28948626-9","bioinf.shenwei.me/taxonkit"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){if(this.value){var e=document.location.pathname;ga("send","pageview",e+"?q="+this.value)}})}),document.addEventListener("DOMContentSwitch",function(){ga("send","pageview",document.location.pathname)})</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      



  <li class="md-nav__item">
    <a href=".." class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      



  <li class="md-nav__item">
    <a href="../download/" class="md-nav__link">
      Download
    </a>
  </li>

    
      
      
      



  <li class="md-nav__item">
    <a href="../usage/" class="md-nav__link">
      Usage
    </a>
  </li>

    
      
      
      


  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Tutorial
        <span class="md-nav__icon md-icon"></span>
      </label>
    
    <a href="./" class="md-nav__link md-nav__link--active">
      Tutorial
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formating-lineage" class="md-nav__link">
    Formating lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      



  <li class="md-nav__item">
    <a href="/bench/" class="md-nav__link">
      Benchmark
    </a>
  </li>

    
      
      
      



  <li class="md-nav__item">
    <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
      Taxid-Changelog
    </a>
  </li>

    
      
      
      



  <li class="md-nav__item">
    <a href="https://github.com/shenwei356" class="md-nav__link">
      More tools
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formating-lineage" class="md-nav__link">
    Formating lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/shenwei356/taxonkit/edit/master/docs/tutorial.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="tutorial">Tutorial</h1>
<h2 id="table-of-contents">Table of Contents</h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#summaries-of-taxonomy-data">Summaries of taxonomy data</a></li>
<li><a href="#formating-lineage">Formating lineage</a></li>
<li><a href="#parsing-krakenbraken-result">Parsing kraken/bracken result</a></li>
<li><a href="#making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="summaries-of-taxonomy-data">Summaries of taxonomy data</h2>
<p>You can change the taxID of interest.</p>
<ol>
<li>
<p>Rank counts of common categories.</p>
<pre><code>$ echo Archaea Bacteria Eukaryota Fungi Metazoa Viridiplantae \
    | rush -D ' ' -T b \
        'taxonkit list --ids $(echo {} | taxonkit name2taxid | cut -f 2) \
            | sed 1d \
            | taxonkit filter -i 2 -E genus -L genus \
            | taxonkit lineage -L -r \
            | csvtk freq -H -t -f 2 -nr \
            &gt; stats.{}.tsv '

$ csvtk -t join --outer-join stats.*.tsv \
    | csvtk add-header -t -n "rank,$(ls stats.*.tsv | rush -k 'echo {@stats.(.+).tsv}' | paste -sd, )" \
    | csvtk csv2md -t
</code></pre>
<p><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=statistics&amp;uncultured=hide&amp;unspecified=hide">Similar data on NCBI Taxonomy</a></p>
<table>
<thead>
<tr>
<th align="left">rank</th>
<th align="left">Archaea</th>
<th align="left">Bacteria</th>
<th align="left">Eukaryota</th>
<th align="left">Fungi</th>
<th align="left">Metazoa</th>
<th align="left">Viridiplantae</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">species</td>
<td align="left">12482</td>
<td align="left">460940</td>
<td align="left">1349648</td>
<td align="left">156908</td>
<td align="left">957297</td>
<td align="left">191026</td>
</tr>
<tr>
<td align="left">strain</td>
<td align="left">354</td>
<td align="left">40643</td>
<td align="left">3486</td>
<td align="left">2352</td>
<td align="left">33</td>
<td align="left">50</td>
</tr>
<tr>
<td align="left">genus</td>
<td align="left">205</td>
<td align="left">4112</td>
<td align="left">90882</td>
<td align="left">6844</td>
<td align="left">64148</td>
<td align="left">16202</td>
</tr>
<tr>
<td align="left">isolate</td>
<td align="left">7</td>
<td align="left">503</td>
<td align="left">809</td>
<td align="left">76</td>
<td align="left">17</td>
<td align="left">3</td>
</tr>
<tr>
<td align="left">species group</td>
<td align="left">2</td>
<td align="left">77</td>
<td align="left">251</td>
<td align="left">22</td>
<td align="left">214</td>
<td align="left">5</td>
</tr>
<tr>
<td align="left">serotype</td>
<td align="left"></td>
<td align="left">218</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">serogroup</td>
<td align="left"></td>
<td align="left">136</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">subsection</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">21</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">21</td>
</tr>
<tr>
<td align="left">subspecies</td>
<td align="left"></td>
<td align="left">632</td>
<td align="left">24523</td>
<td align="left">158</td>
<td align="left">17043</td>
<td align="left">7212</td>
</tr>
<tr>
<td align="left">forma specialis</td>
<td align="left"></td>
<td align="left">521</td>
<td align="left">220</td>
<td align="left">179</td>
<td align="left">33</td>
<td align="left">1</td>
</tr>
<tr>
<td align="left">species subgroup</td>
<td align="left"></td>
<td align="left">23</td>
<td align="left">101</td>
<td align="left"></td>
<td align="left">101</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">biotype</td>
<td align="left"></td>
<td align="left">7</td>
<td align="left">10</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">morph</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
<td align="left">3</td>
<td align="left">4</td>
<td align="left">5</td>
</tr>
<tr>
<td align="left">section</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">437</td>
<td align="left">37</td>
<td align="left">2</td>
<td align="left">398</td>
</tr>
<tr>
<td align="left">genotype</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
</tr>
<tr>
<td align="left">series</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">9</td>
<td align="left"></td>
<td align="left">5</td>
<td align="left">4</td>
</tr>
<tr>
<td align="left">varietas</td>
<td align="left"></td>
<td align="left">25</td>
<td align="left">8499</td>
<td align="left">1100</td>
<td align="left">2</td>
<td align="left">7188</td>
</tr>
<tr>
<td align="left">forma</td>
<td align="left"></td>
<td align="left">4</td>
<td align="left">560</td>
<td align="left">185</td>
<td align="left">6</td>
<td align="left">315</td>
</tr>
<tr>
<td align="left">subgenus</td>
<td align="left"></td>
<td align="left">1</td>
<td align="left">1558</td>
<td align="left">10</td>
<td align="left">1414</td>
<td align="left">112</td>
</tr>
<tr>
<td align="left">pathogroup</td>
<td align="left"></td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">subvariety</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">5</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Count of all ranks</p>
<pre><code>$ time taxonkit list --ids 1 \
    | taxonkit lineage -L -r \
    | csvtk freq -H -t -f 2 -nr \
    | csvtk pretty -t

species            1879659
no rank            222743
genus              96625
strain             44483
subspecies         25174
family             9492
varietas           8524
subfamily          3050
tribe              2213
order              1660
subgenus           1618
isolate            1319
serotype           1216
clade              886
superfamily        865
forma specialis    741
forma              564
subtribe           508
section            437
class              429
suborder           372
species group      330
phylum             272
subclass           156
serogroup          138
infraorder         130
species subgroup   124
superorder         55
subphylum          33
parvorder          26
subsection         21
genotype           20
infraclass         18
biotype            17
morph              12
kingdom            11
series             9
superclass         6
cohort             5
pathogroup         5
subvariety         5
superkingdom       4
subcohort          3
subkingdom         1
superphylum        1

real    0m3.663s
user    0m15.897s
sys     0m1.010s
</code></pre>
</li>
</ol>
<h2 id="formating-lineage">Formating lineage</h2>
<p>Example data.</p>
<pre><code>$ cat taxids3.txt
376619
349741
239935
314101
11932
1327037
</code></pre>
<p>Format to seven-level ranks ("superkingdom phylum class order family genus species").</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit lineage \
    | taxonkit reformat \
    | cut -f 1,3
376619  Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis
349741  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
314101  Bacteria;;;;;;uncultured murine large bowel bacterium BAC 54B
11932   Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
1327037 Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;;Croceibacter phage P2559Y
</code></pre>
<p>Replace missing ranks with <code>Unassigned</code> and output tab-delimited format.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit lineage \
    | taxonkit reformat -r "Unassigned" -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}" \
    | cut -f 1,3-9 \
    | csvtk pretty -t 
376619    Bacteria   Proteobacteria    Gammaproteobacteria   Thiotrichales        Francisellaceae   Francisella                  Francisella tularensis
349741    Bacteria   Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae   Akkermansia                  Akkermansia muciniphila
239935    Bacteria   Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae   Akkermansia                  Akkermansia muciniphila
314101    Bacteria   Unassigned        Unassigned            Unassigned           Unassigned        Unassigned                   uncultured murine large bowel bacterium BAC 54B
11932     Viruses    Artverviricota    Revtraviricetes       Ortervirales         Retroviridae      Intracisternal A-particles   Mouse Intracisternal A-particle
1327037   Viruses    Uroviricota       Caudoviricetes        Caudovirales         Siphoviridae      Unassigned                   Croceibacter phage P2559Y
</code></pre>
<p>Fill missing ranks and add prefixes.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit lineage \
    | taxonkit reformat -F -P \
    | cut -f 1,3
376619  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Thiotrichales;f__Francisellaceae;g__Francisella;s__Francisella tularensis
349741  k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
239935  k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
314101  k__Bacteria;p__unclassified Bacteria phylum;c__unclassified Bacteria class;o__unclassified Bacteria order;f__unclassified Bacteria family;g__unclassified Bacteria genus;s__uncultured murine large bowel bacterium BAC 54B
11932   k__Viruses;p__Artverviricota;c__Revtraviricetes;o__Ortervirales;f__Retroviridae;g__Intracisternal A-particles;s__Mouse Intracisternal A-particle
1327037 k__Viruses;p__Uroviricota;c__Caudoviricetes;o__Caudovirales;f__Siphoviridae;g__unclassified Siphoviridae genus;s__Croceibacter phage P2559Y
</code></pre>
<p>Single prefix of a rank can be set with flag like <code>--prefix-k</code>.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit lineage \
    | taxonkit reformat -F -P --prefix-k "d__" \
    | cut -f 1,3
376619  d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Thiotrichales;f__Francisellaceae;g__Francisella;s__Francisella tularensis
349741  d__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
239935  d__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
314101  d__Bacteria;p__unclassified Bacteria phylum;c__unclassified Bacteria class;o__unclassified Bacteria order;f__unclassified Bacteria family;g__unclassified Bacteria genus;s__uncultured murine large bowel bacterium BAC 54B
11932   d__Viruses;p__Artverviricota;c__Revtraviricetes;o__Ortervirales;f__Retroviridae;g__Intracisternal A-particles;s__Mouse Intracisternal A-particle
1327037 d__Viruses;p__Uroviricota;c__Caudoviricetes;o__Caudovirales;f__Siphoviridae;g__unclassified Siphoviridae genus;s__Croceibacter phage P2559Y
</code></pre>
<h2 id="parsing-krakenbracken-result">Parsing kraken/bracken result</h2>
<p>Example Data.</p>
<pre><code>$ head -n 10  test_bracken_species.kreport 
100.00  11430773        0       R       1       root
92.12   10530410        0       R1      131567    cellular organisms
92.06   10522651        0       D       2           Bacteria
92.04   10520821        0       P       1224          Proteobacteria
92.02   10518133        0       C       1236            Gammaproteobacteria
92.00   10516208        0       O       91347             Enterobacterales
91.81   10494626        0       F       543                 Enterobacteriaceae
86.73   9914368 0       G       561                   Escherichia
84.84   9698064 9698064 S       562                     Escherichia coli
0.93    106494  106494  S       208962                  Escherichia albertii
</code></pre>
<p>Save taxon proportion and taxid, and get lineage, name and rank.</p>
<pre><code>$ time cat test_bracken_species.kreport \
    | csvtk cut -Ht -f 1,5 \
    | taxonkit lineage -i 2 -n -r \
    | csvtk cut -Ht -f 1,2,5,4,3 \
    | head -n 10

100.00  1       no rank root    root
92.12   131567  no rank cellular organisms      cellular organisms
92.06   2       superkingdom    Bacteria        cellular organisms;Bacteria
92.04   1224    phylum  Proteobacteria  cellular organisms;Bacteria;Proteobacteria
92.02   1236    class   Gammaproteobacteria     cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria
92.00   91347   order   Enterobacterales        cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales
91.81   543     family  Enterobacteriaceae      cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae
86.73   561     genus   Escherichia     cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia
84.84   562     species Escherichia coli        cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli
0.93    208962  species Escherichia albertii    cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia albertii

real    0m1.426s
user    0m2.993s
sys     0m0.260s
</code></pre>
<p>Only save species or lower level and get lineage in format of "superkingdom phylum class order family genus species".</p>
<pre><code>$ time cat test_bracken_species.kreport \
    | csvtk cut -Ht -f 1,5 \
    | taxonkit filter -E species -L species -i 2 \
    | taxonkit lineage -i 2 -n -r \
    | taxonkit reformat -i 3 -f "{k};{p};{c};{o};{f};{g};{s}" \
    | csvtk cut -Ht -f 1,2,5,4,6 \
    | csvtk add-header -t -n abundance,taxid,rank,name,lineage \
    | head -n 10 \
    | csvtk pretty -t

abundance   taxid     name      rank                     lineage
84.84       562       species   Escherichia coli         Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli
0.93        208962    species   Escherichia albertii     Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia albertii
0.90        564       species   Escherichia fergusonii   Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia fergusonii
0.05        2725997   species   Escherichia sp. SCLE84   Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia sp. SCLE84
0.00        2044467   species   Escherichia sp. E4742    Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia sp. E4742
0.01        1499973   species   Escherichia marmotae     Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia marmotae
2.68        621       species   Shigella boydii          Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Shigella;Shigella boydii
0.04        622       species   Shigella dysenteriae     Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Shigella;Shigella dysenteriae
0.98        28901     species   Salmonella enterica      Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Salmonella;Salmonella enterica

real    0m4.142s
user    0m12.874s
sys     0m0.808s
</code></pre>
<h2 id="making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids</h2>
<p>Attention:</p>
<ul>
<li><a href="https://ncbiinsights.ncbi.nlm.nih.gov/2019/01/04/blast-2-8-1-with-new-databases-and-better-performance/">BLAST+ 2.8.1 is released with new databases</a>,
which allows you to limit your search by taxonomy using information built into the BLAST databases.
So you don't need to build blastdb for specific taxids now.</li>
</ul>
<p>Changes:</p>
<ul>
<li>2018-09-13 rewritten</li>
<li>2018-12-22 providing faster method for step 3.1</li>
<li>2019-01-07 add note of new blastdb version</li>
<li>2020-10-14 update steps for huge number of accessions belong to high taxon level like bacteria.</li>
</ul>
<p>Data:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db">pre-formated blastdb</a> (09/10/2018)</li>
<li><a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz">prot.accession2taxid.gz</a> (09/07/2018) (optional, but recommended)</li>
</ul>
<p>Hardware in this tutorial</p>
<ul>
<li>CPU: AMD 8-cores/16-threads 3.7Ghz</li>
<li>RAM: 64GB</li>
<li>DISK:<ul>
<li>Taxonomy files stores in NVMe SSD</li>
<li>blastdb files stores in 7200rpm HDD</li>
</ul>
</li>
</ul>
<p>Tools:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">blast+</a></li>
<li><a href="https://zlib.net/pigz/">pigz</a> (recommended, faster than gzip)</li>
<li><a href="https://bioinf.shenwei.me/taxonkit">taxonkit</a></li>
<li><a href="https://bioinf.shenwei.me/seqkit">seqkit</a> (recommended),  version &gt;= 0.14.0</li>
<li><a href="https://github.com/shenwei356/rush">rush</a> (optional, for parallizing filtering sequence)</li>
</ul>
<p>Steps:</p>
<ol>
<li>
<p>Listing all taxids below <code>$id</code> using taxonkit.</p>
<pre><code>id=6656

# 6656 is the phylum Arthropoda
# echo 6656 | taxonkit lineage | taxonkit reformat
# 6656    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Protostomia;Ecdysozoa;Panarthropoda;Arthropoda    Eukaryota;Arthropoda;;;;;

# 2     bacteria
# 2157  archaea
# 4751  fungi
# 10239 virus

# time: 2s
taxonkit list --ids $id --indent "" &gt; $id.taxid.txt

# taxonkit list --ids 2,4751,10239 --indent "" &gt; microbe.taxid.txt

wc -l $id.taxid.txt
# 518373 6656.taxid.txt
</code></pre>
</li>
<li>
<p>Retrieving target accessions. There are two options:</p>
<ol>
<li>
<p>From prot.accession2taxid.gz (<strong>faster, recommended</strong>). <em>Note that some accessions are not in <code>nr</code></em>.</p>
<pre><code># time: 4min
pigz -dc prot.accession2taxid.gz \
    | csvtk grep -t -f taxid -P $id.taxid.txt \
    | csvtk cut -t -f accession.version,taxid \
    | sed 1d \
    &gt; $id.acc2taxid.txt

cut -f 1 $id.acc2taxid.txt &gt; $id.acc.txt

wc -l $id.acc.txt
# 8174609 6656.acc.txt
</code></pre>
</li>
<li>
<p>From pre-formated <code>nr</code> blastdb</p>
<pre><code># time: 40min
blastdbcmd -db nr -entry all -outfmt "%a %T" | pigz -c &gt; nr.acc2taxid.txt.gz

pigz -dc nr.acc2taxid.txt.gz | wc -l
# 555220892

# time: 3min
pigz -dc nr.acc2taxid.txt.gz \
    | csvtk grep -d ' ' -D ' ' -f 2 -P $id.taxid.txt \
    | cut -d ' '  -f 1 \
    &gt; $id.acc.txt

wc -l $id.acc.txt
# 6928021 6656.acc.txt
</code></pre>
</li>
</ol>
</li>
<li>
<p>Retrieving FASTA sequences from pre-formated blastdb. There are two options:</p>
<ol>
<li>
<p>From <code>nr.fa</code> exported from pre-formated blastdb (<strong>faster, smaller output file, recommended</strong>).
   <strong>DO NOT directly download <code>nr.gz</code> from <a href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz">ncbi ftp</a></strong>,
   in which the FASTA headers are not well formated.</p>
<pre><code># 1. exporting nr.fa from pre-formated blastdb

# time: 117min (run only once)
blastdbcmd -db nr -dbtype prot -entry all -outfmt "%f" -out - | pigz -c &gt; nr.fa.gz

# =====================================================================

# 2. filtering sequence belong to $taxid

# ---------------------------------------------------------------------

# methond 1) (for cases where $id.acc.txt is not very huge)
# time: 80min
# perl one-liner is used to unfold records having mulitple accessions
time cat &lt;(echo) &lt;(pigz -dc nr.fa.gz) \
    | perl -e 'BEGIN{ $/ = "\n&gt;"; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, "\n"; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print "&gt;$_"; next; }; $h = "&gt;$h"; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print "&gt;$h1\n$s";} } ' \
    | seqkit grep -f $id.acc.txt -o nr.$id.fa.gz

# ---------------------------------------------------------------------

# method 2) (**faster**)

# 33min (run only once)
# (1). split nr.fa.gz. # Note: I have 16 cpus.
$ time seqkit split2 -p 15 nr.fa.gz

# (2). parallize unfolding
$ cat _unfold_blastdb_fa.sh
#!/bin/sh
perl -e 'BEGIN{ $/ = "\n&gt;"; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, "\n"; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print "&gt;$_"; next; }; $h = "&gt;$h"; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print "&gt;$h1\n$s";} } '

# 10 min
time ls nr.fa.gz.split/nr.part_*.fa.gz \
    | rush -j 15 -v id=$id 'cat &lt;(echo) &lt;(pigz -dc {}) \
        | ./_unfold_blastdb_fa.sh \
        | seqkit grep -f {id}.acc.txt -o nr.{id}.{%@nr\.(.+)$} '

# (3). merge result
cat nr.$id.part*.fa.gz &gt; nr.$id.fa.gz
rm nr.$id.part*.fa.gz

# ---------------------------------------------------------------------

# method 3) (for huge $id.acc.txt file, e.g., bacteria)

# (1). split ${id}.acc.txt into several parts. chunk size depends on lines and RAM (64G for me).
split -d -l 300000000 $id.acc.txt $id.acc.txt.part_

# (2). filter
time ls $id.acc.txt.part_* \
    | rush -j 1 --immediate-output -v id=$id \
        'echo {}; cat &lt;(echo) &lt;(pigz -dc nr.fa.gz ) \
        | ./_unfold_blastdb_fa.sh \
        | seqkit grep -f {} -o nr.{id}.{%@(part_.+)}.fa.gz '

# (3). merge
cat nr.$id.part*.fa.gz &gt; nr.$id.fa.gz

# clean
rm nr.$id.part*.fa.gz
rm $id.acc.txt.part_

# (4). optionally adding taxid, you may edit replacement (-r) below
# split
time split -d -l 200000000 $id.acc2taxid.txt $id.acc2taxid.txt.part_

ln -s nr.$id.fa.gz nr.$id.with-taxid.part0.fa.gz         
i=0
for f in $id.acc2taxid.txt.part_* ; do
    echo $f
    time pigz -cd nr.$id.with-taxid.part$i.fa.gz \
        | seqkit replace -k $f -p "^([^\-]+?) " -r "{kv}-\$1 " -K -U -o nr.$id.with-taxid.part$(($i+1)).fa.gz;
    /bin/rm nr.$id.with-taxid.part$i.fa.gz
    i=$(($i+1));
done
mv nr.$id.with-taxid.part$i.fa.gz nr.$id.with-taxid.fa.gz

# =====================================================================

# 3. counting sequences
#
# ls -lh nr.$id.fa.gz
# -rw-r--r-- 1 shenwei shenwei 902M 9月  13 01:42 nr.6656.fa.gz
#
pigz -dc nr.$id.fa.gz | grep '^&gt;' -c

# 6928017
# Here 6928017 ~=  6928021 ($id.acc.txt)
</code></pre>
</li>
<li>
<p>Directly from pre-formated blastdb</p>
<pre><code># time: 5h20min
blastdbcmd -db nr -entry_batch $id.acc.txt -out - | pigz -c &gt; nr.$id.fa.gz

# counting sequences
#
# Note that the headers of outputed fasta by blastdbcmd are "folded"
# for accessions from different species with same sequences, so the
# number may be small than $(wc -l $id.acc.txt).
pigz -dc nr.$id.fa.gz | grep '^&gt;' -c
# 1577383

# counting accessions
#
# ls -lh nr.$id.fa.gz
# -rw-r--r-- 1 shenwei shenwei 2.1G 9月  13 03:38 nr.6656.fa.gz
#
pigz -dc nr.$id.fa.gz | grep '^&gt;' | sed 's/&gt;/\n&gt;/g' | grep '^&gt;' -c
# 288415413
</code></pre>
</li>
</ol>
</li>
<li>
<p>makeblastdb</p>
<pre><code>pigz -dc nr.$id.fa.gz &gt; nr.$id.fa

# time: 3min ($nr.$id.fa from step 3 option 1)
#
# building $nr.$id.fa from step 3 option 2 with -parse_seqids would produce error:
#
#     BLAST Database creation error: Error: Duplicate seq_ids are found: SP|P29868.1
#
makeblastdb -parse_seqids -in nr.$id.fa -dbtype prot -out nr.$id

# rm nr.$id.fa
</code></pre>
</li>
<li>
<p>blastp (optional)</p>
<pre><code># blastdb nr.$id is built from sequences in step 3 option 1
#
blastp -num_threads 16 -db nr.$id -query t4.fa &gt; t4.fa.blast
# real    0m20.866s

# $ cat t4.fa.blast | grep Query= -A 10
# Query= A0A0J9X1W9.2 RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-Hd1a
#
# Length=35
                                                                     Score     E
# Sequences producing significant alignments:                          (Bits)  Value

# 2MPQ_A  Chain A, Solution structure of the sodium channel toxin Hd1a  72.4    2e-17
# A0A0J9X1W9.2  RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-...  72.4    2e-17
# ADB56726.1  HNTX-IV.2 precursor [Haplopelma hainanum]                 66.6    9e-15
# D2Y233.1  RecName: Full=Mu-theraphotoxin-Hhn1b 2; Short=Mu-TRTX-H...  66.6    9e-15
# ADB56830.1  HNTX-IV.3 precursor [Haplopelma hainanum]                 66.6    9e-15
</code></pre>
</li>
</ol>
<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//taxonkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../usage/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Usage
              </div>
            </div>
          </a>
        
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/vendor.93c04032.min.js"></script>
      <script src="../assets/javascripts/bundle.83e5331e.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "..",
          features: [],
          search: Object.assign({
            worker: "../assets/javascripts/worker/search.8c7e0a7e.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>