
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/taxonkit/tutorial/">
      
      <link rel="icon" href="../files/favicon.ico">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.5">
    
    
      
        <title>Tutorial - TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.ee0f47ba.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#00bdd6">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../download/" class="md-nav__link">
        Download
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        Usage
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Tutorial
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formatting-lineage" class="md-nav__link">
    Formatting lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bench/" class="md-nav__link">
        Benchmark
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../chinese/" class="md-nav__link">
        中文介绍
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../chinese-dev/" class="md-nav__link">
        开发笔记
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
        Taxid-Changelog
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        More tools
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formatting-lineage" class="md-nav__link">
    Formatting lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/shenwei356/taxonkit/edit/master/docs/tutorial.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<h1 id="tutorial">Tutorial</h1>
<h2 id="table-of-contents">Table of Contents</h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#formatting-lineage">Formatting lineage</a></li>
<li><a href="#parsing-krakenbraken-result">Parsing kraken/bracken result</a></li>
<li><a href="#making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids</a></li>
<li><a href="#summaries-of-taxonomy-data">Summaries of taxonomy data</a></li>
<li><a href="#merging-gtdb-and-ncbi-taxonomy">Merging GTDB and NCBI taxonomy</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="formatting-lineage">Formatting lineage</h2>
<p>Show lineage detail of a TaxId.
The command below works on Windows with help of <a href="http://bioinf.shenwei.me/csvtk">csvtk</a>.</p>
<pre><code>$ echo "2697049" \
    | taxonkit lineage -t \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht

10239     superkingdom   Viruses
2559587   clade          Riboviria
2732396   kingdom        Orthornavirae
2732408   phylum         Pisuviricota
2732506   class          Pisoniviricetes
76804     order          Nidovirales
2499399   suborder       Cornidovirineae
11118     family         Coronaviridae
2501931   subfamily      Orthocoronavirinae
694002    genus          Betacoronavirus
2509511   subgenus       Sarbecovirus
694009    species        Severe acute respiratory syndrome-related coronavirus
2697049   no rank        Severe acute respiratory syndrome coronavirus 2
</code></pre>
<p>Example data.</p>
<pre><code>$ cat taxids3.txt
376619
349741
239935
314101
11932
1327037
83333
1408252
2605619
2697049
</code></pre>
<p>Format to 7-level ranks ("superkingdom phylum class order family genus species").</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit reformat -I 1

376619  Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis
349741  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
314101  Bacteria;;;;;;uncultured murine large bowel bacterium BAC 54B
11932   Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
1327037 Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;;Croceibacter phage P2559Y
83333   Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli
1408252 Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli
2605619 Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli
2697049 Viruses;Pisuviricota;Pisoniviricetes;Nidovirales;Coronaviridae;Betacoronavirus;Severe acute respiratory syndrome-related coronavirus
</code></pre>
<p>Format to 8-level ranks ("superkingdom phylum class order family genus species subspecies/rank").</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit reformat -I 1 -f "{k};{p};{c};{o};{f};{g};{s};{t}"

376619  Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis;Francisella tularensis subsp. holarctica LVS
349741  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;
314101  Bacteria;;;;;;uncultured murine large bowel bacterium BAC 54B;
11932   Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle;
1327037 Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;;Croceibacter phage P2559Y;
83333   Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli;Escherichia coli K-12
1408252 Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli;Escherichia coli R178
2605619 Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli;
2697049 Viruses;Pisuviricota;Pisoniviricetes;Nidovirales;Coronaviridae;Betacoronavirus;Severe acute respiratory syndrome-related coronavirus;
</code></pre>
<p>Replace missing ranks with <code>Unassigned</code> and output tab-delimited format.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit reformat -I 1 -r "Unassigned" -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    | csvtk pretty -H -t

376619    Bacteria   Proteobacteria    Gammaproteobacteria   Thiotrichales        Francisellaceae      Francisella                  Francisella tularensis                                  Francisella tularensis subsp. holarctica LVS
349741    Bacteria   Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae      Akkermansia                  Akkermansia muciniphila                                 Akkermansia muciniphila ATCC BAA-835
239935    Bacteria   Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae      Akkermansia                  Akkermansia muciniphila                                 Unassigned
314101    Bacteria   Unassigned        Unassigned            Unassigned           Unassigned           Unassigned                   uncultured murine large bowel bacterium BAC 54B         Unassigned
11932     Viruses    Artverviricota    Revtraviricetes       Ortervirales         Retroviridae         Intracisternal A-particles   Mouse Intracisternal A-particle                         Unassigned
1327037   Viruses    Uroviricota       Caudoviricetes        Caudovirales         Siphoviridae         Unassigned                   Croceibacter phage P2559Y                               Unassigned
83333     Bacteria   Proteobacteria    Gammaproteobacteria   Enterobacterales     Enterobacteriaceae   Escherichia                  Escherichia coli                                        Escherichia coli K-12
1408252   Bacteria   Proteobacteria    Gammaproteobacteria   Enterobacterales     Enterobacteriaceae   Escherichia                  Escherichia coli                                        Escherichia coli R178
2605619   Bacteria   Proteobacteria    Gammaproteobacteria   Enterobacterales     Enterobacteriaceae   Escherichia                  Escherichia coli                                        Unassigned
2697049   Viruses    Pisuviricota      Pisoniviricetes       Nidovirales          Coronaviridae        Betacoronavirus              Severe acute respiratory syndrome-related coronavirus   Unassigned
</code></pre>
<p>Fill missing ranks and add prefixes.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit reformat -I 1 -F -P -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    | csvtk pretty -H -t

376619    k__Bacteria   p__Proteobacteria                 c__Gammaproteobacteria           o__Thiotrichales                 f__Francisellaceae                g__Francisella                       s__Francisella tularensis                                  t__Francisella tularensis subsp. holarctica LVS
349741    k__Bacteria   p__Verrucomicrobia                c__Verrucomicrobiae              o__Verrucomicrobiales            f__Akkermansiaceae                g__Akkermansia                       s__Akkermansia muciniphila                                 t__Akkermansia muciniphila ATCC BAA-835
239935    k__Bacteria   p__Verrucomicrobia                c__Verrucomicrobiae              o__Verrucomicrobiales            f__Akkermansiaceae                g__Akkermansia                       s__Akkermansia muciniphila                                 t__unclassified Akkermansia muciniphila subspecies/strain
314101    k__Bacteria   p__unclassified Bacteria phylum   c__unclassified Bacteria class   o__unclassified Bacteria order   f__unclassified Bacteria family   g__unclassified Bacteria genus       s__uncultured murine large bowel bacterium BAC 54B         t__unclassified uncultured murine large bowel bacterium BAC 54B subspecies/strain
11932     k__Viruses    p__Artverviricota                 c__Revtraviricetes               o__Ortervirales                  f__Retroviridae                   g__Intracisternal A-particles        s__Mouse Intracisternal A-particle                         t__unclassified Mouse Intracisternal A-particle subspecies/strain
1327037   k__Viruses    p__Uroviricota                    c__Caudoviricetes                o__Caudovirales                  f__Siphoviridae                   g__unclassified Siphoviridae genus   s__Croceibacter phage P2559Y                               t__unclassified Croceibacter phage P2559Y subspecies/strain
83333     k__Bacteria   p__Proteobacteria                 c__Gammaproteobacteria           o__Enterobacterales              f__Enterobacteriaceae             g__Escherichia                       s__Escherichia coli                                        t__Escherichia coli K-12
1408252   k__Bacteria   p__Proteobacteria                 c__Gammaproteobacteria           o__Enterobacterales              f__Enterobacteriaceae             g__Escherichia                       s__Escherichia coli                                        t__Escherichia coli R178
2605619   k__Bacteria   p__Proteobacteria                 c__Gammaproteobacteria           o__Enterobacterales              f__Enterobacteriaceae             g__Escherichia                       s__Escherichia coli                                        t__unclassified Escherichia coli subspecies/strain
2697049   k__Viruses    p__Pisuviricota                   c__Pisoniviricetes               o__Nidovirales                   f__Coronaviridae                  g__Betacoronavirus                   s__Severe acute respiratory syndrome-related coronavirus   t__unclassified Severe acute respiratory syndrome-related coronavirus subspecies/strain
</code></pre>
<p><strong>When these's no nodes of rank "subspecies" nor "strain",
we can switch <code>-S/--pseudo-strain</code> to use the node with lowest rank
as subspecies/strain name, if which rank is lower than "species"</strong>.</p>
<pre><code>$ cat taxids3.txt \
    | taxonkit lineage -r -L \
    | taxonkit reformat -I 1 -F -S -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    | cut -f 1,2,9,10 \
    | csvtk add-header -t -n "taxid,rank,species,strain" \
    | csvtk pretty -t

taxid     rank         species                                                 strain
-------   ----------   -----------------------------------------------------   ------------------------------------------------------------------------------
376619    strain       Francisella tularensis                                  Francisella tularensis subsp. holarctica LVS
349741    strain       Akkermansia muciniphila                                 Akkermansia muciniphila ATCC BAA-835
239935    species      Akkermansia muciniphila                                 unclassified Akkermansia muciniphila subspecies/strain
314101    species      uncultured murine large bowel bacterium BAC 54B         unclassified uncultured murine large bowel bacterium BAC 54B subspecies/strain
11932     species      Mouse Intracisternal A-particle                         unclassified Mouse Intracisternal A-particle subspecies/strain
1327037   species      Croceibacter phage P2559Y                               unclassified Croceibacter phage P2559Y subspecies/strain
83333     strain       Escherichia coli                                        Escherichia coli K-12
1408252   subspecies   Escherichia coli                                        Escherichia coli R178
2605619   no rank      Escherichia coli                                        Escherichia coli O16:H48
2697049   no rank      Severe acute respiratory syndrome-related coronavirus   Severe acute respiratory syndrome coronavirus 2
</code></pre>
<p>List eight-level lineage for all TaxIds of rank lower than or equal to species, including some nodes with "no rank".
But when filtering with <code>-L/--lower-than</code>, you can use
<code>-n/--save-predictable-norank</code> to <strong>save some special ranks without order,
where rank of the closest higher node is still lower than rank cutoff</strong>.</p>
<pre><code>$ time taxonkit list --ids 1 \
    | taxonkit filter -L species -E species -R -N -n \
    | taxonkit lineage -n -r -L \
    | taxonkit reformat -I 1 -F -S -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    | csvtk cut -Ht -l -f 1,3,2,1,4-11 \
    | csvtk add-header -t -n "taxid,rank,name,lineage,kingdom,phylum,class,order,family,genus,species,strain" \
    | pigz -c &gt; result.tsv.gz

real    0m25.167s
user    2m14.809s
sys     0m7.197s

$ pigz -cd result.tsv.gz \
    | csvtk grep -t -f taxid -p 2697049 \
    | csvtk transpose -t \
    | csvtk pretty -H -t

taxid     2697049
rank      no rank
name      Severe acute respiratory syndrome coronavirus 2
lineage   Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2
kingdom   Viruses
phylum    Pisuviricota
class     Pisoniviricetes
order     Nidovirales
family    Coronaviridae
genus     Betacoronavirus
species   Severe acute respiratory syndrome-related coronavirus
strain    Severe acute respiratory syndrome coronavirus 2
</code></pre>
<h2 id="parsing-krakenbracken-result">Parsing kraken/bracken result</h2>
<p>Example Data</p>
<ul>
<li><a href="https://github.com/biobakery/biobakery/raw/master/demos/biobakery_demos/data/metaphlan2/input/SRS014459-Stool.fasta.gz">SRS014459-Stool.fasta.gz</a></li>
</ul>
<p>Run Kraken2 and Bracken</p>
<pre><code>KRAKEN_DB=/home/shenwei/ws/db/kraken/k2_pluspf
THREADS=16

CLASSIFICATION_LVL=S
THRESHOLD=10

READ_LEN=100
SAMPLE=SRS014459-Stool.fasta.gz

BRACKEN_OUTPUT_FILE=$SAMPLE

kraken2 --db ${KRAKEN_DB} --threads ${THREADS} -report ${SAMPLE}.kreport $SAMPLE &gt; ${SAMPLE}.kraken

est_abundance.py -i ${SAMPLE}.kreport -k ${KRAKEN_DB}/database${READ_LEN}mers.kmer_distrib \
    -l ${CLASSIFICATION_LVL} -t ${THRESHOLD} -o ${BRACKEN_OUTPUT_FILE}.bracken
</code></pre>
<p>Orignial format</p>
<pre><code>$ head -n 15 SRS014459-Stool.fasta.gz_bracken_species.kreport
100.00  9491    0       R       1       root
99.85   9477    0       R1      131567    cellular organisms
99.85   9477    0       D       2           Bacteria
66.08   6271    0       D1      1783270       FCB group
66.08   6271    0       D2      68336           Bacteroidetes/Chlorobi group
66.08   6271    0       P       976               Bacteroidetes
66.08   6271    0       C       200643              Bacteroidia
66.08   6271    0       O       171549                Bacteroidales
34.45   3270    0       F       815                     Bacteroidaceae
34.45   3270    0       G       816                       Bacteroides
10.43   990     990     S       246787                      Bacteroides cellulosilyticus
7.98    757     757     S       28116                       Bacteroides ovatus
3.10    293     0       G1      2646097                     unclassified Bacteroides
1.06    100     100     S       2755405                       Bacteroides sp. CACC 737
0.49    46      46      S       2650157                       Bacteroides sp. HF-5287
</code></pre>
<p>Converting to MetaPhlAn2 format. (Similar to <a href="https://github.com/jenniferlu717/KrakenTools/blob/master/kreport2mpa.py">kreport2mpa.py</a>)</p>
<pre><code>$ cat SRS014459-Stool.fasta.gz_bracken_species.kreport \
    | csvtk cut -Ht -f 5,1 \
    | taxonkit lineage \
    | taxonkit reformat -i 3 -P -f "{k}|{p}|{c}|{o}|{f}|{g}|{s}" \
    | csvtk cut -Ht -f 4,2 \
    | csvtk replace -Ht -p "(\|[kpcofgs]__)+$" \
    | csvtk replace -Ht -p "\|[kpcofgs]__\|" -r "|" \
    | csvtk uniq -Ht \
    | csvtk grep -Ht -p k__ -v \
    &gt; SRS014459-Stool.fasta.gz_bracken_species.kreport.format

$ head -n 10 SRS014459-Stool.fasta.gz_bracken_species.kreport.format

k__Bacteria     99.85
k__Bacteria|p__Bacteroidetes    66.08
k__Bacteria|p__Bacteroidetes|c__Bacteroidia     66.08
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales    66.08
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae  34.45
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides   34.45
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides cellulosilyticus   10.43
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides ovatus     7.98
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides sp. CACC 737       1.06
k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides sp. HF-5287        0.49
</code></pre>
<p>Converting to Qiime format</p>
<pre><code>$ cat SRS014459-Stool.fasta.gz_bracken_species.kreport \
    | csvtk cut -Ht -f 5,1 \
    | taxonkit lineage \
    | taxonkit reformat -i 3 -P -f "{k}; {p}; {c}; {o}; {f}; {g}; {s}" \
    | csvtk cut -Ht -f 4,2 \
    | csvtk replace -Ht -p "(; [kpcofgs]__)+$" \
    | csvtk replace -Ht -p "; [kpcofgs]__; " -r "; " \
    | csvtk uniq -Ht \
    | csvtk grep -Ht -p k__ -v \
    | head -n 10

k__Bacteria     99.85
k__Bacteria; p__Bacteroidetes   66.08
k__Bacteria; p__Bacteroidetes; c__Bacteroidia   66.08
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales 66.08
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae      34.45
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae; g__Bacteroides      34.45
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae; g__Bacteroides; s__Bacteroides cellulosilyticus     10.43
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae; g__Bacteroides; s__Bacteroides ovatus       7.98
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae; g__Bacteroides; s__Bacteroides sp. CACC 737 1.06
k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Bacteroidaceae; g__Bacteroides; s__Bacteroides sp. HF-5287  0.49
</code></pre>
<p>Save taxon proportion and taxid, and get lineage, name and rank.</p>
<pre><code>$ cat SRS014459-Stool.fasta.gz_bracken_species.kreport \
    | csvtk cut -Ht -f 1,5 \
    | taxonkit lineage -i 2 -n -r \
    | csvtk cut -Ht -f 1,2,5,4,3 \
    | head -n 10 \
    | csvtk pretty -Ht

100.00   1         no rank        root                           root
99.85    131567    no rank        cellular organisms             cellular organisms
99.85    2         superkingdom   Bacteria                       cellular organisms;Bacteria
66.08    1783270   clade          FCB group                      cellular organisms;Bacteria;FCB group
66.08    68336     clade          Bacteroidetes/Chlorobi group   cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group
66.08    976       phylum         Bacteroidetes                  cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group;Bacteroidetes
66.08    200643    class          Bacteroidia                    cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group;Bacteroidetes;Bacteroidia
66.08    171549    order          Bacteroidales                  cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group;Bacteroidetes;Bacteroidia;Bacteroidales
34.45    815       family         Bacteroidaceae                 cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae
34.45    816       genus          Bacteroides                    cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi group;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides
</code></pre>
<p>Only save species or lower level and get lineage in format of "superkingdom phylum class order family genus species".</p>
<pre><code>$ cat SRS014459-Stool.fasta.gz_bracken_species.kreport \
    | csvtk cut -Ht -f 1,5 \
    | taxonkit filter -N -E species -L species -i 2 \
    | taxonkit lineage -i 2 -n -r \
    | taxonkit reformat -i 3 -f "{k};{p};{c};{o};{f};{g};{s}" \
    | csvtk cut -Ht -f 1,2,5,4,6 \
    | csvtk add-header -t -n abundance,taxid,rank,name,lineage \
    | head -n 10 \
    | csvtk pretty -t

abundance   taxid     rank      name                           lineage
---------   -------   -------   ----------------------------   --------------------------------------------------------------------------------------------------------
10.43       246787    species   Bacteroides cellulosilyticus   Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides cellulosilyticus
7.98        28116     species   Bacteroides ovatus             Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides ovatus
1.06        2755405   species   Bacteroides sp. CACC 737       Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. CACC 737
0.49        2650157   species   Bacteroides sp. HF-5287        Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. HF-5287
0.99        2528203   species   Bacteroides sp. A1C1           Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. A1C1
0.28        2763022   species   Bacteroides sp. M10            Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. M10
0.16        2650158   species   Bacteroides sp. HF-5141        Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. HF-5141
0.12        2715212   species   Bacteroides sp. CBA7301        Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides sp. CBA7301
5.10        817       species   Bacteroides fragilis           Bacteria;Bacteroidetes;Bacteroidia;Bacteroidales;Bacteroidaceae;Bacteroides;Bacteroides fragilis
</code></pre>
<h2 id="making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids</h2>
<p>Attention:</p>
<ul>
<li><a href="https://ncbiinsights.ncbi.nlm.nih.gov/2019/01/04/blast-2-8-1-with-new-databases-and-better-performance/">BLAST+ 2.8.1 is released with new databases</a>,
which allows you to limit your search by taxonomy using information built into the BLAST databases.
So you don't need to build blastdb for specific taxids now.</li>
</ul>
<p>Changes:</p>
<ul>
<li>2018-09-13 rewritten</li>
<li>2018-12-22 providing faster method for step 3.1</li>
<li>2019-01-07 add note of new blastdb version</li>
<li>2020-10-14 update steps for huge number of accessions belong to high taxon level like bacteria.</li>
</ul>
<p>Data:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db">pre-formated blastdb</a> (09/10/2018)</li>
<li><a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz">prot.accession2taxid.gz</a> (09/07/2018) (optional, but recommended)</li>
</ul>
<p>Hardware in this tutorial</p>
<ul>
<li>CPU: AMD 8-cores/16-threads 3.7Ghz</li>
<li>RAM: 64GB</li>
<li>DISK:<ul>
<li>Taxonomy files stores in NVMe SSD</li>
<li>blastdb files stores in 7200rpm HDD</li>
</ul>
</li>
</ul>
<p>Tools:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">blast+</a></li>
<li><a href="https://zlib.net/pigz/">pigz</a> (recommended, faster than gzip)</li>
<li><a href="https://bioinf.shenwei.me/taxonkit">taxonkit</a></li>
<li><a href="https://bioinf.shenwei.me/seqkit">seqkit</a> (recommended),  version &gt;= 0.14.0</li>
<li><a href="https://github.com/shenwei356/rush">rush</a> (optional, for parallizing filtering sequence)</li>
</ul>
<p>Steps:</p>
<ol>
<li>
<p>Listing all taxids below <code>$id</code> using taxonkit.</p>
<pre><code>id=6656

# 6656 is the phylum Arthropoda
# echo 6656 | taxonkit lineage | taxonkit reformat
# 6656    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Protostomia;Ecdysozoa;Panarthropoda;Arthropoda    Eukaryota;Arthropoda;;;;;

# 2     bacteria
# 2157  archaea
# 4751  fungi
# 10239 virus

# time: 2s
taxonkit list --ids $id --indent "" &gt; $id.taxid.txt

# taxonkit list --ids 2,4751,10239 --indent "" &gt; microbe.taxid.txt

wc -l $id.taxid.txt
# 518373 6656.taxid.txt
</code></pre>
</li>
<li>
<p>Retrieving target accessions. There are two options:</p>
<ol>
<li>
<p>From prot.accession2taxid.gz (<strong>faster, recommended</strong>). <em>Note that some accessions are not in <code>nr</code></em>.</p>
<pre><code># time: 4min
pigz -dc prot.accession2taxid.gz \
    | csvtk grep -t -f taxid -P $id.taxid.txt \
    | csvtk cut -t -f accession.version,taxid \
    | sed 1d \
    &gt; $id.acc2taxid.txt

cut -f 1 $id.acc2taxid.txt &gt; $id.acc.txt

wc -l $id.acc.txt
# 8174609 6656.acc.txt
</code></pre>
</li>
<li>
<p>From pre-formated <code>nr</code> blastdb</p>
<pre><code># time: 40min
blastdbcmd -db nr -entry all -outfmt "%a %T" | pigz -c &gt; nr.acc2taxid.txt.gz

pigz -dc nr.acc2taxid.txt.gz | wc -l
# 555220892

# time: 3min
pigz -dc nr.acc2taxid.txt.gz \
    | csvtk grep -d ' ' -D ' ' -f 2 -P $id.taxid.txt \
    | cut -d ' '  -f 1 \
    &gt; $id.acc.txt

wc -l $id.acc.txt
# 6928021 6656.acc.txt
</code></pre>
</li>
</ol>
</li>
<li>
<p>Retrieving FASTA sequences from pre-formated blastdb. There are two options:</p>
<ol>
<li>
<p>From <code>nr.fa</code> exported from pre-formated blastdb (<strong>faster, smaller output file, recommended</strong>).
   <strong>DO NOT directly download <code>nr.gz</code> from <a href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz">ncbi ftp</a></strong>,
   in which the FASTA headers are not well formated.</p>
<pre><code># 1. exporting nr.fa from pre-formated blastdb

# time: 117min (run only once)
blastdbcmd -db nr -dbtype prot -entry all -outfmt "%f" -out - | pigz -c &gt; nr.fa.gz

# =====================================================================

# 2. filtering sequence belong to $taxid

# ---------------------------------------------------------------------

# methond 1) (for cases where $id.acc.txt is not very huge)
# time: 80min
# perl one-liner is used to unfold records having mulitple accessions
time cat &lt;(echo) &lt;(pigz -dc nr.fa.gz) \
    | perl -e 'BEGIN{ $/ = "\n&gt;"; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, "\n"; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print "&gt;$_"; next; }; $h = "&gt;$h"; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print "&gt;$h1\n$s";} } ' \
    | seqkit grep -f $id.acc.txt -o nr.$id.fa.gz

# ---------------------------------------------------------------------

# method 2) (**faster**)

# 33min (run only once)
# (1). split nr.fa.gz. # Note: I have 16 cpus.
$ time seqkit split2 -p 15 nr.fa.gz

# (2). parallize unfolding
$ cat _unfold_blastdb_fa.sh
#!/bin/sh
perl -e 'BEGIN{ $/ = "\n&gt;"; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, "\n"; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print "&gt;$_"; next; }; $h = "&gt;$h"; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print "&gt;$h1\n$s";} } '

# 10 min
time ls nr.fa.gz.split/nr.part_*.fa.gz \
    | rush -j 15 -v id=$id 'cat &lt;(echo) &lt;(pigz -dc {}) \
        | ./_unfold_blastdb_fa.sh \
        | seqkit grep -f {id}.acc.txt -o nr.{id}.{%@nr\.(.+)$} '

# (3). merge result
cat nr.$id.part*.fa.gz &gt; nr.$id.fa.gz
rm nr.$id.part*.fa.gz

# ---------------------------------------------------------------------

# method 3) (for huge $id.acc.txt file, e.g., bacteria)

# (1). split ${id}.acc.txt into several parts. chunk size depends on lines and RAM (64G for me).
split -d -l 300000000 $id.acc.txt $id.acc.txt.part_

# (2). filter
time ls $id.acc.txt.part_* \
    | rush -j 1 --immediate-output -v id=$id \
        'echo {}; cat &lt;(echo) &lt;(pigz -dc nr.fa.gz ) \
        | ./_unfold_blastdb_fa.sh \
        | seqkit grep -f {} -o nr.{id}.{%@(part_.+)}.fa.gz '

# (3). merge
cat nr.$id.part*.fa.gz &gt; nr.$id.fa.gz

# clean
rm nr.$id.part*.fa.gz
rm $id.acc.txt.part_

# (4). optionally adding taxid, you may edit replacement (-r) below
# split
time split -d -l 200000000 $id.acc2taxid.txt $id.acc2taxid.txt.part_

ln -s nr.$id.fa.gz nr.$id.with-taxid.part0.fa.gz         
i=0
for f in $id.acc2taxid.txt.part_* ; do
    echo $f
    time pigz -cd nr.$id.with-taxid.part$i.fa.gz \
        | seqkit replace -k $f -p "^([^\-]+?) " -r "{kv}-\$1 " -K -U -o nr.$id.with-taxid.part$(($i+1)).fa.gz;
    /bin/rm nr.$id.with-taxid.part$i.fa.gz
    i=$(($i+1));
done
mv nr.$id.with-taxid.part$i.fa.gz nr.$id.with-taxid.fa.gz

# =====================================================================

# 3. counting sequences
#
# ls -lh nr.$id.fa.gz
# -rw-r--r-- 1 shenwei shenwei 902M 9月  13 01:42 nr.6656.fa.gz
#
pigz -dc nr.$id.fa.gz | grep '^&gt;' -c

# 6928017
# Here 6928017 ~=  6928021 ($id.acc.txt)
</code></pre>
</li>
<li>
<p>Directly from pre-formated blastdb</p>
<pre><code># time: 5h20min
blastdbcmd -db nr -entry_batch $id.acc.txt -out - | pigz -c &gt; nr.$id.fa.gz

# counting sequences
#
# Note that the headers of outputed fasta by blastdbcmd are "folded"
# for accessions from different species with same sequences, so the
# number may be small than $(wc -l $id.acc.txt).
pigz -dc nr.$id.fa.gz | grep '^&gt;' -c
# 1577383

# counting accessions
#
# ls -lh nr.$id.fa.gz
# -rw-r--r-- 1 shenwei shenwei 2.1G 9月  13 03:38 nr.6656.fa.gz
#
pigz -dc nr.$id.fa.gz | grep '^&gt;' | sed 's/&gt;/\n&gt;/g' | grep '^&gt;' -c
# 288415413
</code></pre>
</li>
</ol>
</li>
<li>
<p>makeblastdb</p>
<pre><code>pigz -dc nr.$id.fa.gz &gt; nr.$id.fa

# time: 3min ($nr.$id.fa from step 3 option 1)
#
# building $nr.$id.fa from step 3 option 2 with -parse_seqids would produce error:
#
#     BLAST Database creation error: Error: Duplicate seq_ids are found: SP|P29868.1
#
makeblastdb -parse_seqids -in nr.$id.fa -dbtype prot -out nr.$id

# rm nr.$id.fa
</code></pre>
</li>
<li>
<p>blastp (optional)</p>
<pre><code># blastdb nr.$id is built from sequences in step 3 option 1
#
blastp -num_threads 16 -db nr.$id -query t4.fa &gt; t4.fa.blast
# real    0m20.866s

# $ cat t4.fa.blast | grep Query= -A 10
# Query= A0A0J9X1W9.2 RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-Hd1a
#
# Length=35
                                                                     Score     E
# Sequences producing significant alignments:                          (Bits)  Value

# 2MPQ_A  Chain A, Solution structure of the sodium channel toxin Hd1a  72.4    2e-17
# A0A0J9X1W9.2  RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-...  72.4    2e-17
# ADB56726.1  HNTX-IV.2 precursor [Haplopelma hainanum]                 66.6    9e-15
# D2Y233.1  RecName: Full=Mu-theraphotoxin-Hhn1b 2; Short=Mu-TRTX-H...  66.6    9e-15
# ADB56830.1  HNTX-IV.3 precursor [Haplopelma hainanum]                 66.6    9e-15
</code></pre>
</li>
</ol>
<h2 id="summaries-of-taxonomy-data">Summaries of taxonomy data</h2>
<p>You can change the TaxId of interest.</p>
<ol>
<li>
<p>Rank counts of common categories.</p>
<pre><code>$ echo Archaea Bacteria Eukaryota Fungi Metazoa Viridiplantae \
    | rush -D ' ' -T b \
        'taxonkit list --ids $(echo {} | taxonkit name2taxid | cut -f 2) \
            | sed 1d \
            | taxonkit filter -i 2 -E genus -L genus \
            | taxonkit lineage -L -r \
            | csvtk freq -H -t -f 2 -nr \
            &gt; stats.{}.tsv '

$ csvtk -t join --outer-join stats.*.tsv \
    | csvtk add-header -t -n "rank,$(ls stats.*.tsv | rush -k 'echo {@stats.(.+).tsv}' | paste -sd, )" \
    | csvtk csv2md -t
</code></pre>
<p><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=statistics&amp;uncultured=hide&amp;unspecified=hide">Similar data on NCBI Taxonomy</a></p>
<table>
<thead>
<tr>
<th align="left">rank</th>
<th align="left">Archaea</th>
<th align="left">Bacteria</th>
<th align="left">Eukaryota</th>
<th align="left">Fungi</th>
<th align="left">Metazoa</th>
<th align="left">Viridiplantae</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">species</td>
<td align="left">12482</td>
<td align="left">460940</td>
<td align="left">1349648</td>
<td align="left">156908</td>
<td align="left">957297</td>
<td align="left">191026</td>
</tr>
<tr>
<td align="left">strain</td>
<td align="left">354</td>
<td align="left">40643</td>
<td align="left">3486</td>
<td align="left">2352</td>
<td align="left">33</td>
<td align="left">50</td>
</tr>
<tr>
<td align="left">genus</td>
<td align="left">205</td>
<td align="left">4112</td>
<td align="left">90882</td>
<td align="left">6844</td>
<td align="left">64148</td>
<td align="left">16202</td>
</tr>
<tr>
<td align="left">isolate</td>
<td align="left">7</td>
<td align="left">503</td>
<td align="left">809</td>
<td align="left">76</td>
<td align="left">17</td>
<td align="left">3</td>
</tr>
<tr>
<td align="left">species group</td>
<td align="left">2</td>
<td align="left">77</td>
<td align="left">251</td>
<td align="left">22</td>
<td align="left">214</td>
<td align="left">5</td>
</tr>
<tr>
<td align="left">serotype</td>
<td align="left"></td>
<td align="left">218</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">serogroup</td>
<td align="left"></td>
<td align="left">136</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">subsection</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">21</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">21</td>
</tr>
<tr>
<td align="left">subspecies</td>
<td align="left"></td>
<td align="left">632</td>
<td align="left">24523</td>
<td align="left">158</td>
<td align="left">17043</td>
<td align="left">7212</td>
</tr>
<tr>
<td align="left">forma specialis</td>
<td align="left"></td>
<td align="left">521</td>
<td align="left">220</td>
<td align="left">179</td>
<td align="left">33</td>
<td align="left">1</td>
</tr>
<tr>
<td align="left">species subgroup</td>
<td align="left"></td>
<td align="left">23</td>
<td align="left">101</td>
<td align="left"></td>
<td align="left">101</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">biotype</td>
<td align="left"></td>
<td align="left">7</td>
<td align="left">10</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">morph</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
<td align="left">3</td>
<td align="left">4</td>
<td align="left">5</td>
</tr>
<tr>
<td align="left">section</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">437</td>
<td align="left">37</td>
<td align="left">2</td>
<td align="left">398</td>
</tr>
<tr>
<td align="left">genotype</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">12</td>
</tr>
<tr>
<td align="left">series</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">9</td>
<td align="left"></td>
<td align="left">5</td>
<td align="left">4</td>
</tr>
<tr>
<td align="left">varietas</td>
<td align="left"></td>
<td align="left">25</td>
<td align="left">8499</td>
<td align="left">1100</td>
<td align="left">2</td>
<td align="left">7188</td>
</tr>
<tr>
<td align="left">forma</td>
<td align="left"></td>
<td align="left">4</td>
<td align="left">560</td>
<td align="left">185</td>
<td align="left">6</td>
<td align="left">315</td>
</tr>
<tr>
<td align="left">subgenus</td>
<td align="left"></td>
<td align="left">1</td>
<td align="left">1558</td>
<td align="left">10</td>
<td align="left">1414</td>
<td align="left">112</td>
</tr>
<tr>
<td align="left">pathogroup</td>
<td align="left"></td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">subvariety</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
<td align="left">5</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Count of all ranks</p>
<pre><code>$ time taxonkit list --ids 1 \
    | taxonkit lineage -L -r \
    | csvtk freq -H -t -f 2 -nr \
    | csvtk pretty -H -t

species            1879659
no rank            222743
genus              96625
strain             44483
subspecies         25174
family             9492
varietas           8524
subfamily          3050
tribe              2213
order              1660
subgenus           1618
isolate            1319
serotype           1216
clade              886
superfamily        865
forma specialis    741
forma              564
subtribe           508
section            437
class              429
suborder           372
species group      330
phylum             272
subclass           156
serogroup          138
infraorder         130
species subgroup   124
superorder         55
subphylum          33
parvorder          26
subsection         21
genotype           20
infraclass         18
biotype            17
morph              12
kingdom            11
series             9
superclass         6
cohort             5
pathogroup         5
subvariety         5
superkingdom       4
subcohort          3
subkingdom         1
superphylum        1

real    0m3.663s
user    0m15.897s
sys     0m1.010s
</code></pre>
</li>
<li>
<p>Ranks of taxa at or below species.</p>
<pre><code>$ taxonkit list --ids 1 \
    | taxonkit filter --lower-than species --equal-to species \
    | taxonkit lineage -L -r  \
    | csvtk freq -Ht -nr -f 2 \
    | csvtk add-header -t -n rank,count \
    | csvtk pretty -t

rank              count
---------------   -------
species           1880044
no rank           222756
strain            44483
subspecies        25171
varietas          8524
isolate           1319
serotype          1216
clade             885
forma specialis   741
forma             564
serogroup         138
genotype          20
biotype           17
morph             12
pathogroup        5
subvariety        5
</code></pre>
</li>
</ol>
<h2 id="merging-gtdb-and-ncbi-taxonomy">Merging GTDB and NCBI taxonomy</h2>
<p>Sometimes (<a href="https://github.com/shenwei356/gtdb-taxdump/issues/6">1</a>) one needs to build a database including bacteria and archaea (from GTDB) and viral database from NCBI.
The idea is to export lineages from both GTDB and NCBI using <a href="https://bioinf.shenwei.me/taxonkit/usage/#reformat">taxonkit reformat</a>, and then create taxdump files from them with <a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump">taxonkit create-taxdump</a>.</p>
<ol>
<li>
<p>Exporting taxonomic lineages of taxa with rank equal to species from <a href="https://github.com/shenwei356/gtdb-taxdump">GTDB-taxdump</a>.</p>
<pre><code>taxonkit list --data-dir gtdb-taxdump/R207/ --ids 1 --indent "" \
    | taxonkit filter --data-dir gtdb-taxdump/R207/ --equal-to species \
    | taxonkit reformat --data-dir gtdb-taxdump/R207/ --taxid-field 1 \
        --format "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}" \
        -o gtdb.tsv
</code></pre>
</li>
<li>
<p>Exporting taxonomic lineages of viral taxa with rank equal to or lower than species from NCBI taxdump.
   For taxa whose rank is "no rank" below the species, we treat them as tax of strain rank (<code>--pseudo-strain</code>, taxonkit v0.14.1 needed).</p>
<pre><code># taxid of Viruses: 10239
taxonkit list --data-dir ~/.taxonkit --ids 10239 --indent "" \
    | taxonkit filter --data-dir ~/.taxonkit --equal-to species --lower-than species \
    | taxonkit reformat --data-dir ~/.taxonkit --taxid-field 1 \
        --pseudo-strain --format "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
        -o ncbi-viral.tsv
</code></pre>
</li>
<li>
<p>Creating taxdump from lineages above.</p>
<pre><code>(awk '{print $_"\t"}' gtdb.tsv; cat ncbi-viral.tsv) \
    | taxonkit create-taxdump \
        --field-accession 1 \
        -R "superkingdom,phylum,class,order,family,genus,species,strain" \
        -O taxdump

# we use --field-accession  1 to output the mapping file between old taxids and new ones.
$ grep 2697049  taxdump/taxid.map  # SARS-COV-2
2697049 21630522
</code></pre>
</li>
</ol>
<p>Some tests:</p>
<pre><code># SARS-COV-2 in NCBI taxonomy
$ echo 2697049 \
    | taxonkit lineage -t --data-dir ~/.taxonkit \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir ~/.taxonkit \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
10239     superkingdom   Viruses
2559587   clade          Riboviria
2732396   kingdom        Orthornavirae
2732408   phylum         Pisuviricota
2732506   class          Pisoniviricetes
76804     order          Nidovirales
2499399   suborder       Cornidovirineae
11118     family         Coronaviridae
2501931   subfamily      Orthocoronavirinae
694002    genus          Betacoronavirus
2509511   subgenus       Sarbecovirus
694009    species        Severe acute respiratory syndrome-related coronavirus
2697049   no rank        Severe acute respiratory syndrome coronavirus 2

$ echo "Severe acute respiratory syndrome coronavirus 2" | taxonkit name2taxid --data-dir taxdump/
Severe acute respiratory syndrome coronavirus 2 216305222

$ echo 216305222 \
    | taxonkit lineage -t --data-dir taxdump/ \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
1287770734   superkingdom   Viruses
1506901452   phylum         Pisuviricota
1091693597   class          Pisoniviricetes
37745009     order          Nidovirales
738421640    family         Coronaviridae
906833049    genus          Betacoronavirus
1015862491   species        Severe acute respiratory syndrome-related coronavirus
216305222    strain         Severe acute respiratory syndrome coronavirus 2



$ echo "Escherichia coli"  | taxonkit name2taxid --data-dir taxdump/
Escherichia coli        1945799576

$ echo 1945799576 \
    | taxonkit lineage -t --data-dir taxdump/ \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
609216830    superkingdom   Bacteria
1641076285   phylum         Proteobacteria
329474883    class          Gammaproteobacteria
1012954932   order          Enterobacterales
87250111     family         Enterobacteriaceae
1187493883   genus          Escherichia
1945799576   species        Escherichia coli
</code></pre>
<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//taxonkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../usage/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Usage" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Usage
            </div>
          </div>
        </a>
      
      
        
        <a href="../chinese/" class="md-footer__link md-footer__link--next" aria-label="Next: 中文介绍" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              中文介绍
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.b6ff8c07.min.js"}</script>
    
    
      <script src="../assets/javascripts/bundle.5e13d264.min.js"></script>
      
    
  </body>
</html>