
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/taxonkit/tutorial/">
      
      
        <link rel="prev" href="../usage/">
      
      
        <link rel="next" href="../chinese/">
      
      
      <link rel="icon" href="../files/favicon.ico">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.8">
    
    
      
        <title>Tutorial - TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.356b1318.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Tutorial
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Tutorial
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formatting-lineage" class="md-nav__link">
    Formatting lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mapping-old-species-names-to-new-ones" class="md-nav__link">
    Mapping old species names to new ones
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#add-taxonomy-information-to-blast-result" class="md-nav__link">
    Add taxonomy information to BLAST result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filtering-or-subsetting-taxdmp-files-to-make-a-custom-taxdmp-with-given-taxids" class="md-nav__link">
    Filtering or subsetting taxdmp files to make a custom taxdmp with given TaxIDs
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../bench/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Benchmark
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../chinese/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    中文介绍
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../chinese-dev/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    开发笔记
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Taxid-Changelog
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    More tools
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#formatting-lineage" class="md-nav__link">
    Formatting lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mapping-old-species-names-to-new-ones" class="md-nav__link">
    Mapping old species names to new ones
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#add-taxonomy-information-to-blast-result" class="md-nav__link">
    Add taxonomy information to BLAST result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsing-krakenbracken-result" class="md-nav__link">
    Parsing kraken/bracken result
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-nr-blastdb-for-specific-taxids" class="md-nav__link">
    Making nr blastdb for specific taxids
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summaries-of-taxonomy-data" class="md-nav__link">
    Summaries of taxonomy data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filtering-or-subsetting-taxdmp-files-to-make-a-custom-taxdmp-with-given-taxids" class="md-nav__link">
    Filtering or subsetting taxdmp files to make a custom taxdmp with given TaxIDs
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="tutorial">Tutorial<a class="headerlink" href="#tutorial" title="Permanent link">Link</a></h1>
<h2 id="table-of-contents">Table of Contents<a class="headerlink" href="#table-of-contents" title="Permanent link">Link</a></h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#formatting-lineage">Formatting lineage</a></li>
<li><a href="#parsing-krakenbraken-result">Parsing kraken/bracken result</a></li>
<li><a href="#making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids</a></li>
<li><a href="#summaries-of-taxonomy-data">Summaries of taxonomy data</a></li>
<li><a href="#merging-gtdb-and-ncbi-taxonomy">Merging GTDB and NCBI taxonomy</a></li>
<li><a href="#filtering-or-subsetting-taxdmp-files-to-make-a-custom-taxdmp-with-given-taxids">Filtering or subsetting taxdmp files to make a custom taxdmp with given TaxIDs</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="formatting-lineage">Formatting lineage<a class="headerlink" href="#formatting-lineage" title="Permanent link">Link</a></h2>
<p>Show lineage detail of a TaxId.
The command below works on Windows with help of <a href="http://bioinf.shenwei.me/csvtk">csvtk</a>.</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="m">2697049</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-t<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>unfold<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w"> </span>-s<span class="w"> </span><span class="s2">&quot;;&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-r<span class="w"> </span>-n<span class="w"> </span>-L<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,3,2<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-Ht

<span class="m">10239</span><span class="w">     </span>acellular<span class="w"> </span>root<span class="w">   </span>Viruses<span class="w">                                        </span>
<span class="m">2559587</span><span class="w">   </span>realm<span class="w">            </span>Riboviria<span class="w">                                      </span>
<span class="m">2732396</span><span class="w">   </span>kingdom<span class="w">          </span>Orthornavirae<span class="w">                                  </span>
<span class="m">2732408</span><span class="w">   </span>phylum<span class="w">           </span>Pisuviricota<span class="w">                                   </span>
<span class="m">2732506</span><span class="w">   </span>class<span class="w">            </span>Pisoniviricetes<span class="w">                                </span>
<span class="m">76804</span><span class="w">     </span>order<span class="w">            </span>Nidovirales<span class="w">                                    </span>
<span class="m">2499399</span><span class="w">   </span>suborder<span class="w">         </span>Cornidovirineae<span class="w">                                </span>
<span class="m">11118</span><span class="w">     </span>family<span class="w">           </span>Coronaviridae<span class="w">                                  </span>
<span class="m">2501931</span><span class="w">   </span>subfamily<span class="w">        </span>Orthocoronavirinae<span class="w">                             </span>
<span class="m">694002</span><span class="w">    </span>genus<span class="w">            </span>Betacoronavirus<span class="w">                                </span>
<span class="m">2509511</span><span class="w">   </span>subgenus<span class="w">         </span>Sarbecovirus<span class="w">                                   </span>
<span class="m">3418604</span><span class="w">   </span>species<span class="w">          </span>Betacoronavirus<span class="w"> </span>pandemicum<span class="w">                     </span>
<span class="m">2697049</span><span class="w">   </span>no<span class="w"> </span>rank<span class="w">          </span>Severe<span class="w"> </span>acute<span class="w"> </span>respiratory<span class="w"> </span>syndrome<span class="w"> </span>coronavirus<span class="w"> </span><span class="m">2</span>
</code></pre></div>

<p>Example data.</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt
<span class="m">376619</span>
<span class="m">349741</span>
<span class="m">239935</span>
<span class="m">314101</span>
<span class="m">11932</span>
<span class="m">1327037</span>
<span class="m">83333</span>
<span class="m">1408252</span>
<span class="m">2605619</span>
<span class="m">2697049</span>
</code></pre></div>

<p>Format to 7-level ranks ("superkingdom phylum class order family genus species").</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat2<span class="w"> </span>-I<span class="w"> </span><span class="m">1</span>

<span class="m">376619</span><span class="w">  </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Thiotrichales<span class="p">;</span>Francisellaceae<span class="p">;</span>Francisella<span class="p">;</span>Francisella<span class="w"> </span>tularensis
<span class="m">349741</span><span class="w">  </span>Bacteria<span class="p">;</span>Verrucomicrobiota<span class="p">;</span>Verrucomicrobiia<span class="p">;</span>Verrucomicrobiales<span class="p">;</span>Akkermansiaceae<span class="p">;</span>Akkermansia<span class="p">;</span>Akkermansia<span class="w"> </span>muciniphila
<span class="m">239935</span><span class="w">  </span>Bacteria<span class="p">;</span>Verrucomicrobiota<span class="p">;</span>Verrucomicrobiia<span class="p">;</span>Verrucomicrobiales<span class="p">;</span>Akkermansiaceae<span class="p">;</span>Akkermansia<span class="p">;</span>Akkermansia<span class="w"> </span>muciniphila
<span class="m">314101</span><span class="w">  </span>Bacteria<span class="p">;;;;;;</span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B
<span class="m">11932</span><span class="w">   </span>Viruses<span class="p">;</span>Artverviricota<span class="p">;</span>Revtraviricetes<span class="p">;</span>Ortervirales<span class="p">;</span>Retroviridae<span class="p">;</span>Intracisternal<span class="w"> </span>A-particles<span class="p">;</span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle
<span class="m">1327037</span><span class="w"> </span>Viruses<span class="p">;</span>Uroviricota<span class="p">;</span>Caudoviricetes<span class="p">;;;;</span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y
<span class="m">83333</span><span class="w">   </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli
<span class="m">1408252</span><span class="w"> </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli
<span class="m">2605619</span><span class="w"> </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli
<span class="m">2697049</span><span class="w"> </span>Viruses<span class="p">;</span>Pisuviricota<span class="p">;</span>Pisoniviricetes<span class="p">;</span>Nidovirales<span class="p">;</span>Coronaviridae<span class="p">;</span>Betacoronavirus<span class="p">;</span>Betacoronavirus<span class="w"> </span>pandemicum
</code></pre></div>

<p>Format to 8-level ranks ("superkingdom phylum class order family genus species subspecies/rank").</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat2<span class="w"> </span>-I<span class="w"> </span><span class="m">1</span><span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;{domain|acellular root|superkingdom};{phylum};{class};{order};{family};{genus};{species};{strain|subspecies|no rank}&quot;</span>

<span class="m">376619</span><span class="w">  </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Thiotrichales<span class="p">;</span>Francisellaceae<span class="p">;</span>Francisella<span class="p">;</span>Francisella<span class="w"> </span>tularensis<span class="p">;</span>Francisella<span class="w"> </span>tularensis<span class="w"> </span>subsp.<span class="w"> </span>holarctica<span class="w"> </span>LVS
<span class="m">349741</span><span class="w">  </span>Bacteria<span class="p">;</span>Verrucomicrobiota<span class="p">;</span>Verrucomicrobiia<span class="p">;</span>Verrucomicrobiales<span class="p">;</span>Akkermansiaceae<span class="p">;</span>Akkermansia<span class="p">;</span>Akkermansia<span class="w"> </span>muciniphila<span class="p">;</span>Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>ATCC<span class="w"> </span>BAA-835
<span class="m">239935</span><span class="w">  </span>Bacteria<span class="p">;</span>Verrucomicrobiota<span class="p">;</span>Verrucomicrobiia<span class="p">;</span>Verrucomicrobiales<span class="p">;</span>Akkermansiaceae<span class="p">;</span>Akkermansia<span class="p">;</span>Akkermansia<span class="w"> </span>muciniphila<span class="p">;</span>
<span class="m">314101</span><span class="w">  </span>Bacteria<span class="p">;;;;;;</span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="p">;</span>environmental<span class="w"> </span>samples
<span class="m">11932</span><span class="w">   </span>Viruses<span class="p">;</span>Artverviricota<span class="p">;</span>Revtraviricetes<span class="p">;</span>Ortervirales<span class="p">;</span>Retroviridae<span class="p">;</span>Intracisternal<span class="w"> </span>A-particles<span class="p">;</span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="p">;</span>unclassified<span class="w"> </span>Retroviridae
<span class="m">1327037</span><span class="w"> </span>Viruses<span class="p">;</span>Uroviricota<span class="p">;</span>Caudoviricetes<span class="p">;;;;</span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="p">;</span>unclassified<span class="w"> </span>Caudoviricetes
<span class="m">83333</span><span class="w">   </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="w"> </span>K-12
<span class="m">1408252</span><span class="w"> </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="w"> </span>R178
<span class="m">2605619</span><span class="w"> </span>Bacteria<span class="p">;</span>Pseudomonadota<span class="p">;</span>Gammaproteobacteria<span class="p">;</span>Enterobacterales<span class="p">;</span>Enterobacteriaceae<span class="p">;</span>Escherichia<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="p">;</span>Escherichia<span class="w"> </span>coli<span class="w"> </span>O16:H48
<span class="m">2697049</span><span class="w"> </span>Viruses<span class="p">;</span>Pisuviricota<span class="p">;</span>Pisoniviricetes<span class="p">;</span>Nidovirales<span class="p">;</span>Coronaviridae<span class="p">;</span>Betacoronavirus<span class="p">;</span>Betacoronavirus<span class="w"> </span>pandemicum<span class="p">;</span>Severe<span class="w"> </span>acute<span class="w"> </span>respiratory<span class="w"> </span>syndrome<span class="w"> </span>coronavirus<span class="w"> </span><span class="m">2</span>
</code></pre></div>

<p>Replace missing ranks with <code>Unassigned</code> and output tab-delimited format.
<strong>(Warning:  for NCBI taxonomy data since March 2025, <code>reformat</code> can't handle Bacteria's rank <code>domain</code> and Viruses' rank <code>acellular root</code> simutaneously).</strong></p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat2<span class="w"> </span>-I<span class="w"> </span><span class="m">1</span><span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;Unassigned&quot;</span><span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;{d}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-H<span class="w"> </span>-t

<span class="m">376619</span><span class="w">    </span>Bacteria<span class="w">     </span>Pseudomonadota<span class="w">      </span>Gammaproteobacteria<span class="w">   </span>Thiotrichales<span class="w">        </span>Francisellaceae<span class="w">      </span>Francisella<span class="w">                  </span>Francisella<span class="w"> </span>tularensis<span class="w">                            </span>Francisella<span class="w"> </span>tularensis<span class="w"> </span>subsp.<span class="w"> </span>holarctica<span class="w"> </span>LVS
<span class="m">349741</span><span class="w">    </span>Bacteria<span class="w">     </span>Verrucomicrobiota<span class="w">   </span>Verrucomicrobiia<span class="w">      </span>Verrucomicrobiales<span class="w">   </span>Akkermansiaceae<span class="w">      </span>Akkermansia<span class="w">                  </span>Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>ATCC<span class="w"> </span>BAA-835<span class="w">        </span>
<span class="m">239935</span><span class="w">    </span>Bacteria<span class="w">     </span>Verrucomicrobiota<span class="w">   </span>Verrucomicrobiia<span class="w">      </span>Verrucomicrobiales<span class="w">   </span>Akkermansiaceae<span class="w">      </span>Akkermansia<span class="w">                  </span>Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>Unassigned<span class="w">                                  </span>
<span class="m">314101</span><span class="w">    </span>Bacteria<span class="w">     </span>Unassigned<span class="w">          </span>Unassigned<span class="w">            </span>Unassigned<span class="w">           </span>Unassigned<span class="w">           </span>Unassigned<span class="w">                   </span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="w">   </span>Unassigned<span class="w">                                  </span>
<span class="m">11932</span><span class="w">     </span>Unassigned<span class="w">   </span>Artverviricota<span class="w">      </span>Revtraviricetes<span class="w">       </span>Ortervirales<span class="w">         </span>Retroviridae<span class="w">         </span>Intracisternal<span class="w"> </span>A-particles<span class="w">   </span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="w">                   </span>Unassigned<span class="w">                                  </span>
<span class="m">1327037</span><span class="w">   </span>Unassigned<span class="w">   </span>Uroviricota<span class="w">         </span>Caudoviricetes<span class="w">        </span>Unassigned<span class="w">           </span>Unassigned<span class="w">           </span>Unassigned<span class="w">                   </span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="w">                         </span>Unassigned<span class="w">                                  </span>
<span class="m">83333</span><span class="w">     </span>Bacteria<span class="w">     </span>Pseudomonadota<span class="w">      </span>Gammaproteobacteria<span class="w">   </span>Enterobacterales<span class="w">     </span>Enterobacteriaceae<span class="w">   </span>Escherichia<span class="w">                  </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>K-12<span class="w">                       </span>
<span class="m">1408252</span><span class="w">   </span>Bacteria<span class="w">     </span>Pseudomonadota<span class="w">      </span>Gammaproteobacteria<span class="w">   </span>Enterobacterales<span class="w">     </span>Enterobacteriaceae<span class="w">   </span>Escherichia<span class="w">                  </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>R178<span class="w">                       </span>
<span class="m">2605619</span><span class="w">   </span>Bacteria<span class="w">     </span>Pseudomonadota<span class="w">      </span>Gammaproteobacteria<span class="w">   </span>Enterobacterales<span class="w">     </span>Enterobacteriaceae<span class="w">   </span>Escherichia<span class="w">                  </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Unassigned<span class="w">                                  </span>
<span class="m">2697049</span><span class="w">   </span>Unassigned<span class="w">   </span>Pisuviricota<span class="w">        </span>Pisoniviricetes<span class="w">       </span>Nidovirales<span class="w">          </span>Coronaviridae<span class="w">        </span>Betacoronavirus<span class="w">              </span>Betacoronavirus<span class="w"> </span>pandemicum<span class="w">                        </span>Unassigned
</code></pre></div>

<p>Fill missing ranks and add prefixes.
<strong>(Warning:  for NCBI taxonomy data since March 2025, <code>reformat</code> can't handle Bacteria's rank <code>domain</code> and Viruses' rank <code>acellular root</code> simutaneously).</strong></p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat<span class="w"> </span>-I<span class="w"> </span><span class="m">1</span><span class="w"> </span>-F<span class="w"> </span>-P<span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;{d}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-H<span class="w"> </span>-t

<span class="m">376619</span><span class="w">    </span>d__Bacteria<span class="w">                      </span>p__Pseudomonadota<span class="w">                 </span>c__Gammaproteobacteria<span class="w">           </span>o__Thiotrichales<span class="w">                       </span>f__Francisellaceae<span class="w">                      </span>g__Francisella<span class="w">                         </span>s__Francisella<span class="w"> </span>tularensis<span class="w">                            </span>t__Francisella<span class="w"> </span>tularensis<span class="w"> </span>subsp.<span class="w"> </span>holarctica<span class="w"> </span>LVS<span class="w">                                  </span>
<span class="m">349741</span><span class="w">    </span>d__Bacteria<span class="w">                      </span>p__Verrucomicrobiota<span class="w">              </span>c__Verrucomicrobiia<span class="w">              </span>o__Verrucomicrobiales<span class="w">                  </span>f__Akkermansiaceae<span class="w">                      </span>g__Akkermansia<span class="w">                         </span>s__Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>t__Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>ATCC<span class="w"> </span>BAA-835<span class="w">                                          </span>
<span class="m">239935</span><span class="w">    </span>d__Bacteria<span class="w">                      </span>p__Verrucomicrobiota<span class="w">              </span>c__Verrucomicrobiia<span class="w">              </span>o__Verrucomicrobiales<span class="w">                  </span>f__Akkermansiaceae<span class="w">                      </span>g__Akkermansia<span class="w">                         </span>s__Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>t__unclassified<span class="w"> </span>Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>subspecies/strain<span class="w">                        </span>
<span class="m">314101</span><span class="w">    </span>d__Bacteria<span class="w">                      </span>p__unclassified<span class="w"> </span>Bacteria<span class="w"> </span>phylum<span class="w">   </span>c__unclassified<span class="w"> </span>Bacteria<span class="w"> </span>class<span class="w">   </span>o__unclassified<span class="w"> </span>Bacteria<span class="w"> </span>order<span class="w">         </span>f__unclassified<span class="w"> </span>Bacteria<span class="w"> </span>family<span class="w">         </span>g__unclassified<span class="w"> </span>Bacteria<span class="w"> </span>genus<span class="w">         </span>s__uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="w">   </span>t__unclassified<span class="w"> </span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="w"> </span>subspecies/strain
<span class="m">11932</span><span class="w">     </span>d__unclassified<span class="w"> </span>Viruses<span class="w"> </span>domain<span class="w">   </span>p__Artverviricota<span class="w">                 </span>c__Revtraviricetes<span class="w">               </span>o__Ortervirales<span class="w">                        </span>f__Retroviridae<span class="w">                         </span>g__Intracisternal<span class="w"> </span>A-particles<span class="w">          </span>s__Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="w">                   </span>t__unclassified<span class="w"> </span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="w"> </span>subspecies/strain<span class="w">                </span>
<span class="m">1327037</span><span class="w">   </span>d__unclassified<span class="w"> </span>Viruses<span class="w"> </span>domain<span class="w">   </span>p__Uroviricota<span class="w">                    </span>c__Caudoviricetes<span class="w">                </span>o__unclassified<span class="w"> </span>Caudoviricetes<span class="w"> </span>order<span class="w">   </span>f__unclassified<span class="w"> </span>Caudoviricetes<span class="w"> </span>family<span class="w">   </span>g__unclassified<span class="w"> </span>Caudoviricetes<span class="w"> </span>genus<span class="w">   </span>s__Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="w">                         </span>t__unclassified<span class="w"> </span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="w"> </span>subspecies/strain<span class="w">                      </span>
<span class="m">83333</span><span class="w">     </span>d__Bacteria<span class="w">                      </span>p__Pseudomonadota<span class="w">                 </span>c__Gammaproteobacteria<span class="w">           </span>o__Enterobacterales<span class="w">                    </span>f__Enterobacteriaceae<span class="w">                   </span>g__Escherichia<span class="w">                         </span>s__Escherichia<span class="w"> </span>coli<span class="w">                                  </span>t__Escherichia<span class="w"> </span>coli<span class="w"> </span>K-12<span class="w">                                                         </span>
<span class="m">1408252</span><span class="w">   </span>d__Bacteria<span class="w">                      </span>p__Pseudomonadota<span class="w">                 </span>c__Gammaproteobacteria<span class="w">           </span>o__Enterobacterales<span class="w">                    </span>f__Enterobacteriaceae<span class="w">                   </span>g__Escherichia<span class="w">                         </span>s__Escherichia<span class="w"> </span>coli<span class="w">                                  </span>t__Escherichia<span class="w"> </span>coli<span class="w"> </span>R178<span class="w">                                                         </span>
<span class="m">2605619</span><span class="w">   </span>d__Bacteria<span class="w">                      </span>p__Pseudomonadota<span class="w">                 </span>c__Gammaproteobacteria<span class="w">           </span>o__Enterobacterales<span class="w">                    </span>f__Enterobacteriaceae<span class="w">                   </span>g__Escherichia<span class="w">                         </span>s__Escherichia<span class="w"> </span>coli<span class="w">                                  </span>t__unclassified<span class="w"> </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>subspecies/strain<span class="w">                               </span>
<span class="m">2697049</span><span class="w">   </span>d__unclassified<span class="w"> </span>Viruses<span class="w"> </span>domain<span class="w">   </span>p__Pisuviricota<span class="w">                   </span>c__Pisoniviricetes<span class="w">               </span>o__Nidovirales<span class="w">                         </span>f__Coronaviridae<span class="w">                        </span>g__Betacoronavirus<span class="w">                     </span>s__Betacoronavirus<span class="w"> </span>pandemicum<span class="w">                        </span>t__unclassified<span class="w"> </span>Betacoronavirus<span class="w"> </span>pandemicum<span class="w"> </span>subspecies/strain
</code></pre></div>

<p><strong>When these's no nodes of rank "subspecies" nor "strain",
we can switch <code>-S/--pseudo-strain</code> to use the node with lowest rank
as subspecies/strain name, if which rank is lower than "species"</strong>.
<strong>(Warning:  for NCBI taxonomy data since March 2025, <code>reformat</code> can't handle Bacteria's rank <code>domain</code> and Viruses' rank <code>acellular root</code> simutaneously).</strong></p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>taxids3.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-r<span class="w"> </span>-L<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat<span class="w"> </span>-I<span class="w"> </span><span class="m">1</span><span class="w"> </span>-F<span class="w"> </span>-S<span class="w"> </span>-f<span class="w"> </span><span class="s2">&quot;{d}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,2,9,10<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>add-header<span class="w"> </span>-t<span class="w"> </span>-n<span class="w"> </span><span class="s2">&quot;taxid,rank,species,strain&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-t

taxid<span class="w">     </span>rank<span class="w">         </span>species<span class="w">                                           </span>strain<span class="w">                                                                        </span>
-------<span class="w">   </span>----------<span class="w">   </span>-----------------------------------------------<span class="w">   </span>------------------------------------------------------------------------------
<span class="m">376619</span><span class="w">    </span>strain<span class="w">       </span>Francisella<span class="w"> </span>tularensis<span class="w">                            </span>Francisella<span class="w"> </span>tularensis<span class="w"> </span>subsp.<span class="w"> </span>holarctica<span class="w"> </span>LVS<span class="w">                                  </span>
<span class="m">349741</span><span class="w">    </span>strain<span class="w">       </span>Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>ATCC<span class="w"> </span>BAA-835<span class="w">                                          </span>
<span class="m">239935</span><span class="w">    </span>species<span class="w">      </span>Akkermansia<span class="w"> </span>muciniphila<span class="w">                           </span>unclassified<span class="w"> </span>Akkermansia<span class="w"> </span>muciniphila<span class="w"> </span>subspecies/strain<span class="w">                        </span>
<span class="m">314101</span><span class="w">    </span>species<span class="w">      </span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="w">   </span>unclassified<span class="w"> </span>uncultured<span class="w"> </span>murine<span class="w"> </span>large<span class="w"> </span>bowel<span class="w"> </span>bacterium<span class="w"> </span>BAC<span class="w"> </span>54B<span class="w"> </span>subspecies/strain
<span class="m">11932</span><span class="w">     </span>species<span class="w">      </span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="w">                   </span>unclassified<span class="w"> </span>Mouse<span class="w"> </span>Intracisternal<span class="w"> </span>A-particle<span class="w"> </span>subspecies/strain<span class="w">                </span>
<span class="m">1327037</span><span class="w">   </span>species<span class="w">      </span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="w">                         </span>unclassified<span class="w"> </span>Croceibacter<span class="w"> </span>phage<span class="w"> </span>P2559Y<span class="w"> </span>subspecies/strain<span class="w">                      </span>
<span class="m">83333</span><span class="w">     </span>strain<span class="w">       </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>K-12<span class="w">                                                         </span>
<span class="m">1408252</span><span class="w">   </span>subspecies<span class="w">   </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>R178<span class="w">                                                         </span>
<span class="m">2605619</span><span class="w">   </span>no<span class="w"> </span>rank<span class="w">      </span>Escherichia<span class="w"> </span>coli<span class="w">                                  </span>Escherichia<span class="w"> </span>coli<span class="w"> </span>O16:H48<span class="w">                                                      </span>
<span class="m">2697049</span><span class="w">   </span>no<span class="w"> </span>rank<span class="w">      </span>Betacoronavirus<span class="w"> </span>pandemicum<span class="w">                        </span>Severe<span class="w"> </span>acute<span class="w"> </span>respiratory<span class="w"> </span>syndrome<span class="w"> </span>coronavirus<span class="w"> </span><span class="m">2</span>
</code></pre></div>

<p>List eight-level lineage for all TaxIds of rank lower than or equal to species, including some nodes with "no rank".
But when filtering with <code>-L/--lower-than</code>, you can use
<code>-n/--save-predictable-norank</code> to <strong>save some special ranks without order,
where rank of the closest higher node is still lower than rank cutoff</strong>.</p>
<div class="codehilite"><pre><span></span><code><span class="c">$ time taxonkit list </span><span class="nb">--</span><span class="c">ids 1 </span><span class="nb">-</span><span class="c">I &quot;&quot; \</span>
<span class="c">    | taxonkit filter </span><span class="nb">-</span><span class="c">L species </span><span class="nb">-</span><span class="c">E species </span><span class="nb">-</span><span class="c">R </span><span class="nb">-</span><span class="c">N </span><span class="nb">-</span><span class="c">n \</span>
<span class="c">    | taxonkit lineage </span><span class="nb">-</span><span class="c">n </span><span class="nb">-</span><span class="c">r \</span>
<span class="c">    | taxonkit reformat2 </span><span class="nb">-</span><span class="c">I 1 </span><span class="nb">-</span><span class="c">f &quot;{domain|acellular root|superkingdom}\t{phylum}\t{class}\t{order}\t{family}\t{genus}\t{species}\t{strain|subspecies|no rank}&quot; \</span>
<span class="c">    | csvtk cut </span><span class="nb">-</span><span class="c">Ht </span><span class="nb">-</span><span class="c">l </span><span class="nb">-</span><span class="c">f 1</span><span class="nt">,</span><span class="c">3</span><span class="nt">,</span><span class="c">2</span><span class="nt">,</span><span class="c">1</span><span class="nt">,</span><span class="c">4</span><span class="nb">-</span><span class="c">11 \</span>
<span class="c">    | csvtk add</span><span class="nb">-</span><span class="c">header </span><span class="nb">-</span><span class="c">t </span><span class="nb">-</span><span class="c">n &quot;taxid</span><span class="nt">,</span><span class="c">rank</span><span class="nt">,</span><span class="c">name</span><span class="nt">,</span><span class="c">lineage</span><span class="nt">,</span><span class="c">kingdom</span><span class="nt">,</span><span class="c">phylum</span><span class="nt">,</span><span class="c">class</span><span class="nt">,</span><span class="c">order</span><span class="nt">,</span><span class="c">family</span><span class="nt">,</span><span class="c">genus</span><span class="nt">,</span><span class="c">species</span><span class="nt">,</span><span class="c">strain&quot; \</span>
<span class="c">    | pigz </span><span class="nb">-</span><span class="c">c </span><span class="nv">&gt;</span><span class="c"> result</span><span class="nt">.</span><span class="c">tsv</span><span class="nt">.</span><span class="c">gz</span>

<span class="c">real    0m9</span><span class="nt">.</span><span class="c">778s</span>
<span class="c">user    1m22</span><span class="nt">.</span><span class="c">211s</span>
<span class="c">sys     0m8</span><span class="nt">.</span><span class="c">489s</span>

<span class="c">$ pigz </span><span class="nb">-</span><span class="c">cd result</span><span class="nt">.</span><span class="c">tsv</span><span class="nt">.</span><span class="c">gz \</span>
<span class="c">    | csvtk grep </span><span class="nb">-</span><span class="c">t </span><span class="nb">-</span><span class="c">f taxid </span><span class="nb">-</span><span class="c">p 2697049 \</span>
<span class="c">    | csvtk transpose </span><span class="nb">-</span><span class="c">t \</span>
<span class="c">    | csvtk pretty </span><span class="nb">-</span><span class="c">H </span><span class="nb">-</span><span class="c">t </span><span class="nb">-</span><span class="c">W 70 </span><span class="nb">-</span><span class="c">x &#39;;&#39; </span><span class="nb">-</span><span class="c">S round</span>

<span class="c">╭</span><span class="nb">---------</span><span class="c">┬</span><span class="nb">------------------------------------------------------------------------</span><span class="c">╮</span>
<span class="c">| taxid   | 2697049                                                                |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| rank    | Severe acute respiratory syndrome coronavirus 2                        |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| name    | Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;          |</span>
<span class="c">|         | Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;          |</span>
<span class="c">|         | Betacoronavirus;Sarbecovirus;Betacoronavirus pandemicum;               |</span>
<span class="c">|         | Severe acute respiratory syndrome coronavirus 2                        |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| lineage | 2697049                                                                |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| kingdom | no rank                                                                |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| phylum  | Viruses                                                                |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| class   | Pisuviricota                                                           |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| order   | Pisoniviricetes                                                        |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| family  | Nidovirales                                                            |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| genus   | Coronaviridae                                                          |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| species | Betacoronavirus                                                        |</span>
<span class="c">├</span><span class="nb">---------</span><span class="c">┼</span><span class="nb">------------------------------------------------------------------------</span><span class="c">┤</span>
<span class="c">| strain  | Betacoronavirus pandemicum                                             |</span>
<span class="c">╰</span><span class="nb">---------</span><span class="c">┴</span><span class="nb">------------------------------------------------------------------------</span><span class="c">╯</span>
</code></pre></div>

<h2 id="mapping-old-species-names-to-new-ones">Mapping old species names to new ones<a class="headerlink" href="#mapping-old-species-names-to-new-ones" title="Permanent link">Link</a></h2>
<p>Some species names in papers or websites might changed, we can try querying their TaxIds via their old new names
and then retrieve the new ones.</p>
<div class="codehilite"><pre><span></span><code>cat example/changed_species_names.txt
Lactobacillus fermentum
Mycoplasma gallinaceum

<span class="gh">#</span>  TaxonKit &gt;= v0.15.1
cat example/changed_species_names.txt \
    | taxonkit name2taxid \
    | taxonkit lineage -i 2 -n \
    | cut -f 1,4

Lactobacillus fermentum Limosilactobacillus fermentum
Mycoplasma gallinaceum
</code></pre></div>

<p>Woops, there's no information of <code>Mycoplasma gallinaceum</code>.
Then we check the <a href="https://github.com/shenwei356/taxid-changelog">taxid-changelog</a>.</p>
<div class="codehilite"><pre><span></span><code>zcat taxonkit/taxid-changelog.csv.gz \
    | csvtk grep -f name -P example/changed_species_names.txt
    | csvtk cut -f taxid,version,change,name,rank \
    | csvtk pretty

taxid   version      change           name                      rank
-----   ----------   --------------   -----------------------   -------
1613    2013-02-21   NEW              Lactobacillus fermentum   species
1613    2016-03-01   ABSORB           Lactobacillus fermentum   species
1613    2016-03-01   CHANGE_LIN_LEN   Lactobacillus fermentum   species
29556   2013-02-21   NEW              Mycoplasma gallinaceum    species
29556   2016-03-01   CHANGE_LIN_LEN   Mycoplasma gallinaceum    species
29556   2021-01-01   CHANGE_NAME      Mycoplasma gallinaceum    species
29556   2021-01-01   CHANGE_LIN_LIN   Mycoplasma gallinaceum    species
</code></pre></div>

<p>We can see the names are changed. Full changes can be queried with the taxid. e.g.,</p>
<div class="codehilite"><pre><span></span><code>taxid   version      change           change-value   name                        rank
-----   ----------   --------------   ------------   -------------------------   -------
29556   2013-02-21   NEW                             Mycoplasma gallinaceum      species
29556   2016-03-01   CHANGE_LIN_LEN                  Mycoplasma gallinaceum      species
29556   2020-09-01   CHANGE_NAME                     Mycoplasmopsis gallinacea   species
29556   2020-09-01   CHANGE_LIN_TAX                  Mycoplasmopsis gallinacea   species
29556   2021-01-01   CHANGE_NAME                     Mycoplasma gallinaceum      species
29556   2021-01-01   CHANGE_LIN_LIN                  Mycoplasma gallinaceum      species
29556   2021-09-01   CHANGE_NAME                     Mycoplasmopsis gallinacea   species
29556   2021-09-01   CHANGE_LIN_LIN                  Mycoplasmopsis gallinacea   species
29556   2023-03-01   CHANGE_LIN_LIN                  Mycoplasmopsis gallinacea   species
</code></pre></div>

<p>Then we just use their TaxIds to rertrieve the new names. <strong>The final commands are</strong>:</p>
<div class="codehilite"><pre><span></span><code>zcat taxonkit/taxid-changelog.csv.gz \
    | csvtk grep -f name -P example/changed_species_names.txt \
    | csvtk uniq -f taxid \
    | csvtk cut -f name,taxid \
    | csvtk del-header \
    | csvtk csv2tab \
    | taxonkit lineage -i 2 -n \
    | cut -f 1,4

Lactobacillus fermentum Limosilactobacillus fermentum
Mycoplasma gallinaceum  Mycoplasmopsis gallinacea
</code></pre></div>

<h2 id="add-taxonomy-information-to-blast-result">Add taxonomy information to BLAST result<a class="headerlink" href="#add-taxonomy-information-to-blast-result" title="Permanent link">Link</a></h2>
<p>An blast result file <code>blast_result.txt</code>, where the second column is the accession of matched sequences.</p>
<div class="codehilite"><pre><span></span><code>head -n 5 blast_result.txt | csvtk pretty -Ht

xxxxxxxxxxxxxxxxxxxxx/2/ccs    XM_013496560.1   78.745    494   99    3    6361    6851    895        1385       6.53e-83    326 
xxxxxxxxxxxxxxxxxxxxx/2/ccs    XM_013496560.1   78.543    494   100   3    17168   17658   895        1385       3.04e-81    320 
xxxxxxxxxxxxxxxxxxxxx/76/ccs   LR699760.1       100.000   37    0     0    8139    8175    14507874   14507910   4.27e-06    69.4
xxxxxxxxxxxxxxxxxxxxx/80/ccs   HG994975.1       80.556    540   81    16   8269    8798    3821290    3820765    8.65e-104   394 
xxxxxxxxxxxxxxxxxxxxx/80/ccs   HG994975.1       77.805    410   89    2    9590    9998    3819858    3819450    5.51e-61    252
</code></pre></div>

<p>Prepare <code>acc2taxid.tsv</code> file from <a href="https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz">nucl_gb.accession2taxid.gz</a> file.
Here we use the <code>accession</code> column instead of <code>accession.version</code> column, in case of unmatched versions for some accessions.</p>
<div class="codehilite"><pre><span></span><code>zcat nucl_gb.accession2taxid.gz | cut -f 1,3 | gzip -c &gt; acc2taxid.tsv.gz
</code></pre></div>

<p>Extract needed acc2taxid subset to reduce memory usage.</p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> extract accession and deduplicate and remove versions
cut -f 2 blast_result.txt | csvtk uniq -Ht | csvtk replace -Ht -p &#39;\.\d+$&#39; &gt; acc.txt

<span class="gh">#</span> grep from acc2taxid.tsv.gz
zcat acc2taxid.tsv.gz | grep -w -f acc.txt &gt;  hit.acc2taxid.tsv
</code></pre></div>

<p>Prepare <code>taxid2name.tsv</code>, species name are retrived for the taxids.</p>
<div class="codehilite"><pre><span></span><code>cut -f 2 hit.acc2taxid.tsv | taxonkit reformat -f &#39;{s}&#39; -I 1 &gt; hit.taxid2name.tsv
</code></pre></div>

<p>Append taxids according to the accessions, and append species names for the taxids.</p>
<div class="codehilite"><pre><span></span><code>csvtk add-header -t --names &quot;qseqid,sseqid,pident,length,mismatch,gapopen,qstart,qend,sstart,send,evalue,bitscore&quot;  blast_result.txt \
    | csvtk mutate -t -f sseqid -n taxid \
    | csvtk replace -t -k hit.acc2taxid.tsv -f taxid -p &#39;(.+)\.\d+&#39; -r &#39;{kv}&#39; \
    | csvtk mutate -t -f taxid -n species \
    | csvtk replace -t -k hit.taxid2name.tsv -f species -p &#39;(.+)&#39; -r &#39;{kv}&#39; \
    | head -n 5 | csvtk pretty -t

qseqid                         sseqid           pident    length   mismatch   gapopen   qstart   qend    sstart     send       evalue      bitscore   taxid   species             
----------------------------   --------------   -------   ------   --------   -------   ------   -----   --------   --------   ---------   --------   -----   --------------------
xxxxxxxxxxxxxxxxxxxxx/2/ccs    XM_013496560.1   78.745    494      99         3         6361     6851    895        1385       6.53e-83    326        44415   Eimeria mitis       
xxxxxxxxxxxxxxxxxxxxx/2/ccs    XM_013496560.1   78.543    494      100        3         17168    17658   895        1385       3.04e-81    320        44415   Eimeria mitis       
xxxxxxxxxxxxxxxxxxxxx/76/ccs   LR699760.1       100.000   37       0          0         8139     8175    14507874   14507910   4.27e-06    69.4       3702    Arabidopsis thaliana
xxxxxxxxxxxxxxxxxxxxx/80/ccs   HG994975.1       80.556    540      81         16        8269     8798    3821290    3820765    8.65e-104   394        5802    Eimeria tenella
</code></pre></div>

<h2 id="parsing-krakenbracken-result">Parsing kraken/bracken result<a class="headerlink" href="#parsing-krakenbracken-result" title="Permanent link">Link</a></h2>
<p>Example Data</p>
<ul>
<li><a href="https://github.com/biobakery/biobakery/raw/master/demos/biobakery_demos/data/metaphlan2/input/SRS014459-Stool.fasta.gz">SRS014459-Stool.fasta.gz</a></li>
</ul>
<p>Run Kraken2 and Bracken</p>
<div class="codehilite"><pre><span></span><code><span class="n">KRAKEN_DB</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">shenwei</span><span class="o">/</span><span class="n">ws</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">kraken</span><span class="o">/</span><span class="n">k2_pluspf</span>
<span class="n">THREADS</span><span class="o">=</span><span class="mi">16</span>

<span class="n">CLASSIFICATION_LVL</span><span class="o">=</span><span class="n">S</span>
<span class="n">THRESHOLD</span><span class="o">=</span><span class="mi">10</span>

<span class="n">READ_LEN</span><span class="o">=</span><span class="mi">100</span>
<span class="n">SAMPLE</span><span class="o">=</span><span class="n">SRS014459</span><span class="o">-</span><span class="n">Stool</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">gz</span>

<span class="n">BRACKEN_OUTPUT_FILE</span><span class="o">=$</span><span class="n">SAMPLE</span>

<span class="n">kraken2</span><span class="w"> </span><span class="o">--</span><span class="n">db</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">KRAKEN_DB</span><span class="p">}</span><span class="w"> </span><span class="o">--</span><span class="n">threads</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">THREADS</span><span class="p">}</span><span class="w"> </span><span class="o">-</span><span class="n">report</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">SAMPLE</span><span class="p">}</span><span class="o">.</span><span class="n">kreport</span><span class="w"> </span><span class="o">$</span><span class="n">SAMPLE</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">SAMPLE</span><span class="p">}</span><span class="o">.</span><span class="n">kraken</span>

<span class="n">est_abundance</span><span class="o">.</span><span class="n">py</span><span class="w"> </span><span class="o">-</span><span class="n">i</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">SAMPLE</span><span class="p">}</span><span class="o">.</span><span class="n">kreport</span><span class="w"> </span><span class="o">-</span><span class="n">k</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">KRAKEN_DB</span><span class="p">}</span><span class="o">/</span><span class="n">database</span><span class="o">$</span><span class="p">{</span><span class="n">READ_LEN</span><span class="p">}</span><span class="n">mers</span><span class="o">.</span><span class="n">kmer_distrib</span><span class="w"> </span>\
<span class="w">    </span><span class="o">-</span><span class="n">l</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">CLASSIFICATION_LVL</span><span class="p">}</span><span class="w"> </span><span class="o">-</span><span class="n">t</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">THRESHOLD</span><span class="p">}</span><span class="w"> </span><span class="o">-</span><span class="n">o</span><span class="w"> </span><span class="o">$</span><span class="p">{</span><span class="n">BRACKEN_OUTPUT_FILE</span><span class="p">}</span><span class="o">.</span><span class="n">bracken</span>
</code></pre></div>

<p>Orignial format</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">15</span><span class="w"> </span>SRS014459-Stool.fasta.gz_bracken_species.kreport
<span class="m">100</span>.00<span class="w">  </span><span class="m">9491</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>R<span class="w">       </span><span class="m">1</span><span class="w">       </span>root
<span class="m">99</span>.85<span class="w">   </span><span class="m">9477</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>R1<span class="w">      </span><span class="m">131567</span><span class="w">    </span>cellular<span class="w"> </span>organisms
<span class="m">99</span>.85<span class="w">   </span><span class="m">9477</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>D<span class="w">       </span><span class="m">2</span><span class="w">           </span>Bacteria
<span class="m">66</span>.08<span class="w">   </span><span class="m">6271</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>D1<span class="w">      </span><span class="m">1783270</span><span class="w">       </span>FCB<span class="w"> </span>group
<span class="m">66</span>.08<span class="w">   </span><span class="m">6271</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>D2<span class="w">      </span><span class="m">68336</span><span class="w">           </span>Bacteroidetes/Chlorobi<span class="w"> </span>group
<span class="m">66</span>.08<span class="w">   </span><span class="m">6271</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>P<span class="w">       </span><span class="m">976</span><span class="w">               </span>Bacteroidetes
<span class="m">66</span>.08<span class="w">   </span><span class="m">6271</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>C<span class="w">       </span><span class="m">200643</span><span class="w">              </span>Bacteroidia
<span class="m">66</span>.08<span class="w">   </span><span class="m">6271</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>O<span class="w">       </span><span class="m">171549</span><span class="w">                </span>Bacteroidales
<span class="m">34</span>.45<span class="w">   </span><span class="m">3270</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>F<span class="w">       </span><span class="m">815</span><span class="w">                     </span>Bacteroidaceae
<span class="m">34</span>.45<span class="w">   </span><span class="m">3270</span><span class="w">    </span><span class="m">0</span><span class="w">       </span>G<span class="w">       </span><span class="m">816</span><span class="w">                       </span>Bacteroides
<span class="m">10</span>.43<span class="w">   </span><span class="m">990</span><span class="w">     </span><span class="m">990</span><span class="w">     </span>S<span class="w">       </span><span class="m">246787</span><span class="w">                      </span>Bacteroides<span class="w"> </span>cellulosilyticus
<span class="m">7</span>.98<span class="w">    </span><span class="m">757</span><span class="w">     </span><span class="m">757</span><span class="w">     </span>S<span class="w">       </span><span class="m">28116</span><span class="w">                       </span>Bacteroides<span class="w"> </span>ovatus
<span class="m">3</span>.10<span class="w">    </span><span class="m">293</span><span class="w">     </span><span class="m">0</span><span class="w">       </span>G1<span class="w">      </span><span class="m">2646097</span><span class="w">                     </span>unclassified<span class="w"> </span>Bacteroides
<span class="m">1</span>.06<span class="w">    </span><span class="m">100</span><span class="w">     </span><span class="m">100</span><span class="w">     </span>S<span class="w">       </span><span class="m">2755405</span><span class="w">                       </span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>CACC<span class="w"> </span><span class="m">737</span>
<span class="m">0</span>.49<span class="w">    </span><span class="m">46</span><span class="w">      </span><span class="m">46</span><span class="w">      </span>S<span class="w">       </span><span class="m">2650157</span><span class="w">                       </span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>HF-5287
</code></pre></div>

<p>Converting to MetaPhlAn2 format. (Similar to <a href="https://github.com/jenniferlu717/KrakenTools/blob/master/kreport2mpa.py">kreport2mpa.py</a>)</p>
<div class="codehilite"><pre><span></span><code><span class="err">$</span><span class="w"> </span><span class="n">cat</span><span class="w"> </span><span class="n">SRS014459</span><span class="o">-</span><span class="n">Stool</span><span class="p">.</span><span class="n">fasta</span><span class="p">.</span><span class="n">gz_bracken_species</span><span class="p">.</span><span class="n">kreport</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">5</span><span class="p">,</span><span class="mi">1</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">reformat2</span><span class="w"> </span><span class="o">-</span><span class="n">I</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="ss">&quot;k__{domain|acellular root|superkingdom}|p__{phylum}|c__{class}|o__{order}|f__{family}|g__{genus}|s__{species}&quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="nf">replace</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="ss">&quot;(\|[kpcofgs]__)+$&quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="nf">replace</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="ss">&quot;\|([kpcofgs]__\|)+&quot;</span><span class="w"> </span><span class="o">-</span><span class="n">r</span><span class="w"> </span><span class="ss">&quot;|&quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">uniq</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">grep</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="n">k__</span><span class="w"> </span><span class="o">-</span><span class="n">v</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">tee</span><span class="w"> </span><span class="n">SRS014459</span><span class="o">-</span><span class="n">Stool</span><span class="p">.</span><span class="n">fasta</span><span class="p">.</span><span class="n">gz_bracken_species</span><span class="p">.</span><span class="n">kreport</span><span class="p">.</span><span class="nf">format</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">head</span><span class="w"> </span><span class="o">-</span><span class="n">n</span><span class="w"> </span><span class="mi">10</span>

<span class="n">k__Bacteria</span><span class="w">     </span><span class="mf">99.85</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="w">     </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="w">      </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="w">     </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="w">   </span><span class="mf">34.45</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="o">|</span><span class="n">g__Bacteroides</span><span class="w">    </span><span class="mf">34.45</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="o">|</span><span class="n">g__Bacteroides</span><span class="o">|</span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">cellulosilyticus</span><span class="w">    </span><span class="mf">10.43</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="o">|</span><span class="n">g__Bacteroides</span><span class="o">|</span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">ovatus</span><span class="w">      </span><span class="mf">7.98</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="o">|</span><span class="n">g__Bacteroides</span><span class="o">|</span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">sp</span><span class="p">.</span><span class="w"> </span><span class="n">CACC</span><span class="w"> </span><span class="mi">737</span><span class="w">        </span><span class="mf">1.06</span>
<span class="n">k__Bacteria</span><span class="o">|</span><span class="n">p__Bacteroidota</span><span class="o">|</span><span class="n">c__Bacteroidia</span><span class="o">|</span><span class="n">o__Bacteroidales</span><span class="o">|</span><span class="n">f__Bacteroidaceae</span><span class="o">|</span><span class="n">g__Bacteroides</span><span class="o">|</span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">zhangwenhongii</span><span class="w">      </span><span class="mf">0.49</span>
</code></pre></div>

<p>Converting to Qiime format</p>
<div class="codehilite"><pre><span></span><code><span class="err">$</span><span class="w"> </span><span class="n">cat</span><span class="w"> </span><span class="n">SRS014459</span><span class="o">-</span><span class="n">Stool</span><span class="p">.</span><span class="n">fasta</span><span class="p">.</span><span class="n">gz_bracken_species</span><span class="p">.</span><span class="n">kreport</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">5</span><span class="p">,</span><span class="mi">1</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">reformat2</span><span class="w"> </span><span class="o">-</span><span class="n">I</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="ss">&quot;k__{domain|acellular root|superkingdom}; p__{phylum}; c__{class}; o__{order}; f__{family}; g__{genus}; s__{species}&quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="nf">replace</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="ss">&quot;(; [kpcofgs]__)+$&quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="nf">replace</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="ss">&quot;; ([kpcofgs]__; )+&quot;</span><span class="w"> </span><span class="o">-</span><span class="n">r</span><span class="w"> </span><span class="ss">&quot;; &quot;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">uniq</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">grep</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="n">k__</span><span class="w"> </span><span class="o">-</span><span class="n">v</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">head</span><span class="w"> </span><span class="o">-</span><span class="n">n</span><span class="w"> </span><span class="mi">10</span>

<span class="n">k__Bacteria</span><span class="w">     </span><span class="mf">99.85</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="w">    </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="w">    </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="w">  </span><span class="mf">66.08</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="w">       </span><span class="mf">34.45</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="p">;</span><span class="w"> </span><span class="n">g__Bacteroides</span><span class="w">       </span><span class="mf">34.45</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="p">;</span><span class="w"> </span><span class="n">g__Bacteroides</span><span class="p">;</span><span class="w"> </span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">cellulosilyticus</span><span class="w">      </span><span class="mf">10.43</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="p">;</span><span class="w"> </span><span class="n">g__Bacteroides</span><span class="p">;</span><span class="w"> </span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">ovatus</span><span class="w">        </span><span class="mf">7.98</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="p">;</span><span class="w"> </span><span class="n">g__Bacteroides</span><span class="p">;</span><span class="w"> </span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">sp</span><span class="p">.</span><span class="w"> </span><span class="n">CACC</span><span class="w"> </span><span class="mi">737</span><span class="w">  </span><span class="mf">1.06</span>
<span class="n">k__Bacteria</span><span class="p">;</span><span class="w"> </span><span class="n">p__Bacteroidota</span><span class="p">;</span><span class="w"> </span><span class="n">c__Bacteroidia</span><span class="p">;</span><span class="w"> </span><span class="n">o__Bacteroidales</span><span class="p">;</span><span class="w"> </span><span class="n">f__Bacteroidaceae</span><span class="p">;</span><span class="w"> </span><span class="n">g__Bacteroides</span><span class="p">;</span><span class="w"> </span><span class="n">s__Bacteroides</span><span class="w"> </span><span class="n">zhangwenhongii</span><span class="w">        </span><span class="mf">0.49</span>
</code></pre></div>

<p>Save taxon proportion and taxid, and get lineage, name and rank.</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>SRS014459-Stool.fasta.gz_bracken_species.kreport<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,5<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-i<span class="w"> </span><span class="m">2</span><span class="w"> </span>-n<span class="w"> </span>-r<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,2,5,4,3<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-Ht

<span class="m">100</span>.00<span class="w">   </span><span class="m">1</span><span class="w">         </span>no<span class="w"> </span>rank<span class="w">         </span>root<span class="w">                             </span>root<span class="w">                                                                                                                                                 </span>
<span class="m">99</span>.85<span class="w">    </span><span class="m">131567</span><span class="w">    </span>cellular<span class="w"> </span>root<span class="w">   </span>cellular<span class="w"> </span>organisms<span class="w">               </span>cellular<span class="w"> </span>organisms<span class="w">                                                                                                                                   </span>
<span class="m">99</span>.85<span class="w">    </span><span class="m">2</span><span class="w">         </span>domain<span class="w">          </span>Bacteria<span class="w">                         </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="w">                                                                                                                          </span>
<span class="m">66</span>.08<span class="w">    </span><span class="m">1783270</span><span class="w">   </span>clade<span class="w">           </span>FCB<span class="w"> </span>group<span class="w">                        </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="w">                                                                                                 </span>
<span class="m">66</span>.08<span class="w">    </span><span class="m">68336</span><span class="w">     </span>clade<span class="w">           </span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="w">   </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="w">                                                                  </span>
<span class="m">66</span>.08<span class="w">    </span><span class="m">976</span><span class="w">       </span>phylum<span class="w">          </span>Bacteroidota<span class="w">                     </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="p">;</span>Bacteroidota<span class="w">                                                     </span>
<span class="m">66</span>.08<span class="w">    </span><span class="m">200643</span><span class="w">    </span>class<span class="w">           </span>Bacteroidia<span class="w">                      </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="w">                                         </span>
<span class="m">66</span>.08<span class="w">    </span><span class="m">171549</span><span class="w">    </span>order<span class="w">           </span>Bacteroidales<span class="w">                    </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="w">                           </span>
<span class="m">34</span>.45<span class="w">    </span><span class="m">815</span><span class="w">       </span>family<span class="w">          </span>Bacteroidaceae<span class="w">                   </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="w">            </span>
<span class="m">34</span>.45<span class="w">    </span><span class="m">816</span><span class="w">       </span>genus<span class="w">           </span>Bacteroides<span class="w">                      </span>cellular<span class="w"> </span>organisms<span class="p">;</span>Bacteria<span class="p">;</span>Pseudomonadati<span class="p">;</span>FCB<span class="w"> </span>group<span class="p">;</span>Bacteroidota/Chlorobiota<span class="w"> </span>group<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides
</code></pre></div>

<p>Only save species or lower level and get lineage in format of "superkingdom phylum class order family genus species".</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>cat<span class="w"> </span>SRS014459-Stool.fasta.gz_bracken_species.kreport<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,5<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>filter<span class="w"> </span>-N<span class="w"> </span>-E<span class="w"> </span>species<span class="w"> </span>-L<span class="w"> </span>species<span class="w"> </span>-i<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-i<span class="w"> </span><span class="m">2</span><span class="w"> </span>-n<span class="w"> </span>-r<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>reformat2<span class="w"> </span>-I<span class="w"> </span><span class="m">2</span><span class="w">  </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>cut<span class="w"> </span>-Ht<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span>,2,5,4,6<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>add-header<span class="w"> </span>-t<span class="w"> </span>-n<span class="w"> </span>abundance,taxid,rank,name,lineage<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-t

abundance<span class="w">   </span>taxid<span class="w">     </span>rank<span class="w">      </span>name<span class="w">                           </span>lineage<span class="w">                                                                                                </span>
---------<span class="w">   </span>-------<span class="w">   </span>-------<span class="w">   </span>----------------------------<span class="w">   </span>-------------------------------------------------------------------------------------------------------
<span class="m">10</span>.43<span class="w">       </span><span class="m">246787</span><span class="w">    </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>cellulosilyticus<span class="w">   </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>cellulosilyticus
<span class="m">7</span>.98<span class="w">        </span><span class="m">28116</span><span class="w">     </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>ovatus<span class="w">             </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>ovatus<span class="w">          </span>
<span class="m">1</span>.06<span class="w">        </span><span class="m">2755405</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>CACC<span class="w"> </span><span class="m">737</span><span class="w">       </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>CACC<span class="w"> </span><span class="m">737</span><span class="w">    </span>
<span class="m">0</span>.49<span class="w">        </span><span class="m">2650157</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>zhangwenhongii<span class="w">     </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>zhangwenhongii<span class="w">  </span>
<span class="m">0</span>.99<span class="w">        </span><span class="m">2528203</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>A1C1<span class="w">           </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>A1C1<span class="w">        </span>
<span class="m">0</span>.28<span class="w">        </span><span class="m">2763022</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>M10<span class="w">            </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>sp.<span class="w"> </span>M10<span class="w">         </span>
<span class="m">0</span>.16<span class="w">        </span><span class="m">2650158</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>luhongzhouii<span class="w">       </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>luhongzhouii<span class="w">    </span>
<span class="m">0</span>.12<span class="w">        </span><span class="m">2715212</span><span class="w">   </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>faecium<span class="w">            </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>faecium<span class="w">         </span>
<span class="m">5</span>.10<span class="w">        </span><span class="m">817</span><span class="w">       </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>fragilis<span class="w">           </span>Bacteria<span class="p">;</span>Bacteroidota<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>fragilis<span class="w">        </span><span class="m">817</span><span class="w">       </span>species<span class="w">   </span>Bacteroides<span class="w"> </span>fragilis<span class="w">           </span>Bacteria<span class="p">;</span>Bacteroidetes<span class="p">;</span>Bacteroidia<span class="p">;</span>Bacteroidales<span class="p">;</span>Bacteroidaceae<span class="p">;</span>Bacteroides<span class="p">;</span>Bacteroides<span class="w"> </span>fragilis
</code></pre></div>

<h2 id="making-nr-blastdb-for-specific-taxids">Making nr blastdb for specific taxids<a class="headerlink" href="#making-nr-blastdb-for-specific-taxids" title="Permanent link">Link</a></h2>
<p>Attention:</p>
<ul>
<li>
<p>(2023-11-27) BLAST+ 2.2.15 supports limiting a group of organisms without first using a custom script to get all species-level Taxonomy IDs (taxids) for the group. <a href="https://www.ncbi.nlm.nih.gov/books/NBK569839/#_usrman_BLAST_feat_Taxonomic_filtering_fo_">Details</a>.</p>
<p>E.g., Search of the nr BLAST database limited to Bacteria (taxID 2).</p>
<div class="codehilite"><pre><span></span><code>blastp -db nr -taxids 2 -query ...
</code></pre></div>

</li>
<li>
<p>(2019) <a href="https://ncbiinsights.ncbi.nlm.nih.gov/2019/01/04/blast-2-8-1-with-new-databases-and-better-performance/">BLAST+ 2.8.1 is released with new databases</a>,
which allows you to limit your search by taxonomy using information built into the BLAST databases.
So you don't need to build blastdb for specific taxids now.</p>
</li>
</ul>
<p>Changes:</p>
<ul>
<li>2018-09-13 rewritten</li>
<li>2018-12-22 providing faster method for step 3.1</li>
<li>2019-01-07 add note of new blastdb version</li>
<li>2020-10-14 update steps for huge number of accessions belong to high taxon level like bacteria.</li>
</ul>
<p>Data:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db">pre-formated blastdb</a> (09/10/2018)</li>
<li><a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz">prot.accession2taxid.gz</a> (09/07/2018) (optional, but recommended)</li>
</ul>
<p>Hardware in this tutorial</p>
<ul>
<li>CPU: AMD 8-cores/16-threads 3.7Ghz</li>
<li>RAM: 64GB</li>
<li>DISK:<ul>
<li>Taxonomy files stores in NVMe SSD</li>
<li>blastdb files stores in 7200rpm HDD</li>
</ul>
</li>
</ul>
<p>Tools:</p>
<ul>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">blast+</a></li>
<li><a href="https://zlib.net/pigz/">pigz</a> (recommended, faster than gzip)</li>
<li><a href="https://bioinf.shenwei.me/taxonkit">taxonkit</a></li>
<li><a href="https://bioinf.shenwei.me/seqkit">seqkit</a> (recommended),  version &gt;= 0.14.0</li>
<li><a href="https://github.com/shenwei356/rush">rush</a> (optional, for parallizing filtering sequence)</li>
</ul>
<p>Steps:</p>
<ol>
<li>
<p>Listing all taxids below <code>$id</code> using taxonkit.</p>
<div class="codehilite"><pre><span></span><code>id=6656

# 6656 is the phylum Arthropoda
# echo 6656 | taxonkit lineage | taxonkit reformat
# 6656    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Protostomia;Ecdysozoa;Panarthropoda;Arthropoda    Eukaryota;Arthropoda;;;;;

# 2     bacteria
# 2157  archaea
# 4751  fungi
# 10239 virus

# time: 2s
taxonkit list --ids $id --indent &quot;&quot; &gt; $id.taxid.txt

# taxonkit list --ids 2,4751,10239 --indent &quot;&quot; &gt; microbe.taxid.txt

wc -l $id.taxid.txt
# 518373 6656.taxid.txt
</code></pre></div>

</li>
<li>
<p>Retrieving target accessions. There are two options:</p>
<ol>
<li>
<p>From prot.accession2taxid.gz (<strong>faster, recommended</strong>). <em>Note that some accessions are not in <code>nr</code></em>.</p>
<div class="codehilite"><pre><span></span><code># time: 4min
pigz -dc prot.accession2taxid.gz \
    | csvtk grep -t -f taxid -P $id.taxid.txt \
    | csvtk cut -t -f accession.version,taxid \
    | sed 1d \
    &gt; $id.acc2taxid.txt

cut -f 1 $id.acc2taxid.txt &gt; $id.acc.txt

wc -l $id.acc.txt
# 8174609 6656.acc.txt
</code></pre></div>

</li>
<li>
<p>From pre-formated <code>nr</code> blastdb</p>
<div class="codehilite"><pre><span></span><code># time: 40min
blastdbcmd -db nr -entry all -outfmt &quot;%a %T&quot; | pigz -c &gt; nr.acc2taxid.txt.gz

pigz -dc nr.acc2taxid.txt.gz | wc -l
# 555220892

# time: 3min
pigz -dc nr.acc2taxid.txt.gz \
    | csvtk grep -d &#39; &#39; -D &#39; &#39; -f 2 -P $id.taxid.txt \
    | cut -d &#39; &#39;  -f 1 \
    &gt; $id.acc.txt

wc -l $id.acc.txt
# 6928021 6656.acc.txt
</code></pre></div>

</li>
</ol>
</li>
<li>
<p>Retrieving FASTA sequences from pre-formated blastdb. There are two options:</p>
<ol>
<li>
<p>From <code>nr.fa</code> exported from pre-formated blastdb (<strong>faster, smaller output file, recommended</strong>).
   <strong>DO NOT directly download <code>nr.gz</code> from <a href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz">ncbi ftp</a></strong>,
   in which the FASTA headers are not well formated.</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># 1. exporting nr.fa from pre-formated blastdb</span>

<span class="c1"># time: 117min (run only once)</span>
<span class="n">blastdbcmd</span><span class="w"> </span><span class="o">-</span><span class="n">db</span><span class="w"> </span><span class="n">nr</span><span class="w"> </span><span class="o">-</span><span class="n">dbtype</span><span class="w"> </span><span class="n">prot</span><span class="w"> </span><span class="o">-</span><span class="n">entry</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">-</span><span class="n">outfmt</span><span class="w"> </span><span class="s2">&quot;</span><span class="si">%f</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">-</span><span class="n">out</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">pigz</span><span class="w"> </span><span class="o">-</span><span class="n">c</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">nr</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># =====================================================================</span>

<span class="c1"># 2. filtering sequence belong to $taxid</span>

<span class="c1"># ---------------------------------------------------------------------</span>

<span class="c1"># methond 1) (for cases where $id.acc.txt is not very huge)</span>
<span class="c1"># time: 80min</span>
<span class="c1"># perl one-liner is used to unfold records having mulitple accessions</span>
<span class="n">time</span><span class="w"> </span><span class="n">cat</span><span class="w"> </span><span class="o">&lt;</span><span class="p">(</span><span class="n">echo</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="p">(</span><span class="n">pigz</span><span class="w"> </span><span class="o">-</span><span class="n">dc</span><span class="w"> </span><span class="n">nr</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="p">)</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">perl</span><span class="w"> </span><span class="o">-</span><span class="n">e</span><span class="w"> </span><span class="s1">&#39;BEGIN{ $/ = &quot;</span><span class="se">\n</span><span class="s1">&gt;&quot;; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, &quot;</span><span class="se">\n</span><span class="s1">&quot;; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print &quot;&gt;$_&quot;; next; }; $h = &quot;&gt;$h&quot;; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print &quot;&gt;$h1</span><span class="se">\n</span><span class="s1">$s&quot;;} } &#39;</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">seqkit</span><span class="w"> </span><span class="n">grep</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc</span><span class="o">.</span><span class="n">txt</span><span class="w"> </span><span class="o">-</span><span class="n">o</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># ---------------------------------------------------------------------</span>

<span class="c1"># method 2) (**faster**)</span>

<span class="c1"># 33min (run only once)</span>
<span class="c1"># (1). split nr.fa.gz. # Note: I have 16 cpus.</span>
<span class="o">$</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="n">seqkit</span><span class="w"> </span><span class="n">split2</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="mi">15</span><span class="w"> </span><span class="n">nr</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># (2). parallize unfolding</span>
<span class="o">$</span><span class="w"> </span><span class="n">cat</span><span class="w"> </span><span class="n">_unfold_blastdb_fa</span><span class="o">.</span><span class="n">sh</span>
<span class="c1">#!/bin/sh</span>
<span class="n">perl</span><span class="w"> </span><span class="o">-</span><span class="n">e</span><span class="w"> </span><span class="s1">&#39;BEGIN{ $/ = &quot;</span><span class="se">\n</span><span class="s1">&gt;&quot;; &lt;&gt;; } while(&lt;&gt;){s/&gt;$//;  $i = index $_, &quot;</span><span class="se">\n</span><span class="s1">&quot;; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ /&gt;/) { print &quot;&gt;$_&quot;; next; }; $h = &quot;&gt;$h&quot;; while($h =~ /&gt;([^ ]+ .+?) ?(?=&gt;|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print &quot;&gt;$h1</span><span class="se">\n</span><span class="s1">$s&quot;;} } &#39;</span>

<span class="c1"># 10 min</span>
<span class="n">time</span><span class="w"> </span><span class="n">ls</span><span class="w"> </span><span class="n">nr</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">split</span><span class="o">/</span><span class="n">nr</span><span class="o">.</span><span class="n">part_</span><span class="o">*.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">rush</span><span class="w"> </span><span class="o">-</span><span class="n">j</span><span class="w"> </span><span class="mi">15</span><span class="w"> </span><span class="o">-</span><span class="n">v</span><span class="w"> </span><span class="n">id</span><span class="o">=$</span><span class="n">id</span><span class="w"> </span><span class="s1">&#39;cat &lt;(echo) &lt;(pigz -dc {}) </span><span class="se">\</span>
<span class="s1">        | ./_unfold_blastdb_fa.sh </span><span class="se">\</span>
<span class="s1">        | seqkit grep -f {id}.acc.txt -o nr.{id}.{%@nr\.(.+)$} &#39;</span>

<span class="c1"># (3). merge result</span>
<span class="n">cat</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">part</span><span class="o">*.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="n">rm</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">part</span><span class="o">*.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># ---------------------------------------------------------------------</span>

<span class="c1"># method 3) (for huge $id.acc.txt file, e.g., bacteria)</span>

<span class="c1"># (1). split ${id}.acc.txt into several parts. chunk size depends on lines and RAM (64G for me).</span>
<span class="n">split</span><span class="w"> </span><span class="o">-</span><span class="n">d</span><span class="w"> </span><span class="o">-</span><span class="n">l</span><span class="w"> </span><span class="mi">300000000</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc</span><span class="o">.</span><span class="n">txt</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">part_</span>

<span class="c1"># (2). filter</span>
<span class="n">time</span><span class="w"> </span><span class="n">ls</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">part_</span><span class="o">*</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">rush</span><span class="w"> </span><span class="o">-</span><span class="n">j</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="o">--</span><span class="n">immediate</span><span class="o">-</span><span class="n">output</span><span class="w"> </span><span class="o">-</span><span class="n">v</span><span class="w"> </span><span class="n">id</span><span class="o">=$</span><span class="n">id</span><span class="w"> </span>\
<span class="w">        </span><span class="s1">&#39;echo {}; cat &lt;(echo) &lt;(pigz -dc nr.fa.gz ) </span><span class="se">\</span>
<span class="s1">        | ./_unfold_blastdb_fa.sh </span><span class="se">\</span>
<span class="s1">        | seqkit grep -f {} -o nr.{id}.{%@(part_.+)}.fa.gz &#39;</span>

<span class="c1"># (3). merge</span>
<span class="n">cat</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">part</span><span class="o">*.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># clean</span>
<span class="n">rm</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">part</span><span class="o">*.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="n">rm</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">part_</span>

<span class="c1"># (4). optionally adding taxid, you may edit replacement (-r) below</span>
<span class="c1"># split</span>
<span class="n">time</span><span class="w"> </span><span class="n">split</span><span class="w"> </span><span class="o">-</span><span class="n">d</span><span class="w"> </span><span class="o">-</span><span class="n">l</span><span class="w"> </span><span class="mi">200000000</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc2taxid</span><span class="o">.</span><span class="n">txt</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc2taxid</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">part_</span>

<span class="n">ln</span><span class="w"> </span><span class="o">-</span><span class="n">s</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">part0</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w">         </span>
<span class="n">i</span><span class="o">=</span><span class="mi">0</span>
<span class="k">for</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="o">$</span><span class="n">id</span><span class="o">.</span><span class="n">acc2taxid</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">part_</span><span class="o">*</span><span class="w"> </span><span class="p">;</span><span class="w"> </span><span class="n">do</span>
<span class="w">    </span><span class="n">echo</span><span class="w"> </span><span class="o">$</span><span class="n">f</span>
<span class="w">    </span><span class="n">time</span><span class="w"> </span><span class="n">pigz</span><span class="w"> </span><span class="o">-</span><span class="n">cd</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">part</span><span class="o">$</span><span class="n">i</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span>\
<span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">seqkit</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">-</span><span class="n">k</span><span class="w"> </span><span class="o">$</span><span class="n">f</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="s2">&quot;^([^\-]+?) &quot;</span><span class="w"> </span><span class="o">-</span><span class="n">r</span><span class="w"> </span><span class="s2">&quot;{kv}-\$1 &quot;</span><span class="w"> </span><span class="o">-</span><span class="n">K</span><span class="w"> </span><span class="o">-</span><span class="n">U</span><span class="w"> </span><span class="o">-</span><span class="n">o</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">part</span><span class="o">$</span><span class="p">((</span><span class="o">$</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="p">;</span>
<span class="w">    </span><span class="o">/</span><span class="n">bin</span><span class="o">/</span><span class="n">rm</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">part</span><span class="o">$</span><span class="n">i</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="w">    </span><span class="n">i</span><span class="o">=$</span><span class="p">((</span><span class="o">$</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">));</span>
<span class="n">done</span>
<span class="n">mv</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">part</span><span class="o">$</span><span class="n">i</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">with</span><span class="o">-</span><span class="n">taxid</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="c1"># =====================================================================</span>

<span class="c1"># 3. counting sequences</span>
<span class="c1">#</span>
<span class="c1"># ls -lh nr.$id.fa.gz</span>
<span class="c1"># -rw-r--r-- 1 shenwei shenwei 902M 9月  13 01:42 nr.6656.fa.gz</span>
<span class="c1">#</span>
<span class="n">pigz</span><span class="w"> </span><span class="o">-</span><span class="n">dc</span><span class="w"> </span><span class="n">nr</span><span class="o">.$</span><span class="n">id</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">grep</span><span class="w"> </span><span class="s1">&#39;^&gt;&#39;</span><span class="w"> </span><span class="o">-</span><span class="n">c</span>

<span class="c1"># 6928017</span>
<span class="c1"># Here 6928017 ~=  6928021 ($id.acc.txt)</span>
</code></pre></div>

</li>
<li>
<p>Directly from pre-formated blastdb</p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> time: 5h20min
blastdbcmd -db nr -entry_batch $id.acc.txt -out - | pigz -c &gt; nr.$id.fa.gz

<span class="gh">#</span> counting sequences
<span class="gh">#</span>
# Note that the headers of outputed fasta by blastdbcmd are &quot;folded&quot;
<span class="gh">#</span> for accessions from different species with same sequences, so the
<span class="gh">#</span> number may be small than $(wc -l $id.acc.txt).
pigz -dc nr.$id.fa.gz | grep &#39;^&gt;&#39; -c
<span class="gh">#</span> 1577383

<span class="gh">#</span> counting accessions
<span class="gh">#</span>
# ls -lh nr.$id.fa.gz
<span class="gh">#</span> -rw-r--r-- 1 shenwei shenwei 2.1G 9月  13 03:38 nr.6656.fa.gz
<span class="gh">#</span>
pigz -dc nr.$id.fa.gz | grep &#39;^&gt;&#39; | sed &#39;s/&gt;/\n&gt;/g&#39; | grep &#39;^&gt;&#39; -c
<span class="gh">#</span> 288415413
</code></pre></div>

</li>
</ol>
</li>
<li>
<p>makeblastdb</p>
<div class="codehilite"><pre><span></span><code><span class="nx">pigz</span><span class="w"> </span><span class="o">-</span><span class="nx">dc</span><span class="w"> </span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span><span class="p">.</span><span class="nx">gz</span><span class="w"> </span><span class="p">&gt;</span><span class="w"> </span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span>

<span class="err">#</span><span class="w"> </span><span class="nx">time</span><span class="p">:</span><span class="w"> </span><span class="mi">3</span><span class="nx">min</span><span class="w"> </span><span class="p">(</span><span class="err">$</span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span><span class="w"> </span><span class="nx">from</span><span class="w"> </span><span class="nx">step</span><span class="w"> </span><span class="mi">3</span><span class="w"> </span><span class="nx">option</span><span class="w"> </span><span class="mi">1</span><span class="p">)</span>
<span class="err">#</span>
<span class="err">#</span><span class="w"> </span><span class="nx">building</span><span class="w"> </span><span class="err">$</span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span><span class="w"> </span><span class="nx">from</span><span class="w"> </span><span class="nx">step</span><span class="w"> </span><span class="mi">3</span><span class="w"> </span><span class="nx">option</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span><span class="nx">with</span><span class="w"> </span><span class="o">-</span><span class="nx">parse_seqids</span><span class="w"> </span><span class="nx">would</span><span class="w"> </span><span class="nx">produce</span><span class="w"> </span><span class="nx">error</span><span class="p">:</span>
<span class="err">#</span>
<span class="err">#</span><span class="w">     </span><span class="nx">BLAST</span><span class="w"> </span><span class="nx">Database</span><span class="w"> </span><span class="nx">creation</span><span class="w"> </span><span class="nx">error</span><span class="p">:</span><span class="w"> </span><span class="nx">Error</span><span class="p">:</span><span class="w"> </span><span class="nx">Duplicate</span><span class="w"> </span><span class="nx">seq_ids</span><span class="w"> </span><span class="nx">are</span><span class="w"> </span><span class="nx">found</span><span class="p">:</span><span class="w"> </span><span class="nx">SP</span><span class="o">|</span><span class="nx">P29868</span><span class="m m-Double">.1</span>
<span class="err">#</span>
<span class="nx">makeblastdb</span><span class="w"> </span><span class="o">-</span><span class="nx">parse_seqids</span><span class="w"> </span><span class="o">-</span><span class="k">in</span><span class="w"> </span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span><span class="w"> </span><span class="o">-</span><span class="nx">dbtype</span><span class="w"> </span><span class="nx">prot</span><span class="w"> </span><span class="o">-</span><span class="nx">out</span><span class="w"> </span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span>

<span class="err">#</span><span class="w"> </span><span class="nx">rm</span><span class="w"> </span><span class="nx">nr</span><span class="p">.</span><span class="err">$</span><span class="nx">id</span><span class="p">.</span><span class="nx">fa</span>
</code></pre></div>

</li>
<li>
<p>blastp (optional)</p>
<div class="codehilite"><pre><span></span><code><span class="cp"># blastdb nr.$id is built from sequences in step 3 option 1</span>
<span class="cp">#</span>
<span class="n">blastp</span><span class="w"> </span><span class="o">-</span><span class="n">num_threads</span><span class="w"> </span><span class="mi">16</span><span class="w"> </span><span class="o">-</span><span class="n">db</span><span class="w"> </span><span class="n">nr</span><span class="p">.</span><span class="n">$id</span><span class="w"> </span><span class="o">-</span><span class="n">query</span><span class="w"> </span><span class="n">t4</span><span class="p">.</span><span class="n">fa</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">t4</span><span class="p">.</span><span class="n">fa</span><span class="p">.</span><span class="n">blast</span>
<span class="cp"># real    0m20.866s</span>

<span class="cp"># $ cat t4.fa.blast | grep Query= -A 10</span>
<span class="cp"># Query= A0A0J9X1W9.2 RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-Hd1a</span>
<span class="cp">#</span>
<span class="cp"># Length=35</span>
<span class="w">                                                                     </span><span class="n">Score</span><span class="w">     </span><span class="n">E</span>
<span class="cp"># Sequences producing significant alignments:                          (Bits)  Value</span>

<span class="cp"># 2MPQ_A  Chain A, Solution structure of the sodium channel toxin Hd1a  72.4    2e-17</span>
<span class="cp"># A0A0J9X1W9.2  RecName: Full=Mu-theraphotoxin-Hd1a; Short=Mu-TRTX-...  72.4    2e-17</span>
<span class="cp"># ADB56726.1  HNTX-IV.2 precursor [Haplopelma hainanum]                 66.6    9e-15</span>
<span class="cp"># D2Y233.1  RecName: Full=Mu-theraphotoxin-Hhn1b 2; Short=Mu-TRTX-H...  66.6    9e-15</span>
<span class="cp"># ADB56830.1  HNTX-IV.3 precursor [Haplopelma hainanum]                 66.6    9e-15</span>
</code></pre></div>

</li>
</ol>
<h2 id="summaries-of-taxonomy-data">Summaries of taxonomy data<a class="headerlink" href="#summaries-of-taxonomy-data" title="Permanent link">Link</a></h2>
<p>You can change the TaxId of interest.</p>
<ol>
<li>
<p>Rank counts of common categories.</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span><span class="nb">echo</span><span class="w"> </span>Archaea<span class="w"> </span>Bacteria<span class="w"> </span>Eukaryota<span class="w"> </span>Fungi<span class="w"> </span>Metazoa<span class="w"> </span>Viridiplantae<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>rush<span class="w"> </span>-D<span class="w"> </span><span class="s1">&#39; &#39;</span><span class="w"> </span>-T<span class="w"> </span>b<span class="w"> </span><span class="se">\</span>
<span class="w">        </span><span class="s1">&#39;taxonkit list --ids $(echo {} | taxonkit name2taxid | cut -f 2) \</span>
<span class="s1">            | sed 1d \</span>
<span class="s1">            | taxonkit filter -i 2 -E genus -L genus \</span>
<span class="s1">            | taxonkit lineage -L -r \</span>
<span class="s1">            | csvtk freq -H -t -f 2 -nr \</span>
<span class="s1">            &gt; stats.{}.tsv &#39;</span>

$<span class="w"> </span>csvtk<span class="w"> </span>-t<span class="w"> </span>join<span class="w"> </span>--outer-join<span class="w"> </span>stats.*.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>add-header<span class="w"> </span>-t<span class="w"> </span>-n<span class="w"> </span><span class="s2">&quot;rank,</span><span class="k">$(</span>ls<span class="w"> </span>stats.*.tsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>rush<span class="w"> </span>-k<span class="w"> </span><span class="s1">&#39;echo {@stats.(.+).tsv}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd,<span class="w"> </span><span class="k">)</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>csv2md<span class="w"> </span>-t
</code></pre></div>

<p><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=statistics&amp;uncultured=hide&amp;unspecified=hide">Similar data on NCBI Taxonomy</a></p>
<table>
<thead>
<tr>
<th style="text-align: left;">rank</th>
<th style="text-align: left;">Archaea</th>
<th style="text-align: left;">Bacteria</th>
<th style="text-align: left;">Eukaryota</th>
<th style="text-align: left;">Fungi</th>
<th style="text-align: left;">Metazoa</th>
<th style="text-align: left;">Viridiplantae</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">species</td>
<td style="text-align: left;">12482</td>
<td style="text-align: left;">460940</td>
<td style="text-align: left;">1349648</td>
<td style="text-align: left;">156908</td>
<td style="text-align: left;">957297</td>
<td style="text-align: left;">191026</td>
</tr>
<tr>
<td style="text-align: left;">strain</td>
<td style="text-align: left;">354</td>
<td style="text-align: left;">40643</td>
<td style="text-align: left;">3486</td>
<td style="text-align: left;">2352</td>
<td style="text-align: left;">33</td>
<td style="text-align: left;">50</td>
</tr>
<tr>
<td style="text-align: left;">genus</td>
<td style="text-align: left;">205</td>
<td style="text-align: left;">4112</td>
<td style="text-align: left;">90882</td>
<td style="text-align: left;">6844</td>
<td style="text-align: left;">64148</td>
<td style="text-align: left;">16202</td>
</tr>
<tr>
<td style="text-align: left;">isolate</td>
<td style="text-align: left;">7</td>
<td style="text-align: left;">503</td>
<td style="text-align: left;">809</td>
<td style="text-align: left;">76</td>
<td style="text-align: left;">17</td>
<td style="text-align: left;">3</td>
</tr>
<tr>
<td style="text-align: left;">species group</td>
<td style="text-align: left;">2</td>
<td style="text-align: left;">77</td>
<td style="text-align: left;">251</td>
<td style="text-align: left;">22</td>
<td style="text-align: left;">214</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">serotype</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">218</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">serogroup</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">136</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">subsection</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">21</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">21</td>
</tr>
<tr>
<td style="text-align: left;">subspecies</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">632</td>
<td style="text-align: left;">24523</td>
<td style="text-align: left;">158</td>
<td style="text-align: left;">17043</td>
<td style="text-align: left;">7212</td>
</tr>
<tr>
<td style="text-align: left;">forma specialis</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">521</td>
<td style="text-align: left;">220</td>
<td style="text-align: left;">179</td>
<td style="text-align: left;">33</td>
<td style="text-align: left;">1</td>
</tr>
<tr>
<td style="text-align: left;">species subgroup</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">23</td>
<td style="text-align: left;">101</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">101</td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">biotype</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">7</td>
<td style="text-align: left;">10</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">morph</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">12</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">section</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">437</td>
<td style="text-align: left;">37</td>
<td style="text-align: left;">2</td>
<td style="text-align: left;">398</td>
</tr>
<tr>
<td style="text-align: left;">genotype</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">12</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">12</td>
</tr>
<tr>
<td style="text-align: left;">series</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">9</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">5</td>
<td style="text-align: left;">4</td>
</tr>
<tr>
<td style="text-align: left;">varietas</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">25</td>
<td style="text-align: left;">8499</td>
<td style="text-align: left;">1100</td>
<td style="text-align: left;">2</td>
<td style="text-align: left;">7188</td>
</tr>
<tr>
<td style="text-align: left;">forma</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">560</td>
<td style="text-align: left;">185</td>
<td style="text-align: left;">6</td>
<td style="text-align: left;">315</td>
</tr>
<tr>
<td style="text-align: left;">subgenus</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">1558</td>
<td style="text-align: left;">10</td>
<td style="text-align: left;">1414</td>
<td style="text-align: left;">112</td>
</tr>
<tr>
<td style="text-align: left;">pathogroup</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">5</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: left;">subvariety</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">5</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">5</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Count of all ranks</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>taxonkit<span class="w"> </span>list<span class="w"> </span>--ids<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-L<span class="w"> </span>-r<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>freq<span class="w"> </span>-H<span class="w"> </span>-t<span class="w"> </span>-f<span class="w"> </span><span class="m">2</span><span class="w"> </span>-nr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-H<span class="w"> </span>-t

species<span class="w">            </span><span class="m">1879659</span>
no<span class="w"> </span>rank<span class="w">            </span><span class="m">222743</span>
genus<span class="w">              </span><span class="m">96625</span>
strain<span class="w">             </span><span class="m">44483</span>
subspecies<span class="w">         </span><span class="m">25174</span>
family<span class="w">             </span><span class="m">9492</span>
varietas<span class="w">           </span><span class="m">8524</span>
subfamily<span class="w">          </span><span class="m">3050</span>
tribe<span class="w">              </span><span class="m">2213</span>
order<span class="w">              </span><span class="m">1660</span>
subgenus<span class="w">           </span><span class="m">1618</span>
isolate<span class="w">            </span><span class="m">1319</span>
serotype<span class="w">           </span><span class="m">1216</span>
clade<span class="w">              </span><span class="m">886</span>
superfamily<span class="w">        </span><span class="m">865</span>
forma<span class="w"> </span>specialis<span class="w">    </span><span class="m">741</span>
forma<span class="w">              </span><span class="m">564</span>
subtribe<span class="w">           </span><span class="m">508</span>
section<span class="w">            </span><span class="m">437</span>
class<span class="w">              </span><span class="m">429</span>
suborder<span class="w">           </span><span class="m">372</span>
species<span class="w"> </span>group<span class="w">      </span><span class="m">330</span>
phylum<span class="w">             </span><span class="m">272</span>
subclass<span class="w">           </span><span class="m">156</span>
serogroup<span class="w">          </span><span class="m">138</span>
infraorder<span class="w">         </span><span class="m">130</span>
species<span class="w"> </span>subgroup<span class="w">   </span><span class="m">124</span>
superorder<span class="w">         </span><span class="m">55</span>
subphylum<span class="w">          </span><span class="m">33</span>
parvorder<span class="w">          </span><span class="m">26</span>
subsection<span class="w">         </span><span class="m">21</span>
genotype<span class="w">           </span><span class="m">20</span>
infraclass<span class="w">         </span><span class="m">18</span>
biotype<span class="w">            </span><span class="m">17</span>
morph<span class="w">              </span><span class="m">12</span>
kingdom<span class="w">            </span><span class="m">11</span>
series<span class="w">             </span><span class="m">9</span>
superclass<span class="w">         </span><span class="m">6</span>
cohort<span class="w">             </span><span class="m">5</span>
pathogroup<span class="w">         </span><span class="m">5</span>
subvariety<span class="w">         </span><span class="m">5</span>
superkingdom<span class="w">       </span><span class="m">4</span>
subcohort<span class="w">          </span><span class="m">3</span>
subkingdom<span class="w">         </span><span class="m">1</span>
superphylum<span class="w">        </span><span class="m">1</span>

real<span class="w">    </span>0m3.663s
user<span class="w">    </span>0m15.897s
sys<span class="w">     </span>0m1.010s
</code></pre></div>

</li>
<li>
<p>Ranks of taxa at or below species.</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>taxonkit<span class="w"> </span>list<span class="w"> </span>--ids<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>filter<span class="w"> </span>--lower-than<span class="w"> </span>species<span class="w"> </span>--equal-to<span class="w"> </span>species<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>taxonkit<span class="w"> </span>lineage<span class="w"> </span>-L<span class="w"> </span>-r<span class="w">  </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>freq<span class="w"> </span>-Ht<span class="w"> </span>-nr<span class="w"> </span>-f<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>add-header<span class="w"> </span>-t<span class="w"> </span>-n<span class="w"> </span>rank,count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>csvtk<span class="w"> </span>pretty<span class="w"> </span>-t

rank<span class="w">              </span>count
---------------<span class="w">   </span>-------
species<span class="w">           </span><span class="m">1880044</span>
no<span class="w"> </span>rank<span class="w">           </span><span class="m">222756</span>
strain<span class="w">            </span><span class="m">44483</span>
subspecies<span class="w">        </span><span class="m">25171</span>
varietas<span class="w">          </span><span class="m">8524</span>
isolate<span class="w">           </span><span class="m">1319</span>
serotype<span class="w">          </span><span class="m">1216</span>
clade<span class="w">             </span><span class="m">885</span>
forma<span class="w"> </span>specialis<span class="w">   </span><span class="m">741</span>
forma<span class="w">             </span><span class="m">564</span>
serogroup<span class="w">         </span><span class="m">138</span>
genotype<span class="w">          </span><span class="m">20</span>
biotype<span class="w">           </span><span class="m">17</span>
morph<span class="w">             </span><span class="m">12</span>
pathogroup<span class="w">        </span><span class="m">5</span>
subvariety<span class="w">        </span><span class="m">5</span>
</code></pre></div>

</li>
</ol>
<h2 id="merging-gtdb-and-ncbi-taxonomy">Merging GTDB and NCBI taxonomy<a class="headerlink" href="#merging-gtdb-and-ncbi-taxonomy" title="Permanent link">Link</a></h2>
<p>Sometimes (<a href="https://github.com/shenwei356/gtdb-taxdump/issues/6">1</a>) one needs to build a database including bacteria and archaea (from GTDB) and viral database from NCBI.
The idea is to export lineages from both GTDB and NCBI using <a href="https://bioinf.shenwei.me/taxonkit/usage/#reformat">taxonkit reformat</a>, and then create taxdump files from them with <a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump">taxonkit create-taxdump</a>.</p>
<ol>
<li>
<p>Exporting taxonomic lineages of <strong>taxa with rank equal to species</strong> from <a href="https://github.com/shenwei356/gtdb-taxdump">GTDB-taxdump</a>.</p>
<div class="codehilite"><pre><span></span><code>dir=gtdb-taxdump/R226/
taxonkit list --data-dir $dir --ids 1 --indent &quot;&quot; \
    | taxonkit filter --data-dir $dir --equal-to species \
    | taxonkit reformat2 --data-dir $dir --taxid-field 1 \
        --format &quot;{domain|acellular root|superkingdom}\t{phylum}\t{class}\t{order}\t{family}\t{genus}\t{species}\t{strain|subspecies|no rank}&quot; \
        -o gtdb.tsv
</code></pre></div>

</li>
<li>
<p>Exporting taxonomic lineages of viral/fungal/human <strong>taxa with rank equal to or lower than species</strong> from NCBI taxdump.
   For taxa whose rank is "no rank" below the species, we treat them as tax of strain rank (<code>--pseudo-strain</code>, taxonkit v0.14.1 needed).</p>
<div class="codehilite"><pre><span></span><code># taxid of Viruses: 10239
dir=ncbi-taxdump
taxonkit list --data-dir $dir --ids 10239,4751,9606 --indent &quot;&quot; \
    | taxonkit filter --data-dir $dir --equal-to species --lower-than species \
    | taxonkit reformat --data-dir $dir --taxid-field 1 \
        --format &quot;{a}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}&quot; \
        --pseudo-strain \
        -o ncbi-viral.tsv

# taxid of Fungi: 4751
# taxid of Homo sapiens: 9606
dir=ncbi-taxdump
taxonkit list --data-dir $dir --ids 4751,9606 --indent &quot;&quot; \
    | taxonkit filter --data-dir $dir --equal-to species --lower-than species \
    | taxonkit reformat --data-dir $dir --taxid-field 1 \
        --format &quot;{d}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}&quot; \
        --pseudo-strain \
        -o ncbi-fungal-human.tsv
</code></pre></div>

</li>
<li>
<p>Creating taxdump from lineages above.</p>
<div class="codehilite"><pre><span></span><code>cat gtdb.tsv ncbi-viral.tsv ncbi-fungal-human.tsv \
    | taxonkit create-taxdump \
        --field-accession 1 \
        -R &quot;superkingdom,phylum,class,order,family,genus,species,strain&quot; \
        -O taxdump

# we use --field-accession  1 to output the mapping file between old taxids and new ones.
$ grep -w 2697049  taxdump/taxid.map  # SARS-COV-2
2697049 192491219
</code></pre></div>

</li>
<li>
<p>(Optional) Creating a two-column taxid mappping file for tools like KMCP and LexicMap.</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># -----------------------------------------------</span>
<span class="c1"># gtdb</span>
<span class="n">dir</span><span class="o">=</span><span class="n">gtdb</span><span class="o">-</span><span class="n">taxdump</span><span class="o">-</span><span class="n">R226</span>
<span class="n">cat</span><span class="w"> </span><span class="o">$</span><span class="n">dir</span><span class="o">/</span><span class="n">taxid</span><span class="o">.</span><span class="n">map</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">reformat2</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="o">$</span><span class="n">dir</span><span class="w"> </span><span class="o">-</span><span class="n">I</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="s1">&#39;{species}&#39;</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">name2taxid</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="o">/</span><span class="w"> </span><span class="o">-</span><span class="n">i</span><span class="w"> </span><span class="mi">3</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">gtdb</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># check missing data</span>
<span class="n">awk</span><span class="w"> </span><span class="s1">&#39;$2 == &quot;&quot;&#39;</span><span class="w"> </span><span class="n">gtdb</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># -----------------------------------------------</span>
<span class="c1"># fungi and viral</span>
<span class="c1"># assembly_summary.txt is linked to assembly_summary.txt downloaded by genome_updater.sh</span>
<span class="n">dir</span><span class="o">=</span><span class="n">ncbi</span><span class="o">-</span><span class="n">taxdump</span>
<span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">6</span><span class="w"> </span><span class="n">assembly_summary</span><span class="o">.</span><span class="n">txt</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">lineage</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="o">$</span><span class="n">dir</span><span class="w"> </span><span class="o">-</span><span class="n">i</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span><span class="o">-</span><span class="n">L</span><span class="w"> </span><span class="o">-</span><span class="n">n</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">name2taxid</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="o">/</span><span class="w"> </span><span class="o">-</span><span class="n">i</span><span class="w"> </span><span class="mi">3</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">ncbi</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># check missing data</span>
<span class="n">awk</span><span class="w"> </span><span class="s1">&#39;$2 == &quot;&quot;&#39;</span><span class="w"> </span><span class="n">ncbi</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># -----------------------------------------------</span>
<span class="c1"># human</span>
<span class="c1"># T2T-CHM13v2.0 is linked to a directory containing human genome fasta files,</span>
<span class="c1"># each containing one chromosome or mitochondrion.</span>
<span class="c1">#</span>
<span class="c1">#     # The human genome file is split into separate files for each chromosomes/mitochondrion.</span>
<span class="c1">#     seqkit split2 -s 1 -N GCA_009914755.4_T2T-CHM13v2.0_genomic.fna.gz -O T2T-CHM13v2.0</span>
<span class="c1">#     # reaname</span>
<span class="c1">#     brename -p ^ -r GCA_009914755.4_ T2T-CHM13v2.0/*.fna.gz</span>
<span class="c1">#</span>
<span class="n">ls</span><span class="w"> </span><span class="n">T2T</span><span class="o">-</span><span class="n">CHM13v2</span><span class="o">.</span><span class="mi">0</span><span class="o">/*.</span><span class="n">fna</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">rush</span><span class="w"> </span><span class="o">-</span><span class="n">k</span><span class="w"> </span><span class="s1">&#39;echo -e &quot;{%..}</span><span class="se">\t</span><span class="s1">Homo sapiens&quot;&#39;</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">name2taxid</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="o">/</span><span class="w"> </span><span class="o">-</span><span class="n">i</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span>\
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">human</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># check missing data</span>
<span class="n">awk</span><span class="w"> </span><span class="s1">&#39;$2 == &quot;&quot;&#39;</span><span class="w"> </span><span class="n">human</span><span class="o">.</span><span class="n">map</span>

<span class="c1"># -----------------------------------------------</span>
<span class="c1"># merge</span>
<span class="n">cat</span><span class="w"> </span><span class="n">gtdb</span><span class="o">.</span><span class="n">map</span><span class="w"> </span><span class="n">ncbi</span><span class="o">.</span><span class="n">map</span><span class="w"> </span><span class="n">human</span><span class="o">.</span><span class="n">map</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">taxid</span><span class="o">.</span><span class="n">map</span>
</code></pre></div>

</li>
<li>
<p>(Optional) Creating a 4-column taxid mappping file for tools like Metabuli.</p>
<div class="codehilite"><pre><span></span><code><span class="err">#</span><span class="w"> </span><span class="nt">gtdb</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nt">fungi</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nt">viral</span>
<span class="err">#</span><span class="w"> </span><span class="nt">file</span><span class="w"> </span><span class="nt">name</span><span class="w"> </span><span class="nt">example</span><span class="o">:</span><span class="w"> </span><span class="nt">GCF_000143185</span><span class="p">.</span><span class="nc">2_Schco3_genomic</span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span><span class="o">,</span>
<span class="err">#</span><span class="w"> </span><span class="nt">where</span><span class="w"> </span><span class="nt">GCF_000143185</span><span class="p">.</span><span class="nc">2</span><span class="w"> </span><span class="nt">is</span><span class="w"> </span><span class="nt">mapped</span><span class="w"> </span><span class="nt">to</span><span class="w"> </span><span class="nt">a</span><span class="w"> </span><span class="nt">taxid</span><span class="o">.</span>
<span class="err">#</span>
<span class="nt">fd</span><span class="w"> </span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span><span class="o">$</span><span class="w"> </span><span class="err">\</span>
<span class="w">        </span><span class="nt">fungi-viruses</span><span class="o">/</span><span class="nt">2026-02-12_18-10-55</span><span class="o">/</span><span class="nt">files</span><span class="o">/</span><span class="w"> </span><span class="err">\</span>
<span class="w">        </span><span class="nt">gtdb</span><span class="o">/</span><span class="nt">gtdb_genomes_reps_r226</span><span class="o">/</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="nt">rush</span><span class="w"> </span><span class="nt">--eta</span><span class="w"> </span><span class="nt">-v</span><span class="w"> </span><span class="s1">&#39;acc={%@^(\w\w\w_\d\d\d\d\d\d\d\d\d\.\d+)}&#39;</span><span class="w"> </span><span class="err">\</span>
<span class="w">        </span><span class="s1">&#39;taxid=$(grep {acc} taxdump/taxid.map | cut -f 2); \</span>
<span class="s1">        seqkit seq -ni {} | awk &quot;{print \$1\&quot;\t\&quot;\$1\&quot;\t\&quot;$taxid\&quot;\t0\&quot;}&quot;&#39;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="nt">csvtk</span><span class="w"> </span><span class="nt">replace</span><span class="w"> </span><span class="nt">-Ht</span><span class="w"> </span><span class="nt">-f</span><span class="w"> </span><span class="nt">1</span><span class="w"> </span><span class="nt">-p</span><span class="w"> </span><span class="s1">&#39;\.\d+$&#39;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="nt">csvtk</span><span class="w"> </span><span class="nt">add-header</span><span class="w"> </span><span class="nt">-t</span><span class="w"> </span><span class="nt">-n</span><span class="w"> </span><span class="nt">accession</span><span class="o">,</span><span class="nt">accession</span><span class="p">.</span><span class="nc">version</span><span class="o">,</span><span class="nt">taxid</span><span class="o">,</span><span class="nt">gi</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">&gt;</span><span class="w"> </span><span class="nt">accession2taxid</span><span class="p">.</span><span class="nc">map</span>

<span class="err">#</span><span class="w"> </span><span class="nt">human</span>
<span class="err">#</span>
<span class="err">#</span><span class="w">     </span><span class="err">#</span><span class="w"> </span><span class="nt">The</span><span class="w"> </span><span class="nt">human</span><span class="w"> </span><span class="nt">genome</span><span class="w"> </span><span class="nt">file</span><span class="w"> </span><span class="nt">is</span><span class="w"> </span><span class="nt">split</span><span class="w"> </span><span class="nt">into</span><span class="w"> </span><span class="nt">separate</span><span class="w"> </span><span class="nt">files</span><span class="w"> </span><span class="nt">for</span><span class="w"> </span><span class="nt">each</span><span class="w"> </span><span class="nt">chromosomes</span><span class="o">/</span><span class="nt">mitochondrion</span><span class="o">.</span>
<span class="err">#</span><span class="w">     </span><span class="nt">seqkit</span><span class="w"> </span><span class="nt">split2</span><span class="w"> </span><span class="nt">-s</span><span class="w"> </span><span class="nt">1</span><span class="w"> </span><span class="nt">-N</span><span class="w"> </span><span class="nt">GCA_009914755</span><span class="p">.</span><span class="nc">4_T2T-CHM13v2</span><span class="p">.</span><span class="nc">0_genomic</span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span><span class="w"> </span><span class="nt">-O</span><span class="w"> </span><span class="nt">T2T-CHM13v2</span><span class="p">.</span><span class="nc">0</span>
<span class="err">#</span><span class="w">     </span><span class="err">#</span><span class="w"> </span><span class="nt">reaname</span>
<span class="err">#</span><span class="w">     </span><span class="nt">brename</span><span class="w"> </span><span class="nt">-p</span><span class="w"> </span><span class="o">^</span><span class="w"> </span><span class="nt">-r</span><span class="w"> </span><span class="nt">GCA_009914755</span><span class="p">.</span><span class="nc">4_</span><span class="w"> </span><span class="nt">T2T-CHM13v2</span><span class="p">.</span><span class="nc">0</span><span class="o">/*</span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span>
<span class="err">#</span>
<span class="err">#</span><span class="w"> </span><span class="nt">the</span><span class="w"> </span><span class="nt">accession</span><span class="w"> </span><span class="nt">is</span><span class="w"> </span><span class="nt">sligntly</span><span class="w"> </span><span class="nt">different</span><span class="o">,</span><span class="w"> </span><span class="nt">e</span><span class="p">.</span><span class="nc">g</span><span class="o">.,</span><span class="w"> </span><span class="nt">GCA_009914755</span><span class="p">.</span><span class="nc">4_CP068254</span><span class="p">.</span><span class="nc">1</span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span>
<span class="err">#</span><span class="w"> </span><span class="nt">where</span><span class="w"> </span><span class="nt">GCA_009914755</span><span class="p">.</span><span class="nc">4_CP068254</span><span class="p">.</span><span class="nc">1</span><span class="w"> </span><span class="nt">is</span><span class="w"> </span><span class="nt">mapped</span><span class="w"> </span><span class="nt">to</span><span class="w"> </span><span class="nt">a</span><span class="w"> </span><span class="nt">taxid</span><span class="o">.</span>
<span class="err">#</span>
<span class="nt">fd</span><span class="w"> </span><span class="p">.</span><span class="nc">fna</span><span class="p">.</span><span class="nc">gz</span><span class="o">$</span><span class="w"> </span><span class="nt">human</span><span class="o">/</span><span class="nt">T2T-CHM13v2</span><span class="p">.</span><span class="nc">0</span><span class="o">/</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="nt">rush</span><span class="w"> </span><span class="nt">--eta</span><span class="w">  </span><span class="err">\</span>
<span class="w">        </span><span class="s1">&#39;taxid=$(grep {%..} taxdump/taxid.map | cut -f 2); \</span>
<span class="s1">        seqkit seq -ni {} | awk &quot;{print \$1\&quot;\t\&quot;\$1\&quot;\t\&quot;$taxid\&quot;\t0\&quot;}&quot;&#39;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="nt">csvtk</span><span class="w"> </span><span class="nt">replace</span><span class="w"> </span><span class="nt">-Ht</span><span class="w"> </span><span class="nt">-f</span><span class="w"> </span><span class="nt">1</span><span class="w"> </span><span class="nt">-p</span><span class="w"> </span><span class="s1">&#39;\.\d+$&#39;</span><span class="w"> </span><span class="err">\</span>
<span class="w">    </span><span class="o">&gt;&gt;</span><span class="w"> </span><span class="nt">accession2taxid</span><span class="p">.</span><span class="nc">map</span>
</code></pre></div>

</li>
</ol>
<p>Some tests:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># SARS-COV-2 in NCBI taxonomy</span>
<span class="nv">$</span><span class="w"> </span><span class="nv">echo</span><span class="w"> </span><span class="mi">2697049</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">lineage</span><span class="w"> </span><span class="o">-</span><span class="n">t</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">ncbi</span><span class="o">-</span><span class="n">taxdump</span><span class="sr">/ \</span>
<span class="sr">    | csvtk cut -Ht -f 3 \</span>
<span class="sr">    | csvtk unfold -Ht -f 1 -s &quot;;&quot; \</span>
<span class="sr">    | taxonkit lineage -r -n -L --data-dir ncbi-taxdump/</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">pretty</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span>
<span class="mi">10239</span><span class="w">     </span><span class="n">superkingdom</span><span class="w">   </span><span class="n">Viruses</span>
<span class="mi">2559587</span><span class="w">   </span><span class="n">clade</span><span class="w">          </span><span class="n">Riboviria</span>
<span class="mi">2732396</span><span class="w">   </span><span class="n">kingdom</span><span class="w">        </span><span class="n">Orthornavirae</span>
<span class="mi">2732408</span><span class="w">   </span><span class="n">phylum</span><span class="w">         </span><span class="n">Pisuviricota</span>
<span class="mi">2732506</span><span class="w">   </span><span class="n">class</span><span class="w">          </span><span class="n">Pisoniviricetes</span>
<span class="mi">76804</span><span class="w">     </span><span class="n">order</span><span class="w">          </span><span class="n">Nidovirales</span>
<span class="mi">2499399</span><span class="w">   </span><span class="n">suborder</span><span class="w">       </span><span class="n">Cornidovirineae</span>
<span class="mi">11118</span><span class="w">     </span><span class="n">family</span><span class="w">         </span><span class="n">Coronaviridae</span>
<span class="mi">2501931</span><span class="w">   </span><span class="n">subfamily</span><span class="w">      </span><span class="n">Orthocoronavirinae</span>
<span class="mi">694002</span><span class="w">    </span><span class="n">genus</span><span class="w">          </span><span class="n">Betacoronavirus</span>
<span class="mi">2509511</span><span class="w">   </span><span class="n">subgenus</span><span class="w">       </span><span class="n">Sarbecovirus</span>
<span class="mi">694009</span><span class="w">    </span><span class="n">species</span><span class="w">        </span><span class="n">Severe</span><span class="w"> </span><span class="n">acute</span><span class="w"> </span><span class="n">respiratory</span><span class="w"> </span><span class="n">syndrome</span><span class="o">-</span><span class="n">related</span><span class="w"> </span><span class="n">coronavirus</span>
<span class="mi">2697049</span><span class="w">   </span><span class="k">no</span><span class="w"> </span><span class="nn">rank</span><span class="w">        </span><span class="n">Severe</span><span class="w"> </span><span class="n">acute</span><span class="w"> </span><span class="n">respiratory</span><span class="w"> </span><span class="n">syndrome</span><span class="w"> </span><span class="n">coronavirus</span><span class="w"> </span><span class="mi">2</span>

<span class="nv">$</span><span class="w"> </span><span class="nv">echo</span><span class="w"> </span><span class="s">&quot;Severe acute respiratory syndrome coronavirus 2&quot;</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">name2taxid</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="o">/</span>
<span class="n">Severe</span><span class="w"> </span><span class="n">acute</span><span class="w"> </span><span class="n">respiratory</span><span class="w"> </span><span class="n">syndrome</span><span class="w"> </span><span class="n">coronavirus</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span><span class="mi">192491219</span>

<span class="nv">$</span><span class="w"> </span><span class="nv">echo</span><span class="w"> </span><span class="mi">192491219</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">lineage</span><span class="w"> </span><span class="o">-</span><span class="n">t</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="sr">/ \</span>
<span class="sr">    | csvtk cut -Ht -f 3 \</span>
<span class="sr">    | csvtk unfold -Ht -f 1 -s &quot;;&quot; \</span>
<span class="sr">    | taxonkit lineage -r -n -L --data-dir taxdump/</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">pretty</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span>
<span class="mi">1088277216</span><span class="w">   </span><span class="n">superkingdom</span><span class="w">   </span><span class="n">Viruses</span><span class="w">                                        </span>
<span class="mi">38781089</span><span class="w">     </span><span class="n">phylum</span><span class="w">         </span><span class="n">Pisuviricota</span><span class="w">                                   </span>
<span class="mi">1832208221</span><span class="w">   </span><span class="n">class</span><span class="w">          </span><span class="n">Pisoniviricetes</span><span class="w">                                </span>
<span class="mi">1393610206</span><span class="w">   </span><span class="n">order</span><span class="w">          </span><span class="n">Nidovirales</span><span class="w">                                    </span>
<span class="mi">779314330</span><span class="w">    </span><span class="n">family</span><span class="w">         </span><span class="n">Coronaviridae</span><span class="w">                                  </span>
<span class="mi">68549826</span><span class="w">     </span><span class="n">genus</span><span class="w">          </span><span class="n">Betacoronavirus</span><span class="w">                                </span>
<span class="mi">341128742</span><span class="w">    </span><span class="n">species</span><span class="w">        </span><span class="n">Betacoronavirus</span><span class="w"> </span><span class="n">pandemicum</span><span class="w">                     </span>
<span class="mi">192491219</span><span class="w">    </span><span class="n">strain</span><span class="w">         </span><span class="n">Severe</span><span class="w"> </span><span class="n">acute</span><span class="w"> </span><span class="n">respiratory</span><span class="w"> </span><span class="n">syndrome</span><span class="w"> </span><span class="n">coronavirus</span><span class="w"> </span><span class="mi">2</span>


<span class="nv">$</span><span class="w"> </span><span class="nv">echo</span><span class="w"> </span><span class="s">&quot;Escherichia coli&quot;</span><span class="w">  </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">name2taxid</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="o">/</span>
<span class="n">Escherichia</span><span class="w"> </span><span class="n">coli</span><span class="w">        </span><span class="mi">599451526</span>

<span class="nv">$</span><span class="w"> </span><span class="nv">echo</span><span class="w"> </span><span class="mi">599451526</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">taxonkit</span><span class="w"> </span><span class="n">lineage</span><span class="w"> </span><span class="o">-</span><span class="n">t</span><span class="w"> </span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">dir</span><span class="w"> </span><span class="n">taxdump</span><span class="sr">/ \</span>
<span class="sr">    | csvtk cut -Ht -f 3 \</span>
<span class="sr">    | csvtk unfold -Ht -f 1 -s &quot;;&quot; \</span>
<span class="sr">    | taxonkit lineage -r -n -L --data-dir taxdump/</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">cut</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span><span class="w"> </span><span class="o">-</span><span class="n">f</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="w"> </span><span class="o">\</span>
<span class="w">    </span><span class="o">|</span><span class="w"> </span><span class="n">csvtk</span><span class="w"> </span><span class="n">pretty</span><span class="w"> </span><span class="o">-</span><span class="n">Ht</span>
<span class="mi">81602897</span><span class="w">     </span><span class="n">superkingdom</span><span class="w">   </span><span class="n">Bacteria</span><span class="w">           </span>
<span class="mi">1712663402</span><span class="w">   </span><span class="n">phylum</span><span class="w">         </span><span class="n">Pseudomonadota</span><span class="w">     </span>
<span class="mi">1969409366</span><span class="w">   </span><span class="n">class</span><span class="w">          </span><span class="n">Gammaproteobacteria</span>
<span class="mi">1851777887</span><span class="w">   </span><span class="n">order</span><span class="w">          </span><span class="n">Enterobacterales</span><span class="w">   </span>
<span class="mi">1691888815</span><span class="w">   </span><span class="n">family</span><span class="w">         </span><span class="n">Enterobacteriaceae</span><span class="w"> </span>
<span class="mi">1028471294</span><span class="w">   </span><span class="n">genus</span><span class="w">          </span><span class="n">Escherichia</span><span class="w">        </span>
<span class="mi">599451526</span><span class="w">    </span><span class="n">species</span><span class="w">        </span><span class="n">Escherichia</span><span class="w"> </span><span class="n">coli</span>
</code></pre></div>

<h2 id="filtering-or-subsetting-taxdmp-files-to-make-a-custom-taxdmp-with-given-taxids">Filtering or subsetting taxdmp files to make a custom taxdmp with given TaxIDs<a class="headerlink" href="#filtering-or-subsetting-taxdmp-files-to-make-a-custom-taxdmp-with-given-taxids" title="Permanent link">Link</a></h2>
<blockquote>
<p>You want to create a smaller version of the official NCBI taxonomy taxdmp filtered or subset to just the lineages of certain species, for purposes such as creating small test data for testing of tools using taxdmp files.</p>
<p>https://github.com/shenwei356/taxonkit/issues/112</p>
</blockquote>
<p>Step 1:  preparing taxids in the subset tree</p>
<div class="codehilite"><pre><span></span><code># here, only keep nodes at the rank of species
taxonkit list --ids 707,9606 -I &quot;&quot; \
    | taxonkit filter -E species \
    | taxonkit lineage -t \
    | cut -f 3 \
    | sed -s &#39;s/;/\n/g&#39; \
    &gt; taxids.txt

# the root node
echo 1 &gt;&gt; taxids.txt
</code></pre></div>

<p>Step 2: extracting data of needed nodes</p>
<div class="codehilite"><pre><span></span><code>mkdir subset

grep -w -f &lt;(awk &#39;{print &quot;^&quot;$1}&#39; taxids.txt) ~/.taxonkit/nodes.dmp &gt; subset/nodes.dmp
grep -w -f &lt;(awk &#39;{print &quot;^&quot;$1}&#39; taxids.txt) ~/.taxonkit/names.dmp &gt; subset/names.dmp

touch subset/delnodes.dmp subset/merged.dmp
</code></pre></div>

<p>Checking it. Since there are only two leaves here, we just dump the whole tree</p>
<div class="codehilite"><pre><span></span><code><span class="n">$</span><span class="w"> </span><span class="n">wc</span><span class="w"> </span><span class="o">-</span><span class="n">l</span><span class="w"> </span><span class="n">subset</span><span class="cm">/*.dmp</span>
<span class="cm">   0 subset/delnodes.dmp</span>
<span class="cm">   0 subset/merged.dmp</span>
<span class="cm"> 146 subset/names.dmp</span>
<span class="cm">  40 subset/nodes.dmp</span>
<span class="cm"> 186 total</span>

<span class="cm">$ taxonkit list --ids 1 --data-dir subset/ -nr</span>
<span class="cm">1 [no rank] root</span>
<span class="cm">  131567 [cellular root] cellular organisms</span>
<span class="cm">    2 [domain] Bacteria</span>
<span class="cm">      3379134 [kingdom] Pseudomonadati</span>
<span class="cm">        1224 [phylum] Pseudomonadota</span>
<span class="cm">          1236 [class] Gammaproteobacteria</span>
<span class="cm">            135623 [order] Vibrionales</span>
<span class="cm">              641 [family] Vibrionaceae</span>
<span class="cm">                662 [genus] Vibrio</span>
<span class="cm">                  28174 [species] Vibrio ordalii</span>
<span class="cm">    2759 [domain] Eukaryota</span>
<span class="cm">      33154 [clade] Opisthokonta</span>
<span class="cm">        33208 [kingdom] Metazoa</span>
<span class="cm">          6072 [clade] Eumetazoa</span>
<span class="cm">            33213 [clade] Bilateria</span>
<span class="cm">              33511 [clade] Deuterostomia</span>
<span class="cm">                7711 [phylum] Chordata</span>
<span class="cm">                  89593 [subphylum] Craniata</span>
<span class="cm">                    7742 [clade] Vertebrata</span>
<span class="cm">                      7776 [clade] Gnathostomata</span>
<span class="cm">                        117570 [clade] Teleostomi</span>
<span class="cm">                          117571 [clade] Euteleostomi</span>
<span class="cm">                            8287 [superclass] Sarcopterygii</span>
<span class="cm">                              1338369 [clade] Dipnotetrapodomorpha</span>
<span class="cm">                                32523 [clade] Tetrapoda</span>
<span class="cm">                                  32524 [clade] Amniota</span>
<span class="cm">                                    40674 [class] Mammalia</span>
<span class="cm">                                      32525 [clade] Theria</span>
<span class="cm">                                        9347 [clade] Eutheria</span>
<span class="cm">                                          1437010 [clade] Boreoeutheria</span>
<span class="cm">                                            314146 [superorder] Euarchontoglires</span>
<span class="cm">                                              9443 [order] Primates</span>
<span class="cm">                                                376913 [suborder] Haplorrhini</span>
<span class="cm">                                                  314293 [infraorder] Simiiformes</span>
<span class="cm">                                                    9526 [parvorder] Catarrhini</span>
<span class="cm">                                                      314295 [superfamily] Hominoidea</span>
<span class="cm">                                                        9604 [family] Hominidae</span>
<span class="cm">                                                          207598 [subfamily] Homininae</span>
<span class="cm">                                                            9605 [genus] Homo</span>
<span class="cm">                                                              9606 [species] Homo sapiens</span>


<span class="cm">$ echo 28174 | taxonkit lineage -nr --data-dir subset/</span>
<span class="cm">28174   cellular organisms;Bacteria;Pseudomonadota;Gammaproteobacteria;Vibrionales;Vibrionaceae;Vibrio;Vibrio ordalii       Vibrio ordalii  species</span>
</code></pre></div>

<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//taxonkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.81fa17fe.min.js"></script>
      
    
  </body>
</html>