
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/taxonkit/">
      
      
      
        <link rel="next" href="download/">
      
      
      <link rel="icon" href="files/favicon.ico">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.8">
    
    
      
        <title>TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.356b1318.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#taxonkit-a-practical-and-efficient-ncbi-taxonomy-toolkit" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Home
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmark" class="md-nav__link">
    Benchmark
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-out-of-date" class="md-nav__link">
    Method 3: Install via homebrew (out of date)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-compile-from-source-latest-stabledev-version" class="md-nav__link">
    Method 4: Compile from source (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorial
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="bench/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Benchmark
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="chinese/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    中文介绍
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="chinese-dev/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    开发笔记
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Taxid-Changelog
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    More tools
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmark" class="md-nav__link">
    Benchmark
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-out-of-date" class="md-nav__link">
    Method 3: Install via homebrew (out of date)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-compile-from-source-latest-stabledev-version" class="md-nav__link">
    Method 4: Compile from source (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="taxonkit-a-practical-and-efficient-ncbi-taxonomy-toolkit">TaxonKit - A Practical and Efficient NCBI Taxonomy Toolkit<a class="headerlink" href="#taxonkit-a-practical-and-efficient-ncbi-taxonomy-toolkit" title="Permanent link">Link</a></h1>
<ul>
<li><strong>Documents:</strong> <a href="https://bioinf.shenwei.me/taxonkit">https://bioinf.shenwei.me/taxonkit</a>
(<a href="https://bioinf.shenwei.me/taxonkit/usage/"><strong>Usage&amp;Examples</strong></a>,
<a href="https://bioinf.shenwei.me/taxonkit/tutorial/"><strong>Tutorial</strong></a>, <a href="https://bioinf.shenwei.me/taxonkit/chinese/"><strong>中文介绍</strong></a>)</li>
<li><strong>Source code:</strong> <a href="https://github.com/shenwei356/taxonkit">https://github.com/shenwei356/taxonkit</a>
<a href="https://github.com/shenwei356/taxonkit"><img alt="GitHub stars" src="https://img.shields.io/github/stars/shenwei356/taxonkit.svg?style=social&amp;label=Star&amp;?maxAge=2592000" /></a>
<a href="https://github.com/shenwei356/taxonkit/blob/master/LICENSE"><img alt="license" src="https://img.shields.io/github/license/shenwei356/taxonkit.svg?maxAge=2592000" /></a>
<a href="https://golang.org"><img alt="Built with GoLang" src="https://img.shields.io/badge/powered_by-go-6362c2.svg?style=flat" /></a></li>
<li><strong>Latest version:</strong> <a href="https://github.com/shenwei356/taxonkit/releases"><img alt="Latest Version" src="https://img.shields.io/github/release/shenwei356/taxonkit.svg?style=flat?maxAge=86400" /></a>
<a href="https://bioinf.shenwei.me/taxonkit/download/"><img alt="Github Releases" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/total.svg?maxAge=3600" /></a>
<a href="https://bioinf.shenwei.me/taxonkit/download/"><img alt="Cross-platform" src="https://img.shields.io/badge/platform-any-ec2eb4.svg?style=flat" /></a>
<a href="https://anaconda.org/bioconda/taxonkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/taxonkit/badges/version.svg" /></a></li>
<li><strong><a href="#citation">Please cite</a>:</strong> <a href="https://www.sciencedirect.com/science/article/pii/S1673852721000837">https://doi.org/10.1016/j.jgg.2021.03.006</a>
<a href="https://scholar.google.com/citations?view_op=view_citation&amp;hl=en&amp;user=wHF3Lm8AAAAJ&amp;citation_for_view=wHF3Lm8AAAAJ:ULOm3_A8WrAC"><img alt="Citation Badge" src="https://api.juleskreuer.eu/citation-badge.php?doi=10.1016/j.jgg.2021.03.006" /></a></li>
<li><a href="https://github.com/bioforensics/pytaxonkit">pytaxonkit</a>, Python bindings for TaxonKit.</li>
</ul>
<p>Related projects:</p>
<ul>
<li><a href="https://github.com/shenwei356/taxid-changelog"><strong>Taxid-Changelog</strong></a>: Tracking all changes of TaxIds, including deletion, new adding, merge, reuse, and rank/name changes.</li>
<li><a href="https://github.com/shenwei356/gtdb-taxdump">GTDB taxdump</a>: GTDB taxonomy taxdump files with trackable TaxIds.</li>
<li><a href="https://github.com/shenwei356/ictv-taxdump">ICTV taxdump</a>: NCBI-style taxdump files for International Committee on Taxonomy of Viruses (ICTV)</li>
</ul>
<h2 id="table-of-contents">Table of Contents<a class="headerlink" href="#table-of-contents" title="Permanent link">Link</a></h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#features">Features</a></li>
<li><a href="#subcommands">Subcommands</a></li>
<li><a href="#benchmark">Benchmark</a></li>
<li><a href="#dataset">Dataset</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#command-line-completion">Command-line completion</a></li>
<li><a href="#citation">Citation</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#license">License</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="features">Features<a class="headerlink" href="#features" title="Permanent link">Link</a></h2>
<ul>
<li><strong>Easy to install</strong> (<a href="http://bioinf.shenwei.me/taxonkit/download/">download</a>)<ul>
<li>Statically linked executable binaries for multiple platforms (Linux/Windows/macOS, amd64/arm64) </li>
<li>Light weight and out-of-the-box, no dependencies, no compilation, no configuration</li>
<li>No database building, just download <a href="https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz">NCBI taxonomy data</a> and uncompress to <code>$HOME/.taxonkit</code></li>
</ul>
</li>
<li><strong>Easy to use</strong> (<a href="http://bioinf.shenwei.me/taxonkit/usage/">usages and examples</a>)<ul>
<li>Supporting <a href="#bash-completion">bash-completion</a></li>
<li>Fast (see <a href="#benchmark">benchmark</a>), multiple-CPUs supported, most operations cost 2-10s.</li>
<li>Detailed usages and examples</li>
<li>Supporting STDIN and (gzipped) input/output file, easily integrated in pipe</li>
</ul>
</li>
<li><strong>Versatile commands</strong> <ul>
<li><a href="http://bioinf.shenwei.me/taxonkit/usage/">Usage and examples</a></li>
<li>Featured command: <a href="https://github.com/shenwei356/taxid-changelog">tracking monthly changelog of all TaxIds</a></li>
<li>Featured command: <a href="https://bioinf.shenwei.me/taxonkit/usage/#reformat">reformating lineage into format of seven-level ("superkingdom/kingdom, phylum, class, order, family, genus, species"</a></li>
<li>Featured command: <a href="http://bioinf.shenwei.me/taxonkit/usage/#filter">filtering taxiDs by a rank range</a>, e.g., at or below genus rank.</li>
<li>Featured command: <a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump"><strong>Create NCBI-style taxdump files for custom taxonomy, e.g., GTDB and ICTV</strong></a></li>
</ul>
</li>
</ul>
<h2 id="subcommands">Subcommands<a class="headerlink" href="#subcommands" title="Permanent link">Link</a></h2>
<table>
<thead>
<tr>
<th style="text-align: left;">Subcommand</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#list"><code>list</code></a></td>
<td style="text-align: left;">List taxonomic subtrees (TaxIds) bellow given TaxIds</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#lineage"><code>lineage</code></a></td>
<td style="text-align: left;">Query taxonomic lineage of given TaxIds</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#reformat"><code>reformat</code></a></td>
<td style="text-align: left;">Reformat lineage in canonical ranks</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#name2taxid"><code>name2taxid</code></a></td>
<td style="text-align: left;">Convert scientific names to TaxIds</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#filter"><code>filter</code></a></td>
<td style="text-align: left;">Filter TaxIds by taxonomic rank range</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#lca"><code>lca</code></a></td>
<td style="text-align: left;">Compute lowest common ancestor (LCA) for TaxIds</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#taxid-changelog"><code>taxid-changelog</code></a></td>
<td style="text-align: left;">Create TaxId changelog from dump archives</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#profile2cami"><code>profile2cami</code></a><sup>*</sup></td>
<td style="text-align: left;">Convert metagenomic profile table to CAMI format</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#cami-filter"><code>cami-filter</code></a><sup>*</sup></td>
<td style="text-align: left;">Remove taxa of given TaxIds and their descendants in CAMI metagenomic profile</td>
</tr>
<tr>
<td style="text-align: left;"><a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump"><code>create-taxdump</code></a><sup>*</sup></td>
<td style="text-align: left;">Create NCBI-style taxdump files for custom taxonomy, e.g., GTDB and ICTV</td>
</tr>
</tbody>
</table>
<p>Note: <sup>*</sup>New commands since the publication.</p>
<p><img src="taxonkit.jpg" alt="taxonkit" width="700" align="center" /></p>
<h2 id="benchmark">Benchmark<a class="headerlink" href="#benchmark" title="Permanent link">Link</a></h2>
<ol>
<li>Getting complete lineage for given TaxIds</li>
</ol>
<p><img src="bench/bench.get_lineage.reformat.tsv.png" alt="" width="600" align="center" /></p>
<p>Versions: ETE=3.1.2, taxopy=0.5.0 (<a href="https://github.com/shenwei356/taxonkit/issues/47">faster since 0.6.0</a>), TaxonKit=0.7.2.</p>
<h2 id="dataset">Dataset<a class="headerlink" href="#dataset" title="Permanent link">Link</a></h2>
<ol>
<li>Download and uncompress <code>taxdump.tar.gz</code>: ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz </li>
<li>Copy <code>names.dmp</code>, <code>nodes.dmp</code>, <code>delnodes.dmp</code> and <code>merged.dmp</code> to data directory: <code>$HOME/.taxonkit</code>,
e.g., <code>/home/shenwei/.taxonkit</code> ,</li>
<li>Optionally copy to some other directories, and later you can refer to using flag <code>--data-dir</code>,
or environment variable <code>TAXONKIT_DB</code>.</li>
</ol>
<p>All-in-one command:</p>
<div class="codehilite"><pre><span></span><code>wget -c ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz 
tar -zxvf taxdump.tar.gz

mkdir -p $HOME/.taxonkit
cp names.dmp nodes.dmp delnodes.dmp merged.dmp $HOME/.taxonkit
</code></pre></div>

<p><strong>Update dataset</strong>: Simply re-download the taxdump files, uncompress and override old ones.</p>
<h2 id="installation">Installation<a class="headerlink" href="#installation" title="Permanent link">Link</a></h2>
<p>Go to <a href="https://bioinf.shenwei.me/taxonkit/download">Download Page</a> for more download options and changelogs.</p>
<p><code>TaxonKit</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/taxonkit/releases">release</a> page.</p>
<h4 id="method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)<a class="headerlink" href="#method-1-download-binaries-latest-stabledev-version" title="Permanent link">Link</a></h4>
<p>Just <a href="https://github.com/shenwei356/taxonkit/releases">download</a> compressed
executable file of your operating system,
and uncompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<div class="codehilite"><pre><span></span><code>sudo cp taxonkit /usr/local/bin/
</code></pre></div>

</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<div class="codehilite"><pre><span></span><code>mkdir -p $HOME/bin/; cp taxonkit $HOME/bin/
</code></pre></div>

</li>
</ol>
</li>
<li>
<p><strong>For Windows</strong>, just copy <code>taxonkit.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-conda-latest-stable-version">Method 2: Install via conda  (latest stable version) <a href="https://bioinf.shenwei.me/taxonkit/download/"><img alt="Install-with-conda" src="https://anaconda.org/bioconda/taxonkit/badges/installer/conda.svg" /></a> <a href="https://anaconda.org/bioconda/taxonkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/taxonkit/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/taxonkit"><img alt="downloads" src="https://anaconda.org/bioconda/taxonkit/badges/downloads.svg" /></a><a class="headerlink" href="#method-2-install-via-conda-latest-stable-version" title="Permanent link">Link</a></h4>
<div class="codehilite"><pre><span></span><code>conda install -c bioconda taxonkit
</code></pre></div>

<h4 id="method-3-install-via-homebrew-out-of-date">Method 3: Install via homebrew (out of date)<a class="headerlink" href="#method-3-install-via-homebrew-out-of-date" title="Permanent link">Link</a></h4>
<div class="codehilite"><pre><span></span><code>brew install brewsci/bio/taxonkit
</code></pre></div>

<h4 id="method-4-compile-from-source-latest-stabledev-version">Method 4: Compile from source (latest stable/dev version)<a class="headerlink" href="#method-4-compile-from-source-latest-stabledev-version" title="Permanent link">Link</a></h4>
<ol>
<li>
<p><a href="https://go.dev/doc/install">Install go</a></p>
<div class="codehilite"><pre><span></span><code><span class="n">wget</span><span class="w"> </span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">go</span><span class="o">.</span><span class="n">dev</span><span class="o">/</span><span class="n">dl</span><span class="o">/</span><span class="n">go1</span><span class="o">.</span><span class="mf">17.13</span><span class="o">.</span><span class="n">linux</span><span class="o">-</span><span class="n">amd64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>

<span class="n">tar</span><span class="w"> </span><span class="o">-</span><span class="n">zxf</span><span class="w"> </span><span class="n">go1</span><span class="o">.</span><span class="mf">17.13</span><span class="o">.</span><span class="n">linux</span><span class="o">-</span><span class="n">amd64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="o">-</span><span class="n">C</span><span class="w"> </span><span class="o">$</span><span class="n">HOME</span><span class="o">/</span>

<span class="c1"># or </span>
<span class="c1">#   echo &quot;export PATH=$PATH:$HOME/go/bin&quot; &gt;&gt; ~/.bashrc</span>
<span class="c1">#   source ~/.bashrc</span>
<span class="k">export</span><span class="w"> </span><span class="n">PATH</span><span class="o">=$</span><span class="n">PATH</span><span class="p">:</span><span class="o">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">go</span><span class="o">/</span><span class="n">bin</span>
</code></pre></div>

</li>
<li>
<p>Compile TaxonKit</p>
<div class="codehilite"><pre><span></span><code><span class="err">#</span><span class="w"> </span><span class="o">-------------</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">latest</span><span class="w"> </span><span class="n">stable</span><span class="w"> </span><span class="n">version</span><span class="w"> </span><span class="o">-------------</span>

<span class="k">go</span><span class="w"> </span><span class="k">get</span><span class="w"> </span><span class="o">-</span><span class="n">v</span><span class="w"> </span><span class="o">-</span><span class="n">u</span><span class="w"> </span><span class="n">github</span><span class="p">.</span><span class="n">com</span><span class="o">/</span><span class="n">shenwei356</span><span class="o">/</span><span class="n">taxonkit</span><span class="o">/</span><span class="n">taxonkit</span>

<span class="err">#</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">executable</span><span class="w"> </span><span class="nc">binary</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">located</span><span class="w"> </span><span class="ow">in</span><span class="err">:</span>
<span class="err">#</span><span class="w">   </span><span class="o">~/</span><span class="k">go</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span><span class="n">taxonkit</span>
<span class="err">#</span><span class="w"> </span><span class="n">You</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="n">also</span><span class="w"> </span><span class="n">move</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">anywhere</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="err">$</span><span class="k">PATH</span>
<span class="n">mkdir</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span>
<span class="n">cp</span><span class="w"> </span><span class="o">~/</span><span class="k">go</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span><span class="n">taxonkit</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span>

<span class="err">#</span><span class="w"> </span><span class="o">---------------</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">development</span><span class="w"> </span><span class="n">version</span><span class="w"> </span><span class="o">--------------</span>

<span class="n">git</span><span class="w"> </span><span class="n">clone</span><span class="w"> </span><span class="nl">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="p">.</span><span class="n">com</span><span class="o">/</span><span class="n">shenwei356</span><span class="o">/</span><span class="n">taxonkit</span>
<span class="n">cd</span><span class="w"> </span><span class="n">taxonkit</span><span class="o">/</span><span class="n">taxonkit</span><span class="o">/</span>
<span class="k">go</span><span class="w"> </span><span class="n">build</span>

<span class="err">#</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">executable</span><span class="w"> </span><span class="nc">binary</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">located</span><span class="w"> </span><span class="ow">in</span><span class="err">:</span>
<span class="err">#</span><span class="w">   </span><span class="p">.</span><span class="o">/</span><span class="n">taxonkit</span>
<span class="err">#</span><span class="w"> </span><span class="n">You</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="n">also</span><span class="w"> </span><span class="n">move</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">anywhere</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="err">$</span><span class="k">PATH</span>
<span class="n">mkdir</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span>
<span class="n">cp</span><span class="w"> </span><span class="p">.</span><span class="o">/</span><span class="n">taxonkit</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span>
</code></pre></div>

</li>
</ol>
<h2 id="bash-completion">Bash-completion<a class="headerlink" href="#bash-completion" title="Permanent link">Link</a></h2>
<p>Supported shell: bash|zsh|fish|powershell</p>
<p>Bash:</p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> generate completion shell
taxonkit genautocomplete --shell bash

<span class="gh">#</span> configure if never did.
<span class="gh">#</span> install bash-completion if the &quot;complete&quot; command is not found.
echo &quot;for bcfile in ~/.bash_completion.d/* ; do source \$bcfile; done&quot; &gt;&gt; ~/.bash_completion
echo &quot;source ~/.bash_completion&quot; &gt;&gt; ~/.bashrc
</code></pre></div>

<p>Zsh:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># generate completion shell</span>
<span class="n">taxonkit</span><span class="w"> </span><span class="n">genautocomplete</span><span class="w"> </span><span class="o">--</span><span class="n">shell</span><span class="w"> </span><span class="n">zsh</span><span class="w"> </span><span class="o">--</span><span class="n">file</span><span class="w"> </span><span class="o">~/.</span><span class="n">zfunc</span><span class="o">/</span><span class="n">_taxonkit</span>

<span class="c1"># configure if never did</span>
<span class="n">echo</span><span class="w"> </span><span class="s1">&#39;fpath=( ~/.zfunc &quot;${fpath[@]}&quot; )&#39;</span><span class="w"> </span><span class="o">&gt;&gt;</span><span class="w"> </span><span class="o">~/.</span><span class="n">zshrc</span>
<span class="n">echo</span><span class="w"> </span><span class="s2">&quot;autoload -U compinit; compinit&quot;</span><span class="w"> </span><span class="o">&gt;&gt;</span><span class="w"> </span><span class="o">~/.</span><span class="n">zshrc</span>
</code></pre></div>

<p>fish:</p>
<div class="codehilite"><pre><span></span><code>taxonkit genautocomplete --shell fish --file ~/.config/fish/completions/taxonkit.fish
</code></pre></div>

<h2 id="citation">Citation<a class="headerlink" href="#citation" title="Permanent link">Link</a></h2>
<p>If you use TaxonKit in your work, please cite:</p>
<blockquote>
<p>Shen, W., Ren, H., TaxonKit: a practical and efficient NCBI Taxonomy toolkit,
Journal of Genetics and Genomics, <a href="https://www.sciencedirect.com/science/article/pii/S1673852721000837">https://doi.org/10.1016/j.jgg.2021.03.006</a> <a href="https://scholar.google.com/citations?view_op=view_citation&amp;hl=en&amp;user=wHF3Lm8AAAAJ&amp;citation_for_view=wHF3Lm8AAAAJ:ULOm3_A8WrAC"><img alt="Citation Badge" src="https://api.juleskreuer.eu/citation-badge.php?doi=10.1016/j.jgg.2021.03.006" /></a></p>
</blockquote>
<h2 id="contact">Contact<a class="headerlink" href="#contact" title="Permanent link">Link</a></h2>
<p><a href="https://github.com/shenwei356/taxonkit/issues">Create an issue</a> to report bugs,
propose new functions or ask for help.</p>
<h2 id="license">License<a class="headerlink" href="#license" title="Permanent link">Link</a></h2>
<p><a href="https://github.com/shenwei356/taxonkit/blob/master/LICENSE">MIT License</a></p>
<h2 id="starchart">Starchart<a class="headerlink" href="#starchart" title="Permanent link">Link</a></h2>
<p><img src="https://starchart.cc/shenwei356/taxonkit.svg" alt="Stargazers over time" style="max-width: 100%"></p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.81fa17fe.min.js"></script>
      
    
  </body>
</html>