
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/taxonkit/usage/">
      
      <link rel="icon" href="../files/favicon.ico">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.5">
    
    
      
        <title>Usage - TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.ee0f47ba.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#00bdd6">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage-and-examples" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/taxonkit/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../download/" class="md-nav__link">
        Download
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Usage
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Usage
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#before-use" class="md-nav__link">
    Before use
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonkit" class="md-nav__link">
    taxonkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#list" class="md-nav__link">
    list
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#lineage" class="md-nav__link">
    lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reformat" class="md-nav__link">
    reformat
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#name2taxid" class="md-nav__link">
    name2taxid
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    filter
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#lca" class="md-nav__link">
    lca
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxid-changelog" class="md-nav__link">
    taxid-changelog
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-taxdump" class="md-nav__link">
    create-taxdump
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genautocomplete" class="md-nav__link">
    genautocomplete
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#profile2cami" class="md-nav__link">
    profile2cami
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cami-filter" class="md-nav__link">
    cami-filter
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/" class="md-nav__link">
        Tutorial
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bench/" class="md-nav__link">
        Benchmark
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../chinese/" class="md-nav__link">
        中文介绍
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../chinese-dev/" class="md-nav__link">
        开发笔记
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
        Taxid-Changelog
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        More tools
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#before-use" class="md-nav__link">
    Before use
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonkit" class="md-nav__link">
    taxonkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#list" class="md-nav__link">
    list
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#lineage" class="md-nav__link">
    lineage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reformat" class="md-nav__link">
    reformat
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#name2taxid" class="md-nav__link">
    name2taxid
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    filter
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#lca" class="md-nav__link">
    lca
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxid-changelog" class="md-nav__link">
    taxid-changelog
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-taxdump" class="md-nav__link">
    create-taxdump
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genautocomplete" class="md-nav__link">
    genautocomplete
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#profile2cami" class="md-nav__link">
    profile2cami
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cami-filter" class="md-nav__link">
    cami-filter
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/shenwei356/taxonkit/edit/master/docs/usage.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<h1 id="usage-and-examples">Usage and Examples</h1>
<p>Table of Contents</p>
<!-- TOC -->
<ul>
<li><a href="#usage-and-examples">Usage and Examples</a></li>
<li><a href="#before-use">Before use</a></li>
<li><a href="#taxonkit">taxonkit</a></li>
<li><a href="#list">list</a></li>
<li><a href="#lineage">lineage</a></li>
<li><a href="#reformat">reformat</a></li>
<li><a href="#name2taxid">name2taxid</a></li>
<li><a href="#filter">filter</a></li>
<li><a href="#lca">lca</a></li>
<li><a href="#taxid-changelog">taxid-changelog</a></li>
<li><a href="#profile2cami">profile2cami</a></li>
<li><a href="#create-taxdump">create-taxdump</a></li>
<li><a href="#genautocomplete">genautocomplete</a></li>
</ul>
<!-- /TOC -->

<h2 id="before-use">Before use</h2>
<ol>
<li>Download and uncompress <code>taxdump.tar.gz</code>: ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz </li>
<li>Copy <code>names.dmp</code>, <code>nodes.dmp</code>, <code>delnodes.dmp</code> and <code>merged.dmp</code> to data directory: <code>$HOME/.taxonkit</code>,
e.g., <code>/home/shenwei/.taxonkit</code> ,</li>
<li>Optionally copy to some other directories, and later you can refer to using flag <code>--data-dir</code>,
or environment variable <code>TAXONKIT_DB</code>.</li>
</ol>
<p>All-in-one command:</p>
<pre><code>wget -c ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz 
tar -zxvf taxdump.tar.gz

mkdir -p $HOME/.taxonkit
cp names.dmp nodes.dmp delnodes.dmp merged.dmp $HOME/.taxonkit
</code></pre>
<p><strong>Update dataset</strong>: Simply re-download the taxdump files, uncompress and override old ones.</p>
<h2 id="taxonkit">taxonkit</h2>
<pre><code class="language-text">TaxonKit - A Practical and Efficient NCBI Taxonomy Toolkit

Version: 0.13.0

Author: Wei Shen &lt;shenwei356@gmail.com&gt;

Source code: https://github.com/shenwei356/taxonkit
Documents  : https://bioinf.shenwei.me/taxonkit
Citation   : https://www.sciencedirect.com/science/article/pii/S1673852721000837

Dataset:

    Please download and uncompress &quot;taxdump.tar.gz&quot;:
    ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz

    and copy &quot;names.dmp&quot;, &quot;nodes.dmp&quot;, &quot;delnodes.dmp&quot; and &quot;merged.dmp&quot; to data directory:
    &quot;/home/shenwei/.taxonkit&quot;

    or some other directory, and later you can refer to using flag --data-dir,
    or environment variable TAXONKIT_DB

Usage:
  taxonkit [command] 

Available Commands:
  cami-filter     Remove taxa of given TaxIds and their descendants in CAMI metagenomic profile
  create-taxdump  Create NCBI-style taxdump files for custom taxonomy, e.g., GTDB and ICTV
  filter          Filter TaxIds by taxonomic rank range
  genautocomplete generate shell autocompletion script (bash|zsh|fish|powershell)
  lca             Compute lowest common ancestor (LCA) for TaxIds
  lineage         Query taxonomic lineage of given TaxIds
  list            List taxonomic subtrees of given TaxIds
  name2taxid      Convert scientific names to TaxIds
  profile2cami    Convert metagenomic profile table to CAMI format
  reformat        Reformat lineage in canonical ranks
  taxid-changelog Create TaxId changelog from dump archives
  version         print version information and check for update

Flags:
      --data-dir string   directory containing nodes.dmp and names.dmp (default &quot;/home/shenwei/.taxonkit&quot;)
  -h, --help              help for taxonkit
      --line-buffered     use line buffering on output, i.e., immediately writing to stdin/file for every line of output
  -o, --out-file string   out file (&quot;-&quot; for stdout, suffix .gz for gzipped out) (default &quot;-&quot;)
  -j, --threads int       number of CPUs. 4 is enough (default 4)
      --verbose           print verbose information

</code></pre>
<h2 id="list">list</h2>
<p>Usage</p>
<pre><code class="language-text">List taxonomic subtrees of given TaxIds

Examples:

    $ taxonkit list --ids 9606 -n -r --indent &quot;    &quot;
    9606 [species] Homo sapiens
        63221 [subspecies] Homo sapiens neanderthalensis
        741158 [subspecies] Homo sapiens subsp. 'Denisova'

    $ taxonkit list --ids 9606 --indent &quot;&quot;
    9606
    63221
    741158

Usage:
  taxonkit list [flags]

Flags:
  -h, --help            help for list
  -i, --ids string      TaxId(s), multiple values should be separated by comma
  -I, --indent string   indent (default &quot;  &quot;)
  -J, --json            output in JSON format. you can save the result in file with suffix &quot;.json&quot; and open with modern text editor
  -n, --show-name       output scientific name
  -r, --show-rank       output rank

</code></pre>
<p>Examples</p>
<ol>
<li>
<p>Default usage.</p>
<pre><code>$ taxonkit list --ids 9605,239934
9605
9606
    63221
    741158
1425170
2665952
    2665953

239934
239935
    349741
512293
    512294
    1131822
    1262691
    1263034
1679444
2608915
    1131336
...
</code></pre>
</li>
<li>
<p>Removing indent. The list could be used to extract sequences from BLAST database with <code>blastdbcmd</code> (see <a href="http://bioinf.shenwei.me/taxonkit/tutorial/">tutorial</a>)</p>
<pre><code>$ taxonkit list --ids 9605,239934 --indent ""
9605
9606
63221
741158
1425170
2665952
2665953

239934
239935
349741
512293
512294
1131822
1262691
1263034
1679444
...
</code></pre>
<p><strong>Performance:</strong> Time and memory usage for whole taxon tree:</p>
<pre><code>$ # emptying the buffers cache
$ su -c "free &amp;&amp; sync &amp;&amp; echo 3 &gt; /proc/sys/vm/drop_caches &amp;&amp; free"

$ memusg -t taxonkit list --ids 1 --indent "" --verbose &gt; t0.txt
21:05:01.782 [INFO] parsing merged file: /home/shenwei/.taxonkit/names.dmp
21:05:01.782 [INFO] parsing names file: /home/shenwei/.taxonkit/names.dmp
21:05:01.782 [INFO] parsing delnodes file: /home/shenwei/.taxonkit/names.dmp
21:05:01.816 [INFO] 61023 merged nodes parsed
21:05:01.889 [INFO] 437929 delnodes parsed
21:05:03.178 [INFO] 2303979 names parsed

elapsed time: 3.290s
peak rss: 742.77 MB
</code></pre>
</li>
<li>
<p>Adding names</p>
<pre><code>$ taxonkit list --show-rank --show-name --indent "    " --ids 9605,239934
9605 [genus] Homo
    9606 [species] Homo sapiens
        63221 [subspecies] Homo sapiens neanderthalensis
        741158 [subspecies] Homo sapiens subsp. 'Denisova'
    1425170 [species] Homo heidelbergensis
    2665952 [no rank] environmental samples
        2665953 [species] Homo sapiens environmental sample

239934 [genus] Akkermansia
    239935 [species] Akkermansia muciniphila
        349741 [strain] Akkermansia muciniphila ATCC BAA-835
    512293 [no rank] environmental samples
        512294 [species] uncultured Akkermansia sp.
        1131822 [species] uncultured Akkermansia sp. SMG25
        1262691 [species] Akkermansia sp. CAG:344
        1263034 [species] Akkermansia muciniphila CAG:154
    1679444 [species] Akkermansia glycaniphila
    2608915 [no rank] unclassified Akkermansia
        1131336 [species] Akkermansia sp. KLE1605
        1574264 [species] Akkermansia sp. KLE1797
...
</code></pre>
<p><strong>Performance:</strong> Time and memory usage for whole taxonomy tree:</p>
<pre><code>$ # emptying the buffers cache
$ su -c "free &amp;&amp; sync &amp;&amp; echo 3 &gt; /proc/sys/vm/drop_caches &amp;&amp; free"

$ memusg -t taxonkit list --show-rank --show-name --ids 1 &gt; t1.txt
elapsed time: 5.341s
peak rss: 1.04 GB
</code></pre>
</li>
<li>
<p>Output in JSON format, you can easily collapse and uncollapse taxonomy tree in modern text editor.</p>
<pre><code>$ taxonkit list --show-rank --show-name --indent "    " --ids 9605,239934 --json
{
    "9605 [genus] Homo": {
        "9606 [species] Homo sapiens": {
            "63221 [subspecies] Homo sapiens neanderthalensis": {
            },
            "741158 [subspecies] Homo sapiens subsp. 'Denisova'": {
            }
        },
        "1425170 [species] Homo heidelbergensis": {
        }
    },
    "239934 [genus] Akkermansia": {
        "239935 [species] Akkermansia muciniphila": {
            "349741 [no rank] Akkermansia muciniphila ATCC BAA-835": {
            }
        },
        "512293 [no rank] environmental samples": {
            "512294 [species] uncultured Akkermansia sp.": {
            },
            "1131822 [species] uncultured Akkermansia sp. SMG25": {
            },
            "1262691 [species] Akkermansia sp. CAG:344": {
            },
            "1263034 [species] Akkermansia muciniphila CAG:154": {
            }
        },
        "1679444 [species] Akkermansia glycaniphila": {
        },
        "2608915 [no rank] unclassified Akkermansia": {
            "1131336 [species] Akkermansia sp. KLE1605": {
            },
            "1574264 [species] Akkermansia sp. KLE1797": {
            },
            "1574265 [species] Akkermansia sp. KLE1798": {
            },
            "1638783 [species] Akkermansia sp. UNK.MGS-1": {
            },
            "1755639 [species] Akkermansia sp. MC_55": {
            }
        }
    }
}
</code></pre>
<p>Snapshot of taxonomy (taxid 1) in kate:
<img alt="taxon.json.png" src="../files/taxon.json.png" /></p>
</li>
</ol>
<h2 id="lineage">lineage</h2>
<p>Usage</p>
<pre><code class="language-text">Query taxonomic lineage of given TaxIds

Input:

  - List of TaxIds, one TaxId per line.
  - Or tab-delimited format, please specify TaxId field 
    with flag -i/--taxid-field (default 1).
  - Supporting (gzipped) file or STDIN.

Output:

  1. Input line data.
  2. (Optional) Status code (-c/--show-status-code), values:
     - &quot;-1&quot; for queries not found in whole database.
     - &quot;0&quot; for deleted TaxIds, provided by &quot;delnodes.dmp&quot;.
     - New TaxIds for merged TaxIds, provided by &quot;merged.dmp&quot;.
     - Taxids for these found in &quot;nodes.dmp&quot;.
  3. Lineage, delimiter can be changed with flag -d/--delimiter.
  4. (Optional) TaxIds taxons in the lineage (-t/--show-lineage-taxids)
  5. (Optional) Name (-n/--show-name)
  6. (Optional) Rank (-r/--show-rank)

Filter out invalid and deleted taxids, and replace merged 
taxids with new ones:

    # input is one-column-taxid
    $ taxonkit lineage -c taxids.txt \
        | awk '$2&gt;0' \
        | cut -f 2-

    # taxids are in 3rd field in a 4-columns tab-delimited file,
    # for $5, where 5 = 4 + 1.
    $ cat input.txt \
        | taxonkit lineage -c -i 3 \
        | csvtk filter2 -H -t -f '$5&gt;0' \
        | csvtk -H -t cut -f -3

Usage:
  taxonkit lineage [flags]

Flags:
  -d, --delimiter string      field delimiter in lineage (default &quot;;&quot;)
  -h, --help                  help for lineage
  -L, --no-lineage            do not show lineage, when user just want names or/and ranks
  -R, --show-lineage-ranks    appending ranks of all levels
  -t, --show-lineage-taxids   appending lineage consisting of taxids
  -n, --show-name             appending scientific name
  -r, --show-rank             appending rank of taxids
  -c, --show-status-code      show status code before lineage
  -i, --taxid-field int       field index of taxid. input data should be tab-separated (default 1)

</code></pre>
<p>Examples</p>
<ol>
<li>
<p>Full lineage:</p>
<pre><code># note that 123124124 is a fake taxid, 3 was deleted, 92489,1458427 were merged
$ cat taxids.txt 
9606
9913
376619
349741
239935
314101
11932
1327037
123124124
3
92489
1458427

$ taxonkit lineage taxids.txt | tee lineage.txt 
19:22:13.077 [WARN] taxid 92489 was merged into 796334
19:22:13.077 [WARN] taxid 1458427 was merged into 1458425
19:22:13.077 [WARN] taxid 123124124 not found
19:22:13.077 [WARN] taxid 3 was deleted
9606    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Euarchontoglires;Primates;Haplorrhini;Simiiformes;Catarrhini;Hominoidea;Hominidae;Homininae;Homo;Homo sapiens
9913    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Laurasiatheria;Artiodactyla;Ruminantia;Pecora;Bovidae;Bovinae;Bos;Bos taurus
376619  cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis;Francisella tularensis subsp. holarctica;Francisella tularensis subsp. holarctica LVS
349741  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835
239935  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
314101  cellular organisms;Bacteria;environmental samples;uncultured murine large bowel bacterium BAC 54B
11932   Viruses;Riboviria;Pararnavirae;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;unclassified Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
1327037 Viruses;Duplodnaviria;Heunggongvirae;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;unclassified Siphoviridae;Croceibacter phage P2559Y
123124124
3
92489   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae
1458427 cellular organisms;Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raicheisms;Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei
</code></pre>
</li>
<li>
<p>Speed.</p>
<pre><code>$ time echo 9606 | taxonkit lineage 
9606    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Euarchontoglires;Primates;Haplorrhini;Simiiformes;Catarrhini;Hominoidea;Hominidae;Homininae;Homo;Homo sapiens

real    0m1.190s
user    0m2.365s
sys     0m0.170s

# all TaxIds
$ time taxonkit list --ids 1 --indent "" | taxonkit lineage &gt; t

real    0m4.249s
user    0m16.418s
sys     0m1.221s
</code></pre>
</li>
<li>
<p>Checking deleted or merged taxids</p>
<pre><code>$ taxonkit lineage --show-status-code taxids.txt | tee lineage.withcode.txt

# valid
$ cat lineage.withcode.txt | awk '$2 &gt; 0' | cut -f 1,2
9606    9606
9913    9913
376619  376619
349741  349741
239935  239935
314101  314101
11932   11932
1327037 1327037
92489   796334
1458427 1458425

# merged
$ cat lineage.withcode.txt | awk '$2 &gt; 0 &amp;&amp; $2 != $1' | cut -f 1,2
92489   796334
1458427 1458425

# deleted
$ cat lineage.withcode.txt | awk '$2 == 0' | cut -f 1
3

# invalid
$ cat lineage.withcode.txt | awk '$2 &lt; 0' | cut -f 1
123124124
</code></pre>
</li>
<li>
<p><strong>Filter out invalid and deleted taxids, and replace merged taxids with new ones</strong>,
  you may install <a href="https://github.com/shenwei356/csvtk">csvtk</a>.</p>
<pre><code># input is one-column-taxid
$ taxonkit lineage -c taxids.txt \
    | awk '$2&gt;0' \
    | cut -f 2-

# taxids are in 3rd field in a 4-columns tab-delimited file,
# for $5, where 5 = 4 + 1.
$ cat input.txt \
    | taxonkit lineage -c -i 3 \
    | csvtk filter2 -H -t -f '$5&gt;0' \
    | csvtk -H -t cut -f -3
</code></pre>
</li>
<li>
<p>Only show name and rank.</p>
<pre><code>$ taxonkit lineage -r -n -L taxids.txt \
    | csvtk pretty -H -t
9606        Homo sapiens                                      species
9913        Bos taurus                                        species
376619      Francisella tularensis subsp. holarctica LVS      strain
349741      Akkermansia muciniphila ATCC BAA-835              strain
239935      Akkermansia muciniphila                           species
314101      uncultured murine large bowel bacterium BAC 54B   species
11932       Mouse Intracisternal A-particle                   species
1327037     Croceibacter phage P2559Y                         species
123124124                                                     
3                                                             
92489       Erwinia oleae                                     species
1458427     Serpentinomonas raichei                           species
</code></pre>
</li>
<li>
<p>Show lineage consisting of taxids:</p>
<pre><code>$ taxonkit lineage -t taxids.txt
9606    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Euarchontoglires;Primates;Haplorrhini;Simiiformes;Catarrhini;Hominoidea;Hominidae;Homininae;Homo;Homo sapiens  131567;2759;33154;33208;6072;33213;33511;7711;89593;7742;7776;117570;117571;8287;1338369;32523;32524;40674;32525;9347;1437010;314146;9443;376913;314293;9526;314295;9604;207598;9605;9606
9913    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Laurasiatheria;Artiodactyla;Ruminantia;Pecora;Bovidae;Bovinae;Bos;Bos taurus   131567;2759;33154;33208;6072;33213;33511;7711;89593;7742;7776;117570;117571;8287;1338369;32523;32524;40674;32525;9347;1437010;314145;91561;9845;35500;9895;27592;9903;9913
376619  cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis;Francisella tularensis subsp. holarctica;Francisella tularensis subsp. holarctica LVS     131567;2;1224;1236;72273;34064;262;263;119857;376619
349741  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835      131567;2;1783257;74201;203494;48461;1647988;239934;239935;349741
239935  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila   131567;2;1783257;74201;203494;48461;1647988;239934;239935
314101  cellular organisms;Bacteria;environmental samples;uncultured murine large bowel bacterium BAC 54B       131567;2;48479;314101
11932   Viruses;Riboviria;Pararnavirae;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;unclassified Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle    10239;2559587;2732397;2732409;2732514;2169561;11632;35276;11749;11932
1327037 Viruses;Duplodnaviria;Heunggongvirae;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;unclassified Siphoviridae;Croceibacter phage P2559Y   10239;2731341;2731360;2731618;2731619;28883;10699;196894;1327037
123124124
3
92489   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae       131567;2;1224;1236;91347;1903409;551;796334
1458427 cellular organisms;Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei    131567;2;1224;28216;80840;80864;2490452;1458425
</code></pre>
<p>or read taxids from STDIN:</p>
<pre><code>$ cat taxids.txt | taxonkit lineage
</code></pre>
</li>
<li>
<p>And ranks of all nodes:</p>
<pre><code>$ echo 2697049 \
    | taxonkit lineage -t -R \
    | csvtk transpose -Ht
2697049
Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2
10239;2559587;2732396;2732408;2732506;76804;2499399;11118;2501931;694002;2509511;694009;2697049
superkingdom;clade;kingdom;phylum;class;order;suborder;family;subfamily;genus;subgenus;species;no rank
</code></pre>
<p>Another way to show lineage detail of a TaxId</p>
<pre><code>$ echo 2697049 \
    | taxonkit lineage -t \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -H -t 
10239     superkingdom   Viruses
2559587   clade          Riboviria
2732396   kingdom        Orthornavirae
2732408   phylum         Pisuviricota
2732506   class          Pisoniviricetes
76804     order          Nidovirales
2499399   suborder       Cornidovirineae
11118     family         Coronaviridae
2501931   subfamily      Orthocoronavirinae
694002    genus          Betacoronavirus
2509511   subgenus       Sarbecovirus
694009    species        Severe acute respiratory syndrome-related coronavirus
2697049   no rank        Severe acute respiratory syndrome coronavirus 2
</code></pre>
</li>
</ol>
<h2 id="reformat">reformat</h2>
<p>Usage</p>
<pre><code class="language-text">Reformat lineage in canonical ranks

Input:

  - List of TaxIds or lineages, one record per line.
    The lineage can be a complete lineage or only one taxonomy name.
  - Or tab-delimited format.
    Plese specify the lineage field with flag -i/--lineage-field (default 2).
    Or specify the TaxId field with flag -I/--taxid-field (default 0),
    which overrides -i/--lineage-field.
  - Supporting (gzipped) file or STDIN.

Output:

  1. Input line data.
  2. Reformated lineage.
  3. (Optional) TaxIds taxons in the lineage (-t/--show-lineage-taxids)

Ambiguous names:

  - Some TaxIds have the same complete lineage, empty result is returned 
    by default. You can use the flag -a/--output-ambiguous-result to
    return one possible result

Output format can be formated by flag --format, available placeholders:

    {k}: superkingdom
    {K}: kingdom
    {p}: phylum
    {c}: class
    {o}: order
    {f}: family
    {g}: genus
    {s}: species
    {t}: subspecies/strain

    {S}: subspecies
    {T}: strain

When these's no nodes of rank &quot;subspecies&quot; nor &quot;strain&quot;,
you can switch on -S/--pseudo-strain to use the node with lowest rank
as subspecies/strain name, if which rank is lower than &quot;species&quot;. 
This flag affects {t}, {S}, {T}.

Output format can contains some escape charactors like &quot;\t&quot;.

Usage:
  taxonkit reformat [flags]

Flags:
  -P, --add-prefix                     add prefixes for all ranks, single prefix for a rank is defined by flag --prefix-X
  -d, --delimiter string               field delimiter in input lineage (default &quot;;&quot;)
  -F, --fill-miss-rank                 fill missing rank with lineage information of the next higher rank
  -f, --format string                  output format, placeholders of rank are needed (default &quot;{k};{p};{c};{o};{f};{g};{s}&quot;)
  -h, --help                           help for reformat
  -i, --lineage-field int              field index of lineage. data should be tab-separated (default 2)
  -r, --miss-rank-repl string          replacement string for missing rank
  -p, --miss-rank-repl-prefix string   prefix for estimated taxon level (default &quot;unclassified &quot;)
  -R, --miss-taxid-repl string         replacement string for missing taxid
  -a, --output-ambiguous-result        output one of the ambigous result
      --prefix-K string                prefix for kingdom, used along with flag -P/--add-prefix (default &quot;K__&quot;)
      --prefix-S string                prefix for subspecies, used along with flag -P/--add-prefix (default &quot;S__&quot;)
      --prefix-T string                prefix for strain, used along with flag -P/--add-prefix (default &quot;T__&quot;)
      --prefix-c string                prefix for class, used along with flag -P/--add-prefix (default &quot;c__&quot;)
      --prefix-f string                prefix for family, used along with flag -P/--add-prefix (default &quot;f__&quot;)
      --prefix-g string                prefix for genus, used along with flag -P/--add-prefix (default &quot;g__&quot;)
      --prefix-k string                prefix for superkingdom, used along with flag -P/--add-prefix (default &quot;k__&quot;)
      --prefix-o string                prefix for order, used along with flag -P/--add-prefix (default &quot;o__&quot;)
      --prefix-p string                prefix for phylum, used along with flag -P/--add-prefix (default &quot;p__&quot;)
      --prefix-s string                prefix for species, used along with flag -P/--add-prefix (default &quot;s__&quot;)
      --prefix-t string                prefix for subspecies/strain, used along with flag -P/--add-prefix (default &quot;t__&quot;)
  -S, --pseudo-strain                  use the node with lowest rank as strain name, only if which rank is lower than &quot;species&quot; and not &quot;subpecies&quot; nor &quot;strain&quot;. It affects {t}, {S}, {T}. This flag needs flag -F
  -t, --show-lineage-taxids            show corresponding taxids of reformated lineage
  -I, --taxid-field int                field index of taxid. input data should be tab-separated. it overrides -i/--lineage-field
  -T, --trim                           do not fill missing rank lower than current rank

</code></pre>
<p>Examples:</p>
<ol>
<li>
<p><strong>For version &gt; 0.8.0, <code>reformat</code> accept input of TaxIds via flag <code>-I/--taxid-field</code></strong>.</p>
<pre><code>$ echo 239935 | taxonkit reformat -I 1
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila

$ echo 349741 | taxonkit reformat -I 1 -f "{k}|{p}|{c}|{o}|{f}|{g}|{s}|{t}" -F -t
349741  Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia|Akkermansia muciniphila|Akkermansia muciniphila ATCC BAA-835    2|74201|203494|48461|1647988|239934|239935|349741
</code></pre>
</li>
<li>
<p>Example lineage (produced by: <code>taxonkit lineage taxids.txt | awk '$2!=""' &gt; lineage.txt</code>).</p>
<pre><code>$ cat lineage.txt
9606    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Euarchontoglires;Primates;Haplorrhini;Simiiformes;Catarrhini;Hominoidea;Hominidae;Homininae;Homo;Homo sapiens
9913    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Laurasiatheria;Artiodactyla;Ruminantia;Pecora;Bovidae;Bovinae;Bos;Bos taurus
376619  cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis;Francisella tularensis subsp. holarctica;Francisella tularensis subsp. holarctica LVS
349741  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835
239935  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
314101  cellular organisms;Bacteria;environmental samples;uncultured murine large bowel bacterium BAC 54B
11932   Viruses;Riboviria;Pararnavirae;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;unclassified Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
1327037 Viruses;Duplodnaviria;Heunggongvirae;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;unclassified Siphoviridae;Croceibacter phage P2559Y
92489   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae
1458427 cellular organisms;Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei
</code></pre>
</li>
<li>
<p>Default output format (<code>"{k};{p};{c};{o};{f};{g};{s}"</code>).</p>
<pre><code># reformated lineages are appended to the input data
$ taxonkit reformat lineage.txt 
...
239935  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila     Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
...

$ 
$ taxonkit reformat lineage.txt | tee lineage.txt.reformat

$ cut -f 1,3 lineage.txt.reformat
9606    Eukaryota;Chordata;Mammalia;Primates;Hominidae;Homo;Homo sapiens
9913    Eukaryota;Chordata;Mammalia;Artiodactyla;Bovidae;Bos;Bos taurus
376619  Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis
349741  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
314101  Bacteria;;;;;;uncultured murine large bowel bacterium BAC 54B
11932   Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
1327037 Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;;Croceibacter phage P2559Y
92489   Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae
1458427 Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei

# aligned 
$ cat lineage.txt \
    | taxonkit reformat \
    | csvtk -H -t cut -f 1,3 \
    | csvtk -H -t sep -f 2 -s ';' -R \
    | csvtk add-header -t -n taxid,kindom,phylum,class,order,family,genus,species \
    | csvtk pretty -t

taxid     kindom      phylum            class                 order                family            genus                        species
-------   ---------   ---------------   -------------------   ------------------   ---------------   --------------------------   -----------------------------------------------
9606      Eukaryota   Chordata          Mammalia              Primates             Hominidae         Homo                         Homo sapiens
9913      Eukaryota   Chordata          Mammalia              Artiodactyla         Bovidae           Bos                          Bos taurus
376619    Bacteria    Proteobacteria    Gammaproteobacteria   Thiotrichales        Francisellaceae   Francisella                  Francisella tularensis
349741    Bacteria    Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae   Akkermansia                  Akkermansia muciniphila
239935    Bacteria    Verrucomicrobia   Verrucomicrobiae      Verrucomicrobiales   Akkermansiaceae   Akkermansia                  Akkermansia muciniphila
314101    Bacteria                                                                                                                uncultured murine large bowel bacterium BAC 54B
11932     Viruses     Artverviricota    Revtraviricetes       Ortervirales         Retroviridae      Intracisternal A-particles   Mouse Intracisternal A-particle
1327037   Viruses     Uroviricota       Caudoviricetes        Caudovirales         Siphoviridae                                   Croceibacter phage P2559Y
92489     Bacteria    Proteobacteria    Gammaproteobacteria   Enterobacterales     Erwiniaceae       Erwinia                      Erwinia oleae
1458427   Bacteria    Proteobacteria    Betaproteobacteria    Burkholderiales      Comamonadaceae    Serpentinomonas              Serpentinomonas raichei
</code></pre>
</li>
<li>
<p>And <code>subspecies/strain</code> (<code>{t}</code>), <code>subspecies</code> (<code>{S}</code>), and <code>strain</code> (<code>{T}</code>) are also available.</p>
<pre><code># default operation
$ echo -ne "239935\n83333\n1408252\n2697049\n2605619\n" \
    | taxonkit lineage -n -r \
    | taxonkit reformat -f '{t};{S};{T}' \
    | csvtk -H -t cut -f 1,4,3,5 \
    | csvtk -H -t sep -f 4 -s ';' -R \
    | csvtk -H -t add-header -n "taxid,rank,name,subspecies/strain,subspecies,strain" \
    | csvtk pretty -t

taxid     rank         name                                              subspecies/strain       subspecies              strain
-------   ----------   -----------------------------------------------   ---------------------   ---------------------   ---------------------
239935    species      Akkermansia muciniphila                                                                           
83333     strain       Escherichia coli K-12                             Escherichia coli K-12                           Escherichia coli K-12
1408252   subspecies   Escherichia coli R178                             Escherichia coli R178   Escherichia coli R178   
2697049   no rank      Severe acute respiratory syndrome coronavirus 2                                                   
2605619   no rank      Escherichia coli O16:H48

# fill missing ranks
# see example below for -F/--fill-miss-rank
#
$ echo -ne "239935\n83333\n1408252\n2697049\n2605619\n" \
    | taxonkit lineage -n -r \
    | taxonkit reformat -f '{t};{S};{T}' --fill-miss-rank \
    | csvtk -H -t cut -f 1,4,3,5 \
    | csvtk -H -t sep -f 4 -s ';' -R \
    | csvtk -H -t add-header -n "taxid,rank,name,subspecies/strain,subspecies,strain" \
    | csvtk pretty -t

taxid     rank         name                                              subspecies/strain                                                                      subspecies                                                                      strain
-------   ----------   -----------------------------------------------   ------------------------------------------------------------------------------------   -----------------------------------------------------------------------------   -------------------------------------------------------------------------
239935    species      Akkermansia muciniphila                           unclassified Akkermansia muciniphila subspecies/strain                                 unclassified Akkermansia muciniphila subspecies                                 unclassified Akkermansia muciniphila strain
83333     strain       Escherichia coli K-12                             Escherichia coli K-12                                                                  unclassified Escherichia coli subspecies                                        Escherichia coli K-12
1408252   subspecies   Escherichia coli R178                             Escherichia coli R178                                                                  Escherichia coli R178                                                           unclassified Escherichia coli R178 strain
2697049   no rank      Severe acute respiratory syndrome coronavirus 2   unclassified Severe acute respiratory syndrome-related coronavirus subspecies/strain   unclassified Severe acute respiratory syndrome-related coronavirus subspecies   unclassified Severe acute respiratory syndrome-related coronavirus strain
2605619   no rank      Escherichia coli O16:H48                          unclassified Escherichia coli subspecies/strain                                        unclassified Escherichia coli subspecies                                        unclassified Escherichia coli strain
</code></pre>
</li>
<li>
<p><strong>When these's no nodes of rank "subspecies" nor "strain",
   you can switch <code>-S/--pseudo-strain</code> to use the node with lowest rank
   as subspecies/strain name, if which rank is lower than "species"</strong>.</p>
<pre><code>$ echo -ne "239935\n83333\n1408252\n2697049\n2605619\n" \
    | taxonkit lineage -n -r \
    | taxonkit reformat -f '{t};{S};{T}' --fill-miss-rank  --pseudo-strain \
    | csvtk -H -t cut -f 1,4,3,5 \
    | csvtk -H -t sep -f 4 -s ';' -R \
    | csvtk -H -t add-header -n "taxid,rank,name,subspecies/strain,subspecies,strain" \
    | csvtk pretty -t

taxid     rank         name                                              subspecies/strain                                        subspecies                                        strain
-------   ----------   -----------------------------------------------   ------------------------------------------------------   -----------------------------------------------   -----------------------------------------------
239935    species      Akkermansia muciniphila                           unclassified Akkermansia muciniphila subspecies/strain   unclassified Akkermansia muciniphila subspecies   unclassified Akkermansia muciniphila strain
83333     strain       Escherichia coli K-12                             Escherichia coli K-12                                    unclassified Escherichia coli subspecies          Escherichia coli K-12
1408252   subspecies   Escherichia coli R178                             Escherichia coli R178                                    Escherichia coli R178                             unclassified Escherichia coli R178 strain
2697049   no rank      Severe acute respiratory syndrome coronavirus 2   Severe acute respiratory syndrome coronavirus 2          Severe acute respiratory syndrome coronavirus 2   Severe acute respiratory syndrome coronavirus 2
2605619   no rank      Escherichia coli O16:H48                          Escherichia coli O16:H48                                 Escherichia coli O16:H48                          Escherichia coli O16:H48
</code></pre>
</li>
<li>
<p>Add prefix (<code>-P/--add-prefix</code>).</p>
<pre><code>$ cat lineage.txt \
    | taxonkit reformat -P \
    | csvtk -H -t cut -f 1,3

9606    k__Eukaryota;p__Chordata;c__Mammalia;o__Primates;f__Hominidae;g__Homo;s__Homo sapiens
9913    k__Eukaryota;p__Chordata;c__Mammalia;o__Artiodactyla;f__Bovidae;g__Bos;s__Bos taurus
376619  k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Thiotrichales;f__Francisellaceae;g__Francisella;s__Francisella tularensis
349741  k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
239935  k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila
314101  k__Bacteria;p__;c__;o__;f__;g__;s__uncultured murine large bowel bacterium BAC 54B
11932   k__Viruses;p__Artverviricota;c__Revtraviricetes;o__Ortervirales;f__Retroviridae;g__Intracisternal A-particles;s__Mouse Intracisternal A-particle
1327037 k__Viruses;p__Uroviricota;c__Caudoviricetes;o__Caudovirales;f__Siphoviridae;g__;s__Croceibacter phage P2559Y
92489   k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacterales;f__Erwiniaceae;g__Erwinia;s__Erwinia oleae
1458427 k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Comamonadaceae;g__Serpentinomonas;s__Serpentinomonas raichei
</code></pre>
</li>
<li>
<p>Show corresponding taxids of reformated lineage (flag <code>-t/--show-lineage-taxids</code>)</p>
<pre><code>$ cat lineage.txt \
    | taxonkit reformat -t \
    | csvtk -H -t cut -f 1,4 \
    | csvtk -H -t sep -f 2 -s ';' -R \
    | csvtk add-header -t -n taxid,kindom,phylum,class,order,family,genus,species \
    | csvtk pretty -t

taxid     kindom   phylum    class     order     family    genus     species
-------   ------   -------   -------   -------   -------   -------   -------
9606      2759     7711      40674     9443      9604      9605      9606
9913      2759     7711      40674     91561     9895      9903      9913
376619    2        1224      1236      72273     34064     262       263
349741    2        74201     203494    48461     1647988   239934    239935
239935    2        74201     203494    48461     1647988   239934    239935
314101    2                                                          314101
11932     10239    2732409   2732514   2169561   11632     11749     11932
1327037   10239    2731618   2731619   28883     10699               1327037
92489     2        1224      1236      91347     1903409   551       796334
1458427   2        1224      28216     80840     80864     2490452   1458425
</code></pre>
</li>
<li>
<p>Use custom symbols for unclassfied ranks (<code>-r/--miss-rank-repl</code>)</p>
<pre><code>$ taxonkit reformat lineage.txt -r "__"  | cut -f 3
Eukaryota;Chordata;Mammalia;Primates;Hominidae;Homo;Homo sapiens
Eukaryota;Chordata;Mammalia;Artiodactyla;Bovidae;Bos;Bos taurus
Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis
Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
Bacteria;__;__;__;__;__;uncultured murine large bowel bacterium BAC 54B
Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;__;Croceibacter phage P2559Y
Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae
Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei

$ taxonkit reformat lineage.txt -r Unassigned | cut -f 3
Eukaryota;Chordata;Mammalia;Primates;Hominidae;Homo;Homo sapiens
Eukaryota;Chordata;Mammalia;Artiodactyla;Bovidae;Bos;Bos taurus
Bacteria;Proteobacteria;Gammaproteobacteria;Thiotrichales;Francisellaceae;Francisella;Francisella tularensis
Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
Bacteria;Unassigned;Unassigned;Unassigned;Unassigned;Unassigned;uncultured murine large bowel bacterium BAC 54B
Viruses;Artverviricota;Revtraviricetes;Ortervirales;Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
Viruses;Uroviricota;Caudoviricetes;Caudovirales;Siphoviridae;Unassigned;Croceibacter phage P2559Y
Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Erwinia;Erwinia oleae
Bacteria;Proteobacteria;Betaproteobacteria;Burkholderiales;Comamonadaceae;Serpentinomonas;Serpentinomonas raichei
</code></pre>
</li>
<li>
<p><strong>Estimate and fill missing rank with original lineage information</strong>
  (<code>-F, --fill-miss-rank</code>, very useful for formatting input data for
  <a href="https://bitbucket.org/biobakery/biobakery/wiki/lefse">LEfSe</a>).
  You can change the prefix "unclassified" using flag <code>-p/--miss-rank-repl-prefix</code>.</p>
<pre><code>$ cat lineage.txt \
    | taxonkit reformat -F \
    | csvtk -H -t cut -f 1,3 \
    | csvtk -H -t sep -f 2 -s ';' -R \
    | csvtk add-header -t -n taxid,kindom,phylum,class,order,family,genus,species \
    | csvtk pretty -t

taxid     kindom      phylum                         class                         order                         family                         genus                             species
-------   ---------   ----------------------------   ---------------------------   ---------------------------   ----------------------------   -------------------------------   -----------------------------------------------
9606      Eukaryota   Chordata                       Mammalia                      Primates                      Hominidae                      Homo                              Homo sapiens
9913      Eukaryota   Chordata                       Mammalia                      Artiodactyla                  Bovidae                        Bos                               Bos taurus
376619    Bacteria    Proteobacteria                 Gammaproteobacteria           Thiotrichales                 Francisellaceae                Francisella                       Francisella tularensis
349741    Bacteria    Verrucomicrobia                Verrucomicrobiae              Verrucomicrobiales            Akkermansiaceae                Akkermansia                       Akkermansia muciniphila
239935    Bacteria    Verrucomicrobia                Verrucomicrobiae              Verrucomicrobiales            Akkermansiaceae                Akkermansia                       Akkermansia muciniphila
314101    Bacteria    unclassified Bacteria phylum   unclassified Bacteria class   unclassified Bacteria order   unclassified Bacteria family   unclassified Bacteria genus       uncultured murine large bowel bacterium BAC 54B
11932     Viruses     Artverviricota                 Revtraviricetes               Ortervirales                  Retroviridae                   Intracisternal A-particles        Mouse Intracisternal A-particle
1327037   Viruses     Uroviricota                    Caudoviricetes                Caudovirales                  Siphoviridae                   unclassified Siphoviridae genus   Croceibacter phage P2559Y
92489     Bacteria    Proteobacteria                 Gammaproteobacteria           Enterobacterales              Erwiniaceae                    Erwinia                           Erwinia oleae
1458427   Bacteria    Proteobacteria                 Betaproteobacteria            Burkholderiales               Comamonadaceae                 Serpentinomonas                   Serpentinomonas raichei
</code></pre>
</li>
<li>
<p>Only some ranks.</p>
<pre><code>$ cat lineage.txt \
    | taxonkit reformat -F -f "{s};{p}"\
    | csvtk -H -t cut -f 1,3 \
    | csvtk -H -t sep -f 2 -s ';' -R \
    | csvtk add-header -t -n taxid,species,phylum \
    | csvtk pretty -t

taxid     species                                           phylum
-------   -----------------------------------------------   ----------------------------
9606      Homo sapiens                                      Chordata
9913      Bos taurus                                        Chordata
376619    Francisella tularensis                            Proteobacteria
349741    Akkermansia muciniphila                           Verrucomicrobia
239935    Akkermansia muciniphila                           Verrucomicrobia
314101    uncultured murine large bowel bacterium BAC 54B   unclassified Bacteria phylum
11932     Mouse Intracisternal A-particle                   Artverviricota
1327037   Croceibacter phage P2559Y                         Uroviricota
92489     Erwinia oleae                                     Proteobacteria
1458427   Serpentinomonas raichei                           Proteobacteria
</code></pre>
</li>
<li>
<p><strong>For some taxids which rank is higher than the lowest rank in <code>-f/--format</code>, use <code>-T/--trim</code> to avoid fill missing rank lower than current rank</strong>.</p>
<pre><code>$ echo -ne "2\n239934\n239935\n" \
    | taxonkit lineage \
    | taxonkit reformat -T \
    | sed -r "s/;+$//" \
    | csvtk -H -t cut -f 1,3

2       Bacteria
239934  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia
239935  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
</code></pre>
</li>
<li>
<p>Support tab in format string</p>
<pre><code>$ echo 9606 \
    | taxonkit lineage \
    | taxonkit reformat -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{S}" \
    | csvtk cut -t -f -2

9606    Eukaryota       Chordata        Mammalia        Primates        Hominidae       Homo    Homo sapiens
</code></pre>
</li>
<li>
<p>List seven-level lineage for all TaxIds.</p>
<pre><code># replace empty taxon with "Unassigned"
$ taxonkit list --ids 1 \
    | taxonkit lineage \
    | taxonkit reformat -r Unassigned 
    | gzip -c &gt; all.lineage.tsv.gz

# tab-delimited seven-levels
$ taxonkit list --ids 1 \
    | taxonkit lineage \
    | taxonkit reformat -r Unassigned -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}" \
    | csvtk cut -H -t -f -2 \
    | head -n 5 \
    | csvtk pretty -H -t

# 8-level
$ taxonkit list --ids 1 \
    | taxonkit lineage \
    | taxonkit reformat -r Unassigned -f "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    | csvtk cut -H -t -f -2 \
    | head -n 5 \
    | csvtk pretty -H -t

# Fill and trim
$ memusg -t -s ' taxonkit list --ids 1 \
    | taxonkit lineage \
    | taxonkit reformat -F -T \
    | sed -r "s/;+$//" \
    | gzip -c &gt; all.lineage.tsv.gz '

elapsed time: 19.930s
peak rss: 6.25 GB
</code></pre>
</li>
<li>
<p>From taxid to 7-ranks lineage:</p>
<pre><code>$ cat taxids.txt | taxonkit lineage | taxonkit reformat

# for taxonkit v0.8.0 or later versions
$ cat taxids.txt | taxonkit reformat -I 1
</code></pre>
</li>
<li>
<p>Some TaxIds have the same complete lineage, empty result is returned 
    by default. You can use the flag <code>-a/--output-ambiguous-result</code> to
    return one possible result. <a href="https://github.com/shenwei356/taxonkit/issues/42">see #42</a></p>
<pre><code>$ echo -ne "2507530\n2516889\n" | taxonkit lineage --data-dir . | taxonkit reformat --data-dir . -t 
19:18:29.770 [WARN] we can't distinguish the TaxIds (2507530, 2516889) for lineage: cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019. But you can use -a/--output-ambiguous-result to return one possible result
19:18:29.770 [WARN] we can't distinguish the TaxIds (2507530, 2516889) for lineage: cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019. But you can use -a/--output-ambiguous-result to return one possible result
2507530 cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019
2516889 cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019

$ echo -ne "2507530\n2516889\n" | taxonkit lineage --data-dir . | taxonkit reformat --data-dir . -t -a
2507530 cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019     Eukaryota;Basidiomycota;Agaricomycetes;Russulales;Russulaceae;Russula;Russula sp. 8 KA-2019      2759;5204;155619;452342;5401;5402;2507530
2516889 cellular organisms;Eukaryota;Opisthokonta;Fungi;Dikarya;Basidiomycota;Agaricomycotina;Agaricomycetes;Agaricomycetes incertae sedis;Russulales;Russulaceae;Russula;unclassified Russula;Russula sp. 8 KA-2019     Eukaryota;Basidiomycota;Agaricomycetes;Russulales;Russulaceae;Russula;Russula sp. 8 KA-2019      2759;5204;155619;452342;5401;5402;2507530
</code></pre>
</li>
</ol>
<h2 id="name2taxid">name2taxid</h2>
<p>Usage</p>
<pre><code class="language-text">Convert scientific names to TaxIds

Attention:

  1. Some TaxIds share the same scientific names, e.g, Drosophila.
     These input lines are duplicated with multiple TaxIds.

    $ echo Drosophila | taxonkit name2taxid | taxonkit lineage -i 2 -r -L
    Drosophila      7215    genus
    Drosophila      32281   subgenus
    Drosophila      2081351 genus

Usage:
  taxonkit name2taxid [flags]

Flags:
  -h, --help             help for name2taxid
  -i, --name-field int   field index of name. data should be tab-separated (default 1)
  -s, --sci-name         only searching scientific names
  -r, --show-rank        show rank

</code></pre>
<p>Examples</p>
<p>Example data</p>
<pre><code>$ cat names.txt
Homo sapiens
Akkermansia muciniphila ATCC BAA-835
Akkermansia muciniphila
Mouse Intracisternal A-particle
Wei Shen
uncultured murine large bowel bacterium BAC 54B
Croceibacter phage P2559Y
</code></pre>
<ol>
<li>
<p>Default.</p>
<pre><code># taxonkit name2taxid names.txt
$ cat names.txt | taxonkit name2taxid | csvtk pretty -H -t
Homo sapiens                                      9606
Akkermansia muciniphila ATCC BAA-835              349741
Akkermansia muciniphila                           239935
Mouse Intracisternal A-particle                   11932
Wei Shen                                          
uncultured murine large bowel bacterium BAC 54B   314101
Croceibacter phage P2559Y                         1327037
</code></pre>
</li>
<li>
<p>Show rank.</p>
<pre><code>$ cat names.txt | taxonkit name2taxid --show-rank | csvtk pretty -H -t
Homo sapiens                                      9606      species
Akkermansia muciniphila ATCC BAA-835              349741    strain
Akkermansia muciniphila                           239935    species
Mouse Intracisternal A-particle                   11932     species
Wei Shen                                                    
uncultured murine large bowel bacterium BAC 54B   314101    species
Croceibacter phage P2559Y                         1327037   species
</code></pre>
</li>
<li>
<p>From name to lineage.</p>
<pre><code>$ cat names.txt | taxonkit name2taxid | taxonkit lineage --taxid-field 2
Homo sapiens    9606    cellular organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Euarchontoglires;Primates;Haplorrhini;Simiiformes;Catarrhini;Hominoidea;Hominidae;Homininae;Homo;Homo sapiens
Akkermansia muciniphila ATCC BAA-835    349741  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835
Akkermansia muciniphila 239935  cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
Mouse Intracisternal A-particle 11932   Viruses;Ortervirales;Retroviridae;unclassified Retroviridae;Intracisternal A-particles;Mouse Intracisternal A-particle
Wei Shen
uncultured murine large bowel bacterium BAC 54B 314101  cellular organisms;Bacteria;environmental samples;uncultured murine large bowel bacterium BAC 54B
Croceibacter phage P2559Y       1327037 Viruses;Caudovirales;Siphoviridae;unclassified Siphoviridae;Croceibacter phage P2559Y
</code></pre>
</li>
<li>
<p><strong>Some TaxIds share the same scientific names</strong>, e.g, Drosophila.</p>
<pre><code>$ echo Drosophila \
    | taxonkit name2taxid \
    | taxonkit lineage -i 2 -r \
    | taxonkit reformat -i 3 \
    | csvtk cut -H -t -f 1,2,4,5 \
    | csvtk pretty -H -t
Drosophila   7215      genus      Eukaryota;Arthropoda;Insecta;Diptera;Drosophilidae;Drosophila;
Drosophila   32281     subgenus   Eukaryota;Arthropoda;Insecta;Diptera;Drosophilidae;Drosophila;
Drosophila   2081351   genus      Eukaryota;Basidiomycota;Agaricomycetes;Agaricales;Psathyrellaceae;Drosophila;
</code></pre>
</li>
</ol>
<h2 id="filter">filter</h2>
<p>Usage</p>
<pre><code class="language-text">Filter TaxIds by taxonomic rank range

Attentions:

  1. Flag -L/--lower-than and -H/--higher-than are exclusive, and can be
     used along with -E/--equal-to which values can be different.
  2. A list of pre-ordered ranks is in ~/.taxonkit/ranks.txt, you can use
     your list by -r/--rank-file, the format specification is below.
  3. All ranks in taxonomy database should be defined in rank file.
  4. Ranks can be removed with black list via -B/--black-list.
  5. TaxIDs with no rank can be optionally discarded by -N/--discard-noranks.
  6. But when filtering with -L/--lower-than, you can use
    -n/--save-predictable-norank to save some special ranks without order,
    where rank of the closest higher node is still lower than rank cutoff.

Rank file:

  1. Blank lines or lines starting with &quot;#&quot; are ignored.
  2. Ranks are in decending order and case ignored.
  3. Ranks with same order should be in one line separated with comma (&quot;,&quot;, no space).
  4. Ranks without order should be assigned a prefix symbol &quot;!&quot; for each rank.

Usage:
  taxonkit filter [flags]

Flags:
  -B, --black-list strings        black list of ranks to discard, e.g., '-B &quot;no rank&quot; -B &quot;clade&quot;
  -N, --discard-noranks           discard all ranks without order, type &quot;taxonkit filter --help&quot; for details
  -R, --discard-root              discard root taxid, defined by --root-taxid
  -E, --equal-to strings          output TaxIds with rank equal to some ranks, multiple values can be separated with comma &quot;,&quot; (e.g., -E &quot;genus,species&quot;), or give multiple times (e.g., -E genus -E species)
  -h, --help                      help for filter
  -H, --higher-than string        output TaxIds with rank higher than a rank, exclusive with --lower-than
      --list-order                list user defined ranks in order, from &quot;$HOME/.taxonkit/ranks.txt&quot;
      --list-ranks                list ordered ranks in taxonomy database, sorted in user defined order
  -L, --lower-than string         output TaxIds with rank lower than a rank, exclusive with --higher-than
  -r, --rank-file string          user-defined ordered taxonomic ranks, type &quot;taxonkit filter --help&quot; for details
      --root-taxid uint32         root taxid (default 1)
  -n, --save-predictable-norank   do not discard some special ranks without order when using -L, where rank of the closest higher node is still lower than rank cutoff
  -i, --taxid-field int           field index of taxid. input data should be tab-separated (default 1)

</code></pre>
<p>Examples</p>
<ol>
<li>
<p>Example data</p>
<pre><code>$ echo 349741 | taxonkit lineage -t | cut -f 3 | sed 's/;/\n/g' &gt; taxids2.txt

$ cat taxids2.txt
131567
2
1783257
74201
203494
48461
1647988
239934
239935
349741

$ cat taxids2.txt  | taxonkit lineage -r | csvtk -Ht cut -f 1,3,2 | csvtk pretty -H -t
131567    no rank        cellular organisms
2         superkingdom   cellular organisms;Bacteria
1783257   clade          cellular organisms;Bacteria;PVC group
74201     phylum         cellular organisms;Bacteria;PVC group;Verrucomicrobia
203494    class          cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae
48461     order          cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales
1647988   family         cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae
239934    genus          cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia
239935    species        cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila
349741    strain         cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835
</code></pre>
</li>
<li>
<p>Equal to certain rank(s) (<code>-E/--equal-to</code>)</p>
<pre><code>$ cat taxids2.txt \
    | taxonkit filter -E Phylum -E Class \
    | taxonkit lineage -r \
    | csvtk -Ht cut -f 1,3,2 \
    | csvtk pretty -H -t
74201    phylum   cellular organisms;Bacteria;PVC group;Verrucomicrobia
203494   class    cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae
</code></pre>
</li>
<li>
<p>Lower than a rank (<code>-L/--lower-than</code>)</p>
<pre><code>$ cat taxids2.txt \
    | taxonkit filter -L genus \
    | taxonkit lineage -r -n -L \
    | csvtk -Ht cut -f 1,3,2 \
    | csvtk pretty -H -t
239935   species   Akkermansia muciniphila
349741   strain    Akkermansia muciniphila ATCC BAA-835
</code></pre>
</li>
<li>
<p>Higher than a rank (<code>-H/--higher-than</code>)</p>
<pre><code>$ cat taxids2.txt \
    | taxonkit filter -H phylum \
    | taxonkit lineage -r -n -L \
    | csvtk -Ht cut -f 1,3,2 \
    | csvtk pretty -H -t
2   superkingdom   Bacteria
</code></pre>
</li>
<li>
<p>"no rank" and "clade" have no rank and can be filter out via <code>-N/--discard-noranks</code>.
  Futher ranks can be removed with black list via <code>-B/--black-list</code></p>
<pre><code># 562 is TaxId of Escherichia coli
$ taxonkit list --ids 562 \
    | taxonkit filter -L species \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk freq -Ht -f 2 -nr \
    | csvtk pretty -H -t
strain       2950
no rank      149
serotype     141
serogroup    95
isolate      1
subspecies   1

$ taxonkit list --ids 562 \
    | taxonkit filter -L species -N -B strain \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk freq -Ht -f 2 -nr \
    | csvtk pretty -H -t
serotype     141
serogroup    95
isolate      1
subspecies   1
</code></pre>
</li>
<li>
<p>Combine of <code>-L/-H</code> with <code>-E</code>.</p>
<pre><code>$ cat taxids2.txt \
    | taxonkit filter -L genus -E genus  \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -H -t
239934   genus     Akkermansia
239935   species   Akkermansia muciniphila
349741   strain    Akkermansia muciniphila ATCC BAA-835
</code></pre>
</li>
<li>
<p><strong>Special cases of "no rank"</strong>. (<code>-n/--save-predictable-norank</code>).
    When filtering with <code>-L/--lower-than</code>, you can use
    <code>-n/--save-predictable-norank</code> to save some special ranks without order,
    where rank of the closest higher node is still lower than rank cutoff.</p>
<pre><code>$ echo -ne "2605619\n1327037\n" \
    | taxonkit lineage -t \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -H -t 
131567    no rank        cellular organisms
2         superkingdom   Bacteria
1224      phylum         Proteobacteria
1236      class          Gammaproteobacteria
91347     order          Enterobacterales
543       family         Enterobacteriaceae
561       genus          Escherichia
562       species        Escherichia coli
2605619   no rank        Escherichia coli O16:H48

10239     superkingdom   Viruses
2731341   clade          Duplodnaviria
2731360   clade          Heunggongvirae
2731618   phylum         Uroviricota
2731619   class          Caudoviricetes
28883     order          Caudovirales
10699     family         Siphoviridae
196894    no rank        unclassified Siphoviridae
1327037   species        Croceibacter phage P2559Y

# save taxids
$  echo -ne "2605619\n1327037\n" \
    | taxonkit lineage -t \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | tee taxids4.txt
131567
2
1224
1236
91347
543
561
562
2605619
10239
2731341
2731360
2731618
2731619
28883
10699
196894
1327037
</code></pre>
<p>Now, filter nodes of rank &lt;= species.</p>
<pre><code>$ cat taxids4.txt \
    | taxonkit filter -L species -E species -N -n \
    | taxonkit lineage -r -n -L \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -H -t
562       species   Escherichia coli
2605619   no rank   Escherichia coli O16:H48
1327037   species   Croceibacter phage P2559Y
</code></pre>
<p>Note that 2605619 (no rank) is saved because its parent node 562 is &lt;= species.</p>
</li>
</ol>
<h2 id="lca">lca</h2>
<p>Usage</p>
<pre><code class="language-text">Compute lowest common ancestor (LCA) for TaxIds

Attention:

  1. This command computes LCA TaxId for a list of TaxIds 
     in a field (&quot;-i/--taxids-field) of tab-delimited file or STDIN.
  2. TaxIDs should have the same separater (&quot;-s/--separater&quot;),
     single charactor separater is prefered.
  3. Empty lines or lines without valid TaxIds in the field are omitted.
  4. If some TaxIds are not found in database, it returns 0.

Examples:

    $ echo 239934, 239935, 349741 | taxonkit lca  -s &quot;, &quot;
    239934, 239935, 349741  239934

    $ echo 239934  239935  349741 9606  | taxonkit lca
    239934 239935 349741 9606       131567

Usage:
  taxonkit lca [flags]

Flags:
  -h, --help               help for lca
  -s, --separater string   separater for TaxIds (default &quot; &quot;)
  -D, --skip-deleted       skip deleted TaxIds and compute with left ones
  -U, --skip-unfound       skip unfound TaxIds and compute with left ones
  -i, --taxids-field int   field index of taxid. input data should be tab-separated (default 1)

</code></pre>
<p>Examples:</p>
<ol>
<li>
<p>Example data</p>
<pre><code>$ taxonkit list --ids 9605 -nr --indent "    "
9605 [genus] Homo
    9606 [species] Homo sapiens
        63221 [subspecies] Homo sapiens neanderthalensis
        741158 [subspecies] Homo sapiens subsp. 'Denisova'
    1425170 [species] Homo heidelbergensis
    2665952 [no rank] environmental samples
        2665953 [species] Homo sapiens environmental sample
</code></pre>
</li>
<li>
<p>Simple one</p>
<pre><code>$ echo 63221 2665953 | taxonkit lca
63221 2665953   9605
</code></pre>
</li>
<li>
<p>Custom field (<code>-i/--taxids-field</code>) and separater (<code>-s/--separater</code>).</p>
<pre><code>$ echo -ne "a\t63221,2665953\nb\t63221, 741158\n"
a       63221,2665953
b       63221, 741158

$ echo -ne "a\t63221,2665953\nb\t63221, 741158\n" \
    | taxonkit lca -i 2 -s ","
a       63221,2665953   9605
b       63221, 741158   9606
</code></pre>
</li>
<li>
<p>Merged TaxIds.</p>
<pre><code># merged
$ echo 92487 92488 92489 | taxonkit lca
10:08:26.578 [WARN] taxid 92489 was merged into 796334
92487 92488 92489       1236
</code></pre>
</li>
<li>
<p>Deleted TaxIds, you can ommit theses and continue compute with left onces with (<code>-D/--skip-deleted</code>).</p>
<pre><code>$ echo 1 2 3 | taxonkit lca 
10:30:17.678 [WARN] taxid 3 not found
1 2 3   0

$ time  echo 1 2 3 | taxonkit lca -D
10:29:31.828 [WARN] taxid 3 was deleted
1 2 3   1
</code></pre>
</li>
<li>
<p>TaxIDs not found in database, you can ommit theses and continue compute with left onces with (<code>-U/--skip-unfound</code>).</p>
<pre><code>$ echo 61021 61022 11111111 | taxonkit lca
10:31:44.929 [WARN] taxid 11111111 not found
61021 61022 11111111    0

$ echo 61021 61022 11111111 | taxonkit lca -U
10:32:02.772 [WARN] taxid 11111111 not found
61021 61022 11111111    2628496
</code></pre>
</li>
</ol>
<h2 id="taxid-changelog">taxid-changelog</h2>
<p>Usage</p>
<pre><code class="language-text">Create TaxId changelog from dump archives

Steps:

    # dependencies:
    #   rush - https://github.com/shenwei356/rush/

    mkdir -p archive; cd archive;

    # --------- download ---------

    # option 1
    # for fast network connection
    wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/taxdmp*.zip

    # option 2
    # for slow network connection
    url=https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/
    wget $url -O - -o /dev/null \
        | grep taxdmp | perl -ne '/(taxdmp_.+?.zip)/; print &quot;$1\n&quot;;' \
        | rush -j 2 -v url=$url 'axel -n 5 {url}/{}' \
            --immediate-output  -c -C download.rush

    # --------- unzip ---------

    ls taxdmp*.zip | rush -j 1 'unzip {} names.dmp nodes.dmp merged.dmp delnodes.dmp -d {@_(.+)\.}'

    # --------- create log ---------

    cd ..
    taxonkit taxid-changelog -i archive -o taxid-changelog.csv.gz --verbose

Output format (CSV):

    # fields        comments
    taxid           # taxid
    version         # version / time of archive, e.g, 2019-07-01
    change          # change, values:
                    #   NEW             newly added
                    #   REUSE_DEL       deleted taxids being reused
                    #   REUSE_MER       merged taxids being reused
                    #   DELETE          deleted
                    #   MERGE           merged into another taxid
                    #   ABSORB          other taxids merged into this one
                    #   CHANGE_NAME     scientific name changed
                    #   CHANGE_RANK     rank changed
                    #   CHANGE_LIN_LIN  lineage taxids remain but lineage remain
                    #   CHANGE_LIN_TAX  lineage taxids changed
                    #   CHANGE_LIN_LEN  lineage length changed
    change-value    # variable values for changes: 
                    #   1) new taxid for MERGE
                    #   2) merged taxids for ABSORB
                    #   3) empty for others
    name            # scientific name
    rank            # rank
    lineage         # complete lineage of the taxid
    lineage-taxids  # taxids of the lineage

    # you can use csvtk to investigate them. e.g.,
    csvtk grep -f taxid -p 1390515 taxid-changelog.csv.gz

Usage:
  taxonkit taxid-changelog [flags]

Flags:
  -i, --archive string   directory containing uncompressed dumped archives
  -h, --help             help for taxid-changelog

</code></pre>
<p><a href="https://github.com/shenwei356/taxid-changelog">Details</a></p>
<ol>
<li>
<p>Example 1  (<em>E.coli</em> with taxid <code>562</code>)</p>
<pre><code>$ pigz -cd taxid-changelog.csv.gz \
    | csvtk grep -f taxid -p 562 \
    | csvtk pretty
taxid   version      change           change-value    name               rank      lineage                                                                                                                            lineage-taxids
562     2014-08-01   NEW                              Escherichia coli   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia coli   131567;2;1224;1236;91347;543;561;562
562     2014-08-01   ABSORB           662101;662104   Escherichia coli   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia coli   131567;2;1224;1236;91347;543;561;562
562     2015-11-01   ABSORB           1637691         Escherichia coli   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia coli   131567;2;1224;1236;91347;543;561;562
562     2016-10-01   CHANGE_LIN_LIN                   Escherichia coli   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli    131567;2;1224;1236;91347;543;561;562
562     2018-06-01   ABSORB           469598          Escherichia coli   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli    131567;2;1224;1236;91347;543;561;562

# merged taxids
$ pigz -cd taxid-changelog.csv.gz \
    | csvtk grep -f taxid -p 662101,662104,1637691,469598 \
    | csvtk pretty
taxid     version      change           change-value   name                        rank      lineage                                                                                                                                     lineage-taxids
469598    2014-08-01   NEW                             Escherichia sp. 3_2_53FAA   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia sp. 3_2_53FAA   131567;2;1224;1236;91347;543;561;469598
469598    2016-10-01   CHANGE_LIN_LIN                  Escherichia sp. 3_2_53FAA   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia sp. 3_2_53FAA    131567;2;1224;1236;91347;543;561;469598
469598    2018-06-01   MERGE            562            Escherichia sp. 3_2_53FAA   species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia sp. 3_2_53FAA    131567;2;1224;1236;91347;543;561;469598
662101    2014-08-01   MERGE            562                                                                                                                                                                                              
662104    2014-08-01   MERGE            562                                                                                                                                                                                              
1637691   2015-04-01   DELETE                                                                                                                                                                                                            
1637691   2015-05-01   REUSE_DEL                       Escherichia sp. MAR         species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia sp. MAR         131567;2;1224;1236;91347;543;561;1637691
1637691   2015-11-01   MERGE            562            Escherichia sp. MAR         species   cellular organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacteriales;Enterobacteriaceae;Escherichia;Escherichia sp. MAR         131567;2;1224;1236;91347;543;561;1637691
</code></pre>
</li>
<li>
<p>Example 2 (SARS-CoV-2).</p>
<pre><code>$ time pigz -cd taxid-changelog.csv.gz \
        | csvtk grep -f taxid -p 2697049 \
        | csvtk pretty
taxid     version      change           change-value   name                                              rank      lineage                                                                                                                                                                                                                                                        lineage-taxids
2697049   2020-02-01   NEW                             Wuhan seafood market pneumonia virus              species   Viruses;Riboviria;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;unclassified Betacoronavirus;Wuhan seafood market pneumonia virus                                                                                               10239;2559587;76804;2499399;11118;2501931;694002;696098;2697049
2697049   2020-03-01   CHANGE_NAME                     Severe acute respiratory syndrome coronavirus 2   no rank   Viruses;Riboviria;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2                                              10239;2559587;76804;2499399;11118;2501931;694002;2509511;694009;2697049
2697049   2020-03-01   CHANGE_RANK                     Severe acute respiratory syndrome coronavirus 2   no rank   Viruses;Riboviria;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2                                              10239;2559587;76804;2499399;11118;2501931;694002;2509511;694009;2697049
2697049   2020-03-01   CHANGE_LIN_LEN                  Severe acute respiratory syndrome coronavirus 2   no rank   Viruses;Riboviria;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2                                              10239;2559587;76804;2499399;11118;2501931;694002;2509511;694009;2697049
2697049   2020-06-01   CHANGE_LIN_LEN                  Severe acute respiratory syndrome coronavirus 2   no rank   Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2   10239;2559587;2732396;2732408;2732506;76804;2499399;11118;2501931;694002;2509511;694009;2697049
2697049   2020-07-01   CHANGE_RANK                     Severe acute respiratory syndrome coronavirus 2   isolate   Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2   10239;2559587;2732396;2732408;2732506;76804;2499399;11118;2501931;694002;2509511;694009;2697049
2697049   2020-08-01   CHANGE_RANK                     Severe acute respiratory syndrome coronavirus 2   no rank   Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2   10239;2559587;2732396;2732408;2732506;76804;2499399;11118;2501931;694002;2509511;694009;2697049

real    0m7.644s
user    0m16.749s
sys     0m3.985s
</code></pre>
</li>
<li>
<p>Example 3 (All subspecies and strain in <em>Akkermansia muciniphila</em> 239935)</p>
<pre><code># species in Akkermansia
$ taxonkit list --show-rank --show-name --indent "    " --ids 239935
239935 [species] Akkermansia muciniphila
    349741 [strain] Akkermansia muciniphila ATCC BAA-835

# check them all  
$ pigz -cd taxid-changelog.csv.gz \
    | csvtk grep -f taxid -P &lt;(taxonkit list --indent "" --ids 239935) \
    | csvtk pretty                                                                                                                                                   lineage-taxids
taxid    version      change           change-value   name                                   rank      lineage                                                                                                                                                                                                         lineage-taxids
239935   2014-08-01   NEW                             Akkermansia muciniphila                species   cellular organisms;Bacteria;Chlamydiae/Verrucomicrobia group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Verrucomicrobiaceae;Akkermansia;Akkermansia muciniphila                                        131567;2;51290;74201;203494;48461;203557;239934;239935
239935   2015-05-01   CHANGE_LIN_TAX                  Akkermansia muciniphila                species   cellular organisms;Bacteria;Chlamydiae/Verrucomicrobia group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila                                            131567;2;51290;74201;203494;48461;1647988;239934;239935
239935   2016-03-01   CHANGE_LIN_TAX                  Akkermansia muciniphila                species   cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila                                                                   131567;2;1783257;74201;203494;48461;1647988;239934;239935
239935   2016-05-01   ABSORB           1834199        Akkermansia muciniphila                species   cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila                                                                   131567;2;1783257;74201;203494;48461;1647988;239934;239935
349741   2014-08-01   NEW                             Akkermansia muciniphila ATCC BAA-835   no rank   cellular organisms;Bacteria;Chlamydiae/Verrucomicrobia group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Verrucomicrobiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835   131567;2;51290;74201;203494;48461;203557;239934;239935;349741
349741   2015-05-01   CHANGE_LIN_TAX                  Akkermansia muciniphila ATCC BAA-835   no rank   cellular organisms;Bacteria;Chlamydiae/Verrucomicrobia group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835       131567;2;51290;74201;203494;48461;1647988;239934;239935;349741
349741   2016-03-01   CHANGE_LIN_TAX                  Akkermansia muciniphila ATCC BAA-835   no rank   cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835                              131567;2;1783257;74201;203494;48461;1647988;239934;239935;349741
349741   2020-07-01   CHANGE_RANK                     Akkermansia muciniphila ATCC BAA-835   strain    cellular organisms;Bacteria;PVC group;Verrucomicrobia;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae;Akkermansia;Akkermansia muciniphila;Akkermansia muciniphila ATCC BAA-835                              131567;2;1783257;74201;203494;48461;1647988;239934;239935;349741
</code></pre>
</li>
</ol>
<p><a href="https://github.com/shenwei356/taxid-changelog">More</a></p>
<h2 id="create-taxdump">create-taxdump</h2>
<p>Usage</p>
<pre><code class="language-text">Create NCBI-style taxdump files for custom taxonomy, e.g., GTDB and ICTV

Input format: 
  0. For GTDB taxonomy file, just use --gtdb.
     We use the numeric assembly accession as the taxon at subspecies rank.
     (without the prefix GCA_ and GCF_, and version number).
  1. The input file should be tab-delimited, at least one column is needed.
  2. Ranks can be given either via the first row or the flag --rank-names.
  3. The column containing the genome/assembly accession is recommended to
     generate TaxId mapping file (taxid.map, id -&gt; taxid).
       -A/--field-accession,    field contaning genome/assembly accession      
       --field-accession-re,    regular expression to extract the accession
     Note that mutiple TaxIds pointing to the same accession are listed as
     comma-seperated integers. 

Attentions:
  1. Names should be distinct in taxa of different ranks.
     But for these missing some taxon nodes, using names of parent nodes is allowed:

       GB_GCA_018897955.1      d__Archaea;p__EX4484-52;c__EX4484-52;o__EX4484-52;f__LFW-46;g__LFW-46;s__LFW-46 sp018897155

     It can also detect duplicate names with different ranks, e.g.,
     the Class and Genus have the same name B47-G6, and the Order and Family
     between them have different names. In this case, we reassign a new TaxId
     by increasing the TaxId until it being distinct.

       GB_GCA_003663585.1      d__Archaea;p__Thermoplasmatota;c__B47-G6;o__B47-G6B;f__47-G6;g__B47-G6;s__B47-G6 sp003663585

  2. Taxa from different parents may have the same name.
     We will assign different TaxIds to them. 

     E.g., in ICTV, many viruses from different species have the same names.
     In practice, we set the &quot;Virus names(s)&quot; as a subspecies rank and also
     specify it as the accession.

       Species             Virus name(s)
       Jerseyvirus SETP3   Salmonella phage SETP7
       Jerseyvirus SETP7   Salmonella phage SETP7

  3. The generated TaxIds are not consecutive numbers, however some tools like MMSeqs2
     required this, you can use the script below for convertion:

     https://github.com/apcamargo/ictv-mmseqs2-protein-database/blob/master/fix_taxdump.py

Usage:
  taxonkit create-taxdump [flags] 

Flags:
  -A, --field-accession int         field index of assembly accession (genome ID), for outputting taxid.map
      --field-accession-re string   regular expression to extract assembly accession (default
                                    &quot;^\\w\\w_(.+)$&quot;)
      --force                       overwrite existed output directory
      --gtdb                        input files are GTDB taxonomy file
      --gtdb-re-subs string         regular expression to extract assembly accession as the subspecies
                                    (default &quot;^\\w\\w_GC[AF]_(.+)\\.\\d+$&quot;)
  -h, --help                        help for create-taxdump
      --line-chunk-size int         number of lines to process for each thread, and 4 threads is fast
                                    enough. (default 5000)
      --null strings                null value of taxa (default [,NULL,NA])
  -x, --old-taxdump-dir string      taxdump directory of the previous version, for generating merged.dmp
                                    and delnodes.dmp
  -O, --out-dir string              output directory
  -R, --rank-names strings          names of all ranks, leave it empty to use the first row of input as
                                    rank names

</code></pre>
<p>Examples:</p>
<ol>
<li>
<p>GTDB. See more: https://github.com/shenwei356/gtdb-taxdump</p>
<pre><code>$ taxonkit create-taxdump --gtdb ar53_taxonomy_r207.tsv.gz bac120_taxonomy_r207.tsv.gz --out-dir taxdump
16:42:35.213 [INFO] 317542 records saved to taxdump/taxid.map
16:42:35.460 [INFO] 401815 records saved to taxdump/nodes.dmp
16:42:35.611 [INFO] 401815 records saved to taxdump/names.dmp
16:42:35.611 [INFO] 0 records saved to taxdump/merged.dmp
16:42:35.611 [INFO] 0 records saved to taxdump/delnodes.dmp
</code></pre>
</li>
<li>
<p>ICTV, See more: https://github.com/shenwei356/ictv-taxdump</p>
</li>
<li>
<p><a href="https://www.nature.com/articles/s41564-021-00928-6">MGV</a>. Only Order, Family, Genus information are available.</p>
<pre><code>$ cat mgv_contig_info.tsv \
    | csvtk cut -t -f ictv_order,ictv_family,ictv_genus,votu_id,contig_id \
    | sed 1d \
    &gt; mgv.tsv

$ taxonkit create-taxdump mgv.tsv --out-dir mgv --force -A 5 -R order,family,genus,species,contig
16:45:40.555 [WARN] --field-accession-re failed to extract genome accession, the origninal value is used instead. e.g., MGV-GENOME-0231225
16:45:40.817 [INFO] 189680 records saved to mgv/taxid.map
16:45:40.846 [INFO] 54224 records saved to mgv/nodes.dmp
16:45:40.864 [INFO] 54224 records saved to mgv/names.dmp
16:45:40.864 [INFO] 0 records saved to mgv/merged.dmp
16:45:40.864 [INFO] 0 records saved to mgv/delnodes.dmp

$ head -n 5 mgv/taxid.map 
MGV-GENOME-0364295      3108345579
MGV-GENOME-0364296      2356405276
MGV-GENOME-0364303      1099424244
MGV-GENOME-0364311      4037644503
MGV-GENOME-0364312      452745976

$ echo 677052301 | taxonkit lineage --data-dir mgv/ 
677052301       Caudovirales;crAss-phage;OTU-61123

$ echo 677052301 | taxonkit reformat --data-dir mgv/ -I 1 -P
677052301       k__;p__;c__;o__Caudovirales;f__crAss-phage;g__;s__OTU-61123

$ grep MGV-GENOME-0364295 mgv.tsv 
Caudovirales    crAss-phage     NULL    OTU-61123       MGV-GENOME-0364295
</code></pre>
</li>
<li>
<p>Custom lineages with the first row as rank names and treating one column as accession.</p>
<pre><code>$ csvtk pretty -t example/taxonomy.tsv 
id                superkingdom   phylum           class                 order              family               genus            species
---------------   ------------   --------------   -------------------   ----------------   ------------------   --------------   --------------------------
GCF_001027105.1   Bacteria       Firmicutes       Bacilli               Bacillales         Staphylococcaceae    Staphylococcus   Staphylococcus aureus
GCF_001096185.1   Bacteria       Firmicutes       Bacilli               Lactobacillales    Streptococcaceae     Streptococcus    Streptococcus pneumoniae
GCF_001544255.1   Bacteria       Firmicutes       Bacilli               Lactobacillales    Enterococcaceae      Enterococcus     Enterococcus faecium
GCF_002949675.1   Bacteria       Proteobacteria   Gammaproteobacteria   Enterobacterales   Enterobacteriaceae   Shigella         Shigella dysenteriae
GCF_002950215.1   Bacteria       Proteobacteria   Gammaproteobacteria   Enterobacterales   Enterobacteriaceae   Shigella         Shigella flexneri
GCF_006742205.1   Bacteria       Firmicutes       Bacilli               Bacillales         Staphylococcaceae    Staphylococcus   Staphylococcus epidermidis
GCF_000006945.2   Bacteria       Proteobacteria   Gammaproteobacteria   Enterobacterales   Enterobacteriaceae   Salmonella       Salmonella enterica
GCF_000017205.1   Bacteria       Proteobacteria   Gammaproteobacteria   Pseudomonadales    Pseudomonadaceae     Pseudomonas      Pseudomonas aeruginosa
GCF_003697165.2   Bacteria       Proteobacteria   Gammaproteobacteria   Enterobacterales   Enterobacteriaceae   Escherichia      Escherichia coli
GCF_009759685.1   Bacteria       Proteobacteria   Gammaproteobacteria   Moraxellales       Moraxellaceae        Acinetobacter    Acinetobacter baumannii
GCF_000148585.2   Bacteria       Firmicutes       Bacilli               Lactobacillales    Streptococcaceae     Streptococcus    Streptococcus mitis
GCF_000392875.1   Bacteria       Firmicutes       Bacilli               Lactobacillales    Enterococcaceae      Enterococcus     Enterococcus faecalis
GCF_000742135.1   Bacteria       Proteobacteria   Gammaproteobacteria   Enterobacterales   Enterobacteriaceae   Klebsiella       Klebsiella pneumonia

# the first column as accession
$ taxonkit create-taxdump -A 1 example/taxonomy.tsv -O example/taxdump
16:31:31.828 [INFO] I will use the first row of input as rank names
16:31:31.841 [WARN] --field-accession-re failed to extract genome accession, the origninal value is used instead. e.g., GCF_000742135.1
16:31:31.843 [INFO] 13 records saved to example/taxdump/taxid.map
16:31:31.843 [INFO] 39 records saved to example/taxdump/nodes.dmp
16:31:31.843 [INFO] 39 records saved to example/taxdump/names.dmp
16:31:31.843 [INFO] 0 records saved to example/taxdump/merged.dmp
16:31:31.843 [INFO] 0 records saved to example/taxdump/delnodes.dmp

$ export TAXONKIT_DB=example/taxdump
$ taxonkit list --ids 1 | taxonkit filter -E species | taxonkit lineage -r
1527235303      Bacteria;Firmicutes;Bacilli;Lactobacillales;Streptococcaceae;Streptococcus;Streptococcus mitis  species
2983929374      Bacteria;Firmicutes;Bacilli;Lactobacillales;Streptococcaceae;Streptococcus;Streptococcus pneumoniae     species
3809813362      Bacteria;Firmicutes;Bacilli;Lactobacillales;Enterococcaceae;Enterococcus;Enterococcus faecalis  species
4145431389      Bacteria;Firmicutes;Bacilli;Lactobacillales;Enterococcaceae;Enterococcus;Enterococcus faecium   species
1569132721      Bacteria;Firmicutes;Bacilli;Bacillales;Staphylococcaceae;Staphylococcus;Staphylococcus aureus   species
1920251658      Bacteria;Firmicutes;Bacilli;Bacillales;Staphylococcaceae;Staphylococcus;Staphylococcus epidermidis      species
3843752343      Bacteria;Proteobacteria;Gammaproteobacteria;Pseudomonadales;Pseudomonadaceae;Pseudomonas;Pseudomonas aeruginosa species
72054943        Bacteria;Proteobacteria;Gammaproteobacteria;Moraxellales;Moraxellaceae;Acinetobacter;Acinetobacter baumannii    species
1678121664      Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Salmonella;Salmonella enterica  species
524994882       Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Shigella;Shigella dysenteriae   species
2695851945      Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Shigella;Shigella flexneri      species
3958205156      Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Klebsiella;Klebsiella pneumoniae        species
4093283224      Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia;Escherichia coli    species

$ head -n 3 example/taxdump/taxid.map
GCF_001027105.1 1569132721
GCF_001096185.1 2983929374
GCF_001544255.1 4145431389
</code></pre>
</li>
<li>
<p>Custom lineages with the first row as rank names (pure lineage data)</p>
<pre><code>$ csvtk cut -t -f 2- example/taxonomy.tsv | head -n 2 | csvtk pretty -t 
superkingdom   phylum       class     order        family              genus            species
------------   ----------   -------   ----------   -----------------   --------------   ---------------------
Bacteria       Firmicutes   Bacilli   Bacillales   Staphylococcaceae   Staphylococcus   Staphylococcus aureus

$ csvtk cut -t -f 2- example/taxonomy.tsv \
    | taxonkit create-taxdump -O example/taxdump2
16:53:08.604 [INFO] I will use the first row of input as rank names
16:53:08.614 [INFO] 39 records saved to example/taxdump2/nodes.dmp
16:53:08.614 [INFO] 39 records saved to example/taxdump2/names.dmp
16:53:08.614 [INFO] 0 records saved to example/taxdump2/merged.dmp
16:53:08.615 [INFO] 0 records saved to example/taxdump2/delnodes.dmp

$ export TAXONKIT_DB=example/taxdump2
$ taxonkit list --ids 1 | taxonkit filter -E species | taxonkit lineage -r | head -n 2
1527235303      Bacteria;Firmicutes;Bacilli;Lactobacillales;Streptococcaceae;Streptococcus;Streptococcus mitis  species
2983929374      Bacteria;Firmicutes;Bacilli;Lactobacillales;Streptococcaceae;Streptococcus;Streptococcus pneumoniae     species
</code></pre>
</li>
</ol>
<h2 id="genautocomplete">genautocomplete</h2>
<p>Usage</p>
<pre><code class="language-text">Generate shell autocompletion script

Supported shell: bash|zsh|fish|powershell

Bash:

    # generate completion shell
    taxonkit genautocomplete --shell bash

    # configure if never did.
    # install bash-completion if the &quot;complete&quot; command is not found.
    echo &quot;for bcfile in ~/.bash_completion.d/* ; do source \$bcfile; done&quot; &gt;&gt; ~/.bash_completion
    echo &quot;source ~/.bash_completion&quot; &gt;&gt; ~/.bashrc

Zsh:

    # generate completion shell
    taxonkit genautocomplete --shell zsh --file ~/.zfunc/_taxonkit

    # configure if never did
    echo 'fpath=( ~/.zfunc &quot;${fpath[@]}&quot; )' &gt;&gt; ~/.zshrc
    echo &quot;autoload -U compinit; compinit&quot; &gt;&gt; ~/.zshrc

fish:

    taxonkit genautocomplete --shell fish --file ~/.config/fish/completions/taxonkit.fish

Usage:
  taxonkit genautocomplete [flags]

Flags:
      --file string   autocompletion file (default &quot;/home/shenwei/.bash_completion.d/taxonkit.sh&quot;)
  -h, --help          help for genautocomplete
      --type string   autocompletion type (currently only bash supported) (default &quot;bash&quot;)

</code></pre>
<h2 id="profile2cami">profile2cami</h2>
<p>Usage</p>
<pre><code class="language-text">Convert metagenomic profile table to CAMI format

Input format: 
  1. The input file should be tab-delimited
  2. At least two columns needed:
     a) TaxId of taxon at species or lower rank.
     b) Abundance (could be percentage, automatically detected or use -p/--percentage).

Attentions:
  1. Some TaxIds may be merged to another ones in current taxonomy version,
     the abundances will be summed up.
  2. Some TaxIds may be deleted in current taxonomy version,
     the abundances can be optionally recomputed with the flag -R/--recompute-abd.

Usage:
  taxonkit profile2cami [flags]

Flags:
  -a, --abundance-field int   field index of abundance. input data should be tab-separated (default 2)
  -h, --help                  help for profile2cami
  -0, --keep-zero             keep taxons with abundance of zero
  -p, --percentage            abundance is in percentage
  -R, --recompute-abd         recompute abundance if some TaxIds are deleted in current taxonomy version
  -s, --sample-id string      sample ID in result file
  -r, --show-rank strings     only show TaxIds and names of these ranks (default [superkingdom,phylum,class,order,family,genus,species,strain])
  -i, --taxid-field int       field index of taxid. input data should be tab-separated (default 1)
  -t, --taxonomy-id string    taxonomy ID in result file

</code></pre>
<p>Examples</p>
<ul>
<li>
<p>Test data, note that <code>2824115</code> is merged to <code>483329</code> and <code>1657696</code> is deleted in current taxonomy version.</p>
<pre><code>$ cat example/abundance.tsv 
2824115 0.2     merged to 483329
483329  0.2     absord 2824115
239935  0.5     no change
1657696 0.1     deleted
</code></pre>
</li>
<li>
<p>Example:</p>
<pre><code>$ taxonkit profile2cami -s sample1 -t 2021-10-01 \
    example/abundance.tsv

13:17:40.552 [WARN] taxid is deleted in current taxonomy version: 1657696
13:17:40.552 [WARN] you may recomputed abundance with the flag -R/--recompute-abd
@SampleID:sample1
@Version:0.10.0
@Ranks:superkingdom|phylum|class|order|family|genus|species|strain
@TaxonomyID:2021-10-01
@@TAXID RANK    TAXPATH TAXPATHSN       PERCENTAGE
2       superkingdom    2       Bacteria        50.000000000000000
2759    superkingdom    2759    Eukaryota       40.000000000000000
74201   phylum  2|74201 Bacteria|Verrucomicrobia        50.000000000000000
6656    phylum  2759|6656       Eukaryota|Arthropoda    40.000000000000000
203494  class   2|74201|203494  Bacteria|Verrucomicrobia|Verrucomicrobiae       50.000000000000000
50557   class   2759|6656|50557 Eukaryota|Arthropoda|Insecta    40.000000000000000
48461   order   2|74201|203494|48461    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales    50.000000000000000
7041    order   2759|6656|50557|7041    Eukaryota|Arthropoda|Insecta|Coleoptera 40.000000000000000
1647988 family  2|74201|203494|48461|1647988    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae    50.000000000000000
57514   family  2759|6656|50557|7041|57514      Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae       40.000000000000000
239934  genus   2|74201|203494|48461|1647988|239934     Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia 50.000000000000000
57515   genus   2759|6656|50557|7041|57514|57515        Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae|Nicrophorus   40.000000000000000
239935  species 2|74201|203494|48461|1647988|239934|239935      Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia|Akkermansia muciniphila 50.000000000000000
483329  species 2759|6656|50557|7041|57514|57515|483329 Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae|Nicrophorus|Nicrophorus carolina       40.000000000000000
</code></pre>
</li>
<li>
<p>Recompute (normalize) the abundance</p>
<pre><code>$ taxonkit profile2cami -s sample1 -t 2021-10-01 \
    example/abundance.tsv --recompute-abd
13:19:23.647 [WARN] taxid is deleted in current taxonomy version: 1657696
@SampleID:sample1
@Version:0.10.0
@Ranks:superkingdom|phylum|class|order|family|genus|species|strain
@TaxonomyID:2021-10-01
@@TAXID RANK    TAXPATH TAXPATHSN       PERCENTAGE
2       superkingdom    2       Bacteria        55.555555555555557
2759    superkingdom    2759    Eukaryota       44.444444444444450
74201   phylum  2|74201 Bacteria|Verrucomicrobia        55.555555555555557
6656    phylum  2759|6656       Eukaryota|Arthropoda    44.444444444444450
203494  class   2|74201|203494  Bacteria|Verrucomicrobia|Verrucomicrobiae       55.555555555555557
50557   class   2759|6656|50557 Eukaryota|Arthropoda|Insecta    44.444444444444450
48461   order   2|74201|203494|48461    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales    55.555555555555557
7041    order   2759|6656|50557|7041    Eukaryota|Arthropoda|Insecta|Coleoptera 44.444444444444450
1647988 family  2|74201|203494|48461|1647988    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae    55.555555555555557
57514   family  2759|6656|50557|7041|57514      Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae       44.444444444444450
239934  genus   2|74201|203494|48461|1647988|239934     Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia 55.555555555555557
57515   genus   2759|6656|50557|7041|57514|57515        Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae|Nicrophorus   44.444444444444450
239935  species 2|74201|203494|48461|1647988|239934|239935      Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia|Akkermansia muciniphila 55.555555555555557
483329  species 2759|6656|50557|7041|57514|57515|483329 Eukaryota|Arthropoda|Insecta|Coleoptera|Silphidae|Nicrophorus|Nicrophorus carolina       44.444444444444450
</code></pre>
</li>
<li>
<p>See https://github.com/shenwei356/sun2021-cami-profiles</p>
</li>
</ul>
<h2 id="cami-filter">cami-filter</h2>
<p>Usage</p>
<pre><code class="language-text">Remove taxa of given TaxIds and their descendants in CAMI metagenomic profile

Input format: 
  The CAMI (Taxonomic) Profiling Output Format    
  - https://github.com/CAMI-challenge/contest_information/blob/master/file_formats/CAMI_TP_specification.mkd
  - One file with mutiple samples is also supported.

How to:
  - No extra taxonomy data needed, so the original taxonomic information are
    used and not changed.
  - A mini taxonomic tree is built from records with abundance greater than
    zero, and only leaves are retained for later use. The rank of leaves may
    be &quot;strain&quot;, &quot;species&quot;, or &quot;no rank&quot;.
  - Relative abundances (in percentage) are recomputed for all leaves
    (reference genome).
  - A new taxonomic tree is built from these leaves, and abundances are 
    cumulatively added up from leaves to the root.

Examples:
  1. Remove Archaea, Bacteria, and EukaryoteS, only keep Viruses:
      taxonkit cami-filter -t 2,2157,2759 test.profile -o test.filter.profile
  2. Remove Viruses:
      taxonkit cami-filter -t 10239 test.profile -o test.filter.profile

Usage:
  taxonkit cami-filter [flags]

Flags:
      --field-percentage int   field index of PERCENTAGE (default 5)
      --field-rank int         field index of taxid (default 2)
      --field-taxid int        field index of taxid (default 1)
      --field-taxpath int      field index of TAXPATH (default 3)
      --field-taxpathsn int    field index of TAXPATHSN (default 4)
  -h, --help                   help for cami-filter
      --leaf-ranks strings     only consider leaves at these ranks (default [species,strain,no rank])
      --show-rank strings      only show TaxIds and names of these ranks (default [superkingdom,phylum,class,order,family,genus,species,strain])
      --taxid-sep string       separator of taxid in TAXPATH and TAXPATHSN (default &quot;|&quot;)
  -t, --taxids strings         the parent taxid(s) to filter out
  -f, --taxids-file strings    file(s) for the parent taxid(s) to filter out, one taxid per line

</code></pre>
<p>Examples:</p>
<ol>
<li>Remove Eukaryota<pre><code>taxonkit profile2cami -s sample1 -t 2021-10-01 \
    example/abundance.tsv --recompute-abd \
    | taxonkit cami-filter -t 2759
@SampleID:sample1
@Version:0.10.0
@Ranks:superkingdom|phylum|class|order|family|genus|species|strain
@TaxonomyID:2021-10-01
@@TAXID RANK    TAXPATH TAXPATHSN       PERCENTAGE
2       superkingdom    2       Bacteria        100.000000000000000
74201   phylum  2|74201 Bacteria|Verrucomicrobia        100.000000000000000
203494  class   2|74201|203494  Bacteria|Verrucomicrobia|Verrucomicrobiae       100.000000000000000
48461   order   2|74201|203494|48461    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales    100.000000000000000
1647988 family  2|74201|203494|48461|1647988    Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae    100.000000000000000
239934  genus   2|74201|203494|48461|1647988|239934     Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia       100.000000000000000
239935  species 2|74201|203494|48461|1647988|239934|239935      Bacteria|Verrucomicrobia|Verrucomicrobiae|Verrucomicrobiales|Akkermansiaceae|Akkermansia|Akkermansia muciniphila       100.000000000000000
</code></pre>
</li>
</ol>
<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//taxonkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../download/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Download" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Download
            </div>
          </div>
        </a>
      
      
        
        <a href="../tutorial/" class="md-footer__link md-footer__link--next" aria-label="Next: Tutorial" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Tutorial
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.b6ff8c07.min.js"}</script>
    
    
      <script src="../assets/javascripts/bundle.5e13d264.min.js"></script>
      
    
  </body>
</html>