
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="TaxonKit - NCBI Taxonomy Toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/taxonkit/download/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../usage/">
      
      
      <link rel="icon" href="../files/favicon.ico">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.8">
    
    
      
        <title>Download - TaxonKit - NCBI Taxonomy Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.356b1318.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="cyan" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-header__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TaxonKit - NCBI Taxonomy Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Download
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TaxonKit - NCBI Taxonomy Toolkit" class="md-nav__button md-logo" aria-label="TaxonKit - NCBI Taxonomy Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    TaxonKit - NCBI Taxonomy Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/taxonkit" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Download
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#current-version" class="md-nav__link">
    Current Version
  </a>
  
    <nav class="md-nav" aria-label="Current Version">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#please-cite" class="md-nav__link">
    Please cite
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#links" class="md-nav__link">
    Links
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stable" class="md-nav__link">
    Method 1: Download binaries (latest stable)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-may-not-the-lastest-version" class="md-nav__link">
    Method 3: Install via homebrew (may not the lastest version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-compile-from-source-latest-stabledev-version" class="md-nav__link">
    Method 4: Compile from source (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#release-history" class="md-nav__link">
    Release history
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorial
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../bench/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Benchmark
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../chinese/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    中文介绍
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../chinese-dev/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    开发笔记
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356/taxid-changelog" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Taxid-Changelog
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    More tools
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#current-version" class="md-nav__link">
    Current Version
  </a>
  
    <nav class="md-nav" aria-label="Current Version">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#please-cite" class="md-nav__link">
    Please cite
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#links" class="md-nav__link">
    Links
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stable" class="md-nav__link">
    Method 1: Download binaries (latest stable)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-may-not-the-lastest-version" class="md-nav__link">
    Method 3: Install via homebrew (may not the lastest version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-compile-from-source-latest-stabledev-version" class="md-nav__link">
    Method 4: Compile from source (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#release-history" class="md-nav__link">
    Release history
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="download">Download<a class="headerlink" href="#download" title="Permanent link">Link</a></h1>
<p><code>TaxonKit</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/taxonkit/releases">release</a> page.</p>
<h2 id="current-version">Current Version<a class="headerlink" href="#current-version" title="Permanent link">Link</a></h2>
<ul>
<li>
<p><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.20.0">TaxonKit v0.20.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.20.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.20.0/total.svg" /></a></p>
<p>This version is mainly for maintaining compatibility with NCBI's recent changes(<a href="https://ncbiinsights.ncbi.nlm.nih.gov/2025/02/27/new-ranks-ncbi-taxonomy/">1</a>, <a href="https://ncbiinsights.ncbi.nlm.nih.gov/2025/04/25/ncbi-taxonomy-updates-virus-classification-april-2025/">2</a>).  <br />
<strong>Please remove the <code>ranks.txt</code> file in <code>~/.taxonkit/</code> or other directories containing taxdump files (nodes.dmp)</strong>.</p>
<ul>
<li><code>taxonkit reformat</code>:<ul>
<li>Tried to make it compatible with NCBI's recent changes but failed.</li>
<li><strong>Update doc to recommend using <code>taxonkit reformat2</code></strong>.</li>
</ul>
</li>
<li><code>taxonkit reformat</code>:<ul>
<li>Fix handling of deleted TaxIds. <a href="https://github.com/shenwei356/taxonkit/issues/117">#117</a></li>
<li><strong>Change the default output format to be compatible with NCBI's recent changes</strong>. <a href="https://github.com/shenwei356/taxonkit/issues/115">#115</a>, <a href="https://github.com/shenwei356/taxonkit/issues/116">#116</a>)</li>
</ul>
</li>
<li><code>taxonkit filter2</code>:<ul>
<li>Support new ranks: <code>acellular root</code> and <code>cellular root</code>. 
  <strong>Please remove the <code>ranks.txt</code> file in <code>~/.taxonkit/</code> or other directories containing taxdump files (nodes.dmp)</strong>.</li>
<li>Automaticlly switch on <code>-N/--discard-noranks</code> when only <code>-E/--equal-to</code> is given and the value is not one of ranks without order ("no rank", "clade").</li>
</ul>
</li>
<li><code>taxonkit name2taxid</code>:<ul>
<li>Remove duplicated records caused by capitalization.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h3 id="please-cite">Please cite<a class="headerlink" href="#please-cite" title="Permanent link">Link</a></h3>
<blockquote>
<p>Shen, W., Ren, H., TaxonKit: a practical and efficient NCBI Taxonomy toolkit,
Journal of Genetics and Genomics, <a href="https://www.sciencedirect.com/science/article/pii/S1673852721000837">https://doi.org/10.1016/j.jgg.2021.03.006</a></p>
</blockquote>
<h3 id="links">Links<a class="headerlink" href="#links" title="Permanent link">Link</a></h3>
<p><strong><em>Tips</em></strong></p>
<ul>
<li>run <code>taxonkit version</code> to check update !!!</li>
<li>run <code>taxonkit genautocomplete</code> to update Bash completion !!!</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align: left;">OS</th>
<th style="text-align: left;">Arch</th>
<th style="text-align: left;">File, 中国镜像</th>
<th style="text-align: left;">Download Count</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Linux</td>
<td style="text-align: left;"><strong>64-bit</strong></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_linux_amd64.tar.gz"><strong>taxonkit_linux_amd64.tar.gz</strong></a>,<br/> <a href="http://app.shenwei.me/data/taxonkit/taxonkit_linux_amd64.tar.gz">中国镜像</a></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_linux_amd64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/taxonkit_linux_amd64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td style="text-align: left;">Linux</td>
<td style="text-align: left;"><strong>arm64</strong></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_linux_arm64.tar.gz"><strong>taxonkit_linux_arm64.tar.gz</strong></a>,<br/> <a href="http://app.shenwei.me/data/taxonkit/taxonkit_linux_arm64.tar.gz">中国镜像</a></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_linux_arm64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/taxonkit_linux_arm64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td style="text-align: left;">macOS</td>
<td style="text-align: left;"><strong>64-bit</strong></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_darwin_amd64.tar.gz"><strong>taxonkit_darwin_amd64.tar.gz</strong></a>,<br/> <a href="http://app.shenwei.me/data/taxonkit/taxonkit_darwin_amd64.tar.gz">中国镜像</a></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_darwin_amd64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/taxonkit_darwin_amd64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td style="text-align: left;">macOS</td>
<td style="text-align: left;"><strong>arm64</strong></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_darwin_arm64.tar.gz"><strong>taxonkit_darwin_arm64.tar.gz</strong></a>,<br/> <a href="http://app.shenwei.me/data/taxonkit/taxonkit_darwin_arm64.tar.gz">中国镜像</a></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_darwin_arm64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/taxonkit_darwin_arm64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td style="text-align: left;">Windows</td>
<td style="text-align: left;"><strong>64-bit</strong></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_windows_amd64.exe.tar.gz"><strong>taxonkit_windows_amd64.exe.tar.gz</strong></a>,<br/> <a href="http://app.shenwei.me/data/taxonkit/taxonkit_windows_amd64.exe.tar.gz">中国镜像</a></td>
<td style="text-align: left;"><a href="https://github.com/shenwei356/taxonkit/releases/download/v0.20.0/taxonkit_windows_amd64.exe.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/latest/taxonkit_windows_amd64.exe.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
</tbody>
</table>
<h2 id="installation">Installation<a class="headerlink" href="#installation" title="Permanent link">Link</a></h2>
<p><a href="https://github.com/shenwei356/taxonkit/releases">Download Page</a></p>
<p><code>TaxonKit</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/taxonkit/releases">release</a> page.</p>
<h4 id="method-1-download-binaries-latest-stable">Method 1: Download binaries (latest stable)<a class="headerlink" href="#method-1-download-binaries-latest-stable" title="Permanent link">Link</a></h4>
<p>Just <a href="https://github.com/shenwei356/taxonkit/releases">download</a> compressed
executable file of your operating system,
and uncompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<div class="codehilite"><pre><span></span><code>sudo cp taxonkit /usr/local/bin/
</code></pre></div>

</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<div class="codehilite"><pre><span></span><code>mkdir -p $HOME/bin/; cp taxonkit $HOME/bin/
</code></pre></div>

</li>
</ol>
</li>
<li>
<p><strong>For windows</strong>, just copy <code>taxonkit.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-conda-latest-stable-version">Method 2: Install via conda  (latest stable version) <a href="http://bioinf.shenwei.me/taxonkit/download/"><img alt="Install-with-conda" src="https://anaconda.org/bioconda/taxonkit/badges/installer/conda.svg" /></a> <a href="https://anaconda.org/bioconda/taxonkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/taxonkit/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/taxonkit"><img alt="downloads" src="https://anaconda.org/bioconda/taxonkit/badges/downloads.svg" /></a><a class="headerlink" href="#method-2-install-via-conda-latest-stable-version" title="Permanent link">Link</a></h4>
<div class="codehilite"><pre><span></span><code>conda install -c bioconda taxonkit
</code></pre></div>

<h4 id="method-3-install-via-homebrew-may-not-the-lastest-version">Method 3: Install via homebrew (may not the lastest version)<a class="headerlink" href="#method-3-install-via-homebrew-may-not-the-lastest-version" title="Permanent link">Link</a></h4>
<div class="codehilite"><pre><span></span><code>brew install brewsci/bio/taxonkit
</code></pre></div>

<h4 id="method-4-compile-from-source-latest-stabledev-version">Method 4: Compile from source (latest stable/dev version)<a class="headerlink" href="#method-4-compile-from-source-latest-stabledev-version" title="Permanent link">Link</a></h4>
<ol>
<li>
<p><a href="https://go.dev/doc/install">Install go</a></p>
<div class="codehilite"><pre><span></span><code><span class="n">wget</span><span class="w"> </span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">go</span><span class="o">.</span><span class="n">dev</span><span class="o">/</span><span class="n">dl</span><span class="o">/</span><span class="n">go1</span><span class="o">.</span><span class="mf">24.2</span><span class="o">.</span><span class="n">linux</span><span class="o">-</span><span class="n">amd64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>

<span class="n">tar</span><span class="w"> </span><span class="o">-</span><span class="n">zxf</span><span class="w"> </span><span class="n">go1</span><span class="o">.</span><span class="mf">24.2</span><span class="o">.</span><span class="n">linux</span><span class="o">-</span><span class="n">amd64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span><span class="w"> </span><span class="o">-</span><span class="n">C</span><span class="w"> </span><span class="o">$</span><span class="n">HOME</span><span class="o">/</span>

<span class="c1"># or </span>
<span class="c1">#   echo &quot;export PATH=$PATH:$HOME/go/bin&quot; &gt;&gt; ~/.bashrc</span>
<span class="c1">#   source ~/.bashrc</span>
<span class="k">export</span><span class="w"> </span><span class="n">PATH</span><span class="o">=$</span><span class="n">PATH</span><span class="p">:</span><span class="o">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">go</span><span class="o">/</span><span class="n">bin</span>
</code></pre></div>

</li>
<li>
<p>Compile TaxonKit</p>
<div class="codehilite"><pre><span></span><code><span class="n">git</span><span class="w"> </span><span class="n">clone</span><span class="w"> </span><span class="nl">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="p">.</span><span class="n">com</span><span class="o">/</span><span class="n">shenwei356</span><span class="o">/</span><span class="n">taxonkit</span>
<span class="n">cd</span><span class="w"> </span><span class="n">taxonkit</span><span class="o">/</span><span class="n">taxonkit</span><span class="o">/</span>
<span class="k">go</span><span class="w"> </span><span class="n">build</span>

<span class="err">#</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">executable</span><span class="w"> </span><span class="nc">binary</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">located</span><span class="w"> </span><span class="ow">in</span><span class="err">:</span>
<span class="err">#</span><span class="w">   </span><span class="p">.</span><span class="o">/</span><span class="n">taxonkit</span>
<span class="err">#</span><span class="w"> </span><span class="n">You</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="n">also</span><span class="w"> </span><span class="n">move</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">anywhere</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="err">$</span><span class="k">PATH</span>
<span class="n">mkdir</span><span class="w"> </span><span class="o">-</span><span class="n">p</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span>
<span class="n">cp</span><span class="w"> </span><span class="p">.</span><span class="o">/</span><span class="n">taxonkit</span><span class="w"> </span><span class="err">$</span><span class="n">HOME</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span>
</code></pre></div>

</li>
</ol>
<h2 id="bash-completion">Bash-completion<a class="headerlink" href="#bash-completion" title="Permanent link">Link</a></h2>
<p>Supported shell: bash|zsh|fish|powershell</p>
<p>Bash:</p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> generate completion shell
taxonkit genautocomplete --shell bash

<span class="gh">#</span> configure if never did.
<span class="gh">#</span> install bash-completion if the &quot;complete&quot; command is not found.
echo &quot;for bcfile in ~/.bash_completion.d/* ; do source \$bcfile; done&quot; &gt;&gt; ~/.bash_completion
echo &quot;source ~/.bash_completion&quot; &gt;&gt; ~/.bashrc
</code></pre></div>

<p>Zsh:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># generate completion shell</span>
<span class="n">taxonkit</span><span class="w"> </span><span class="n">genautocomplete</span><span class="w"> </span><span class="o">--</span><span class="n">shell</span><span class="w"> </span><span class="n">zsh</span><span class="w"> </span><span class="o">--</span><span class="n">file</span><span class="w"> </span><span class="o">~/.</span><span class="n">zfunc</span><span class="o">/</span><span class="n">_taxonkit</span>

<span class="c1"># configure if never did</span>
<span class="n">echo</span><span class="w"> </span><span class="s1">&#39;fpath=( ~/.zfunc &quot;${fpath[@]}&quot; )&#39;</span><span class="w"> </span><span class="o">&gt;&gt;</span><span class="w"> </span><span class="o">~/.</span><span class="n">zshrc</span>
<span class="n">echo</span><span class="w"> </span><span class="s2">&quot;autoload -U compinit; compinit&quot;</span><span class="w"> </span><span class="o">&gt;&gt;</span><span class="w"> </span><span class="o">~/.</span><span class="n">zshrc</span>
</code></pre></div>

<p>fish:</p>
<div class="codehilite"><pre><span></span><code>taxonkit genautocomplete --shell fish --file ~/.config/fish/completions/taxonkit.fish
</code></pre></div>

<h2 id="dataset">Dataset<a class="headerlink" href="#dataset" title="Permanent link">Link</a></h2>
<ol>
<li>Download and uncompress <code>taxdump.tar.gz</code>: ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz </li>
<li>Copy <code>names.dmp</code>, <code>nodes.dmp</code>, <code>delnodes.dmp</code> and <code>merged.dmp</code> to data directory: <code>$HOME/.taxonkit</code>,
e.g., <code>/home/shenwei/.taxonkit</code> ,</li>
<li>Optionally copy to some other directories, and later you can refer to using flag <code>--data-dir</code>,
or environment variable <code>TAXONKIT_DB</code>.</li>
</ol>
<p>All-in-one command:</p>
<div class="codehilite"><pre><span></span><code>wget -c ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz 
tar -zxvf taxdump.tar.gz

mkdir -p $HOME/.taxonkit
cp names.dmp nodes.dmp delnodes.dmp merged.dmp $HOME/.taxonkit
</code></pre></div>

<p><strong>Update dataset</strong>: Simply re-download the taxdump files, uncompress and override old ones.</p>
<h2 id="release-history">Release history<a class="headerlink" href="#release-history" title="Permanent link">Link</a></h2>
<ul>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.19.0">TaxonKit v0.19.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.19.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.19.0/total.svg" /></a><ul>
<li>new command <code>taxonkit reformat2</code>: Reformat lineage in chosen ranks, allowing more ranks than 'reformat'</li>
<li><code>taxonkit reformat</code>:<ul>
<li>Fix <code>-T/--trim</code> which did not work for <code>-r/--miss-rank-repl</code>. <a href="https://github.com/shenwei356/taxonkit/issues/106">#106</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.18.0">TaxonKit v0.18.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.18.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.18.0/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li>Add a placeholder for rank "realm", "{r}", which is common in Virus taxonomy like <a href="https://github.com/shenwei356/ictv-taxdump">ictv</a>. <a href="https://github.com/shenwei356/taxonkit/issues/102">#102</a></li>
</ul>
</li>
<li><code>taxonkit name2taxid</code>:<ul>
<li>Show warning for names with multiple taxids. <a href="https://github.com/shenwei356/taxonkit/issues/103">#103</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.17.0">TaxonKit v0.17.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.17.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.17.0/total.svg" /></a><ul>
<li><code>taxonkit filter</code>:<ul>
<li>Fix keeping taxids with a rank of "no rank" or "clade". <a href="https://github.com/shenwei356/taxonkit/issues/97">#97</a></li>
</ul>
</li>
<li><code>taxonkit list</code>:<ul>
<li>Accept input TaxIds from stdin/files, not just the flag <code>--ids</code>. <a href="https://github.com/shenwei356/taxonkit/issues/93">#93</a></li>
</ul>
</li>
<li><code>taxonkit name2taxid</code>:<ul>
<li>Add support of fuzzy match with <code>-f/--fuzzy</code>. <a href="https://github.com/shenwei356/taxonkit/issues/88">#88</a></li>
</ul>
</li>
<li><code>taxonkit profile2cami</code>:<ul>
<li>Add a new flag <code>-S/--no-sum-up</code> to disable summing up abundances. <a href="https://github.com/shenwei356/taxonkit/issues/99">#99</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.16.0">TaxonKit v0.16.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.16.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.16.0/total.svg" /></a><ul>
<li><code>taxonkit create-taxdump</code>:<ul>
<li>TaxIds are generated from the hash value of "rank+taxon_name", this allowed duplicated names with different ranks (common in GTDB). <a href="https://github.com/shenwei356/taxonkit/issues/92">#92</a>.</li>
</ul>
</li>
<li><code>taxonkit taxid-changelog/create-taxdump</code>:<ul>
<li>Add notes about the possible error records in taxid-changelog built from taxonomic data created by <code>create-taxdump</code>. <a href="https://github.com/shenwei356/taxonkit/issues/91">#91</a></li>
<li>Dont't worry, a single version of taxonomic data created by <code>taxonkit create-taxdump</code> has no problem, it's just the changelog might not be perfect.</li>
</ul>
</li>
<li><code>taxonkit lca</code>:<ul>
<li>new flag <code>-K/--keep-invalid</code>: print the query even if no single valid taxid left. <a href="https://github.com/shenwei356/taxonkit/issues/89">#89</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.15.1">TaxonKit v0.15.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.15.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.15.1/total.svg" /></a><ul>
<li><code>taxonkit name2taxid</code>:<ul>
<li>remove the restriction of name types. <a href="https://github.com/shenwei356/taxonkit/issues/87">#87</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.15.0">TaxonKit v0.15.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.15.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.15.0/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li>For lineages with more than one node, if it fails to query TaxId with the parent-child pair, use the last child only. <a href="https://github.com/shenwei356/taxonkit/issues/82">#82</a></li>
<li>The flag <code>-T/--trim</code> also does not add the prefix for missing ranks lower than the current rank. <a href="https://github.com/shenwei356/taxonkit/issues/82">#82</a></li>
<li>New flag <code>-s/--miss-rank-repl-suffix</code> to set the suffix for estimated taxon names. <a href="https://github.com/shenwei356/taxonkit/issues/85">#85</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.2">TaxonKit v0.14.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.14.2/total.svg" /></a><ul>
<li><code>taxonkit filter</code>:<ul>
<li>fix checking merged/deleted/not-found taxids. <a href="https://github.com/shenwei356/taxonkit/issues/80">#80</a></li>
</ul>
</li>
<li><code>taxonkit lca</code>:<ul>
<li>add a new flag <code>-b/--buffer-size</code> to set the size of the line buffer. <a href="https://github.com/shenwei356/taxonkit/issues/75">#75</a></li>
<li>fix typos: <code>--separater</code> -&gt; <code>--separater</code>, the former is still available for backward compatibility.</li>
</ul>
</li>
<li><code>taxonkit reformat</code>:<ul>
<li>output compatible format for TaxIds not found in the database. <a href="https://github.com/shenwei356/taxonkit/issues/79">#79</a></li>
</ul>
</li>
<li><code>taxonkit taxid-changelog</code>:<ul>
<li>support gzip-compressed taxdump files for saving space. <a href="https://github.com/shenwei356/taxonkit/issues/78">#78</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.1">TaxonKit v0.14.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.14.1/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li>The flag <code>-S/--pseudo-strain</code> does not require <code>-F/--fill-miss-rank</code> now.</li>
<li>For taxa of rank &gt;= species, <code>{t}</code>, <code>{S}</code>, and <code>T</code> outputs nothing when using <code>-S/--pseudo-strain</code>.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.0">TaxonKit v0.14.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.14.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.14.0/total.svg" /></a><ul>
<li><code>taxonkit create-taxdump</code>:<ul>
<li>save taxIds in <code>int32</code> instead of <code>uint32</code>, as BLAST and DIAMOND do. <a href="https://github.com/shenwei356/taxonkit/issues/70">#70</a></li>
</ul>
</li>
<li><code>taxonkit list</code>:<ul>
<li>do not skip visited subtrees when some of give taxids are descendants of others. <a href="https://github.com/shenwei356/taxonkit/issues/68">#68</a></li>
</ul>
</li>
<li><code>taxonkit</code>:<ul>
<li>when environment variable <code>TAXONKIT_DB</code> is set, explicitly setting <code>--data-dir</code> will override the value of <code>TAXONKIT_DB</code>.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.13.0">TaxonKit v0.13.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.13.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.13.0/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li>add a new placeholder <code>{K}</code> for rank <code>kingdom</code>. <a href="https://github.com/shenwei356/taxonkit/issues/64">#64</a></li>
<li>do not panic for invalid TaxIds, e.g., the column name, when using <code>-I--taxid-field</code>.</li>
</ul>
</li>
<li><code>taxonkit create-taxdump</code>: <ul>
<li>fix merged.dmp and delnodes.dmp. Thanks to @apcamargo ! <a href="https://github.com/shenwei356/gtdb-taxdump/issues/2">gtdb-taxdump/issues/2</a>.</li>
<li>fix bug of handling non-GTDB data when using <code>-A/--field-accession</code> and no rank names given:
  the colname of the accession column would be treated as one of the ranks, which messed up all the ranks.</li>
<li>fix the default option value of <code>--field-accession-re</code> which wrongly remove prefix like <code>Sp_</code>. <a href="https://github.com/shenwei356/taxonkit/issues/65">#65</a></li>
</ul>
</li>
<li><code>taxonkit list</code>:<ul>
<li>fix warning message of merged taxids.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.12.0">TaxonKit v0.12.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.12.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.12.0/total.svg" /></a><ul>
<li><code>taxonkit create-taxdump</code>: <ul>
<li>accepts arbitrary ranks <a href="https://github.com/shenwei356/taxonkit/issues/60">#60</a></li>
<li>better handle of taxa with same names.</li>
<li>many flags changed.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.11.1">TaxonKit v0.11.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.11.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.11.1/total.svg" /></a><ul>
<li><code>taxonkit create-taxdump</code>: fix bug of missing Class rank, contributed by @apcamargo. The flag <code>--gtdb</code> was not effected. <a href="https://github.com/shenwei356/taxonkit/issues/57">#57</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.11.0">TaxonKit v0.11.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.11.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.11.0/total.svg" /></a><ul>
<li>new command <code>taxonkit create-taxdump</code>: Create NCBI-style taxdump files for custom taxonomy, e.g., GTDB and ICTV. <a href="https://github.com/shenwei356/taxonkit/issues/56">#56</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.10.1">TaxonKit v0.10.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.10.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.10.1/total.svg" /></a><ul>
<li><code>taxonkit cami2-filter</code>: fix option <code>--show-rank</code> which did not work in v0.10.0.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.10.0">TaxonKit v0.10.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.10.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.10.0/total.svg" /></a><ul>
<li>new command <code>taxonkit cami2-filter</code>: Remove taxa of given TaxIds and their descendants in CAMI metagenomic profile</li>
<li><code>taxonkit reformat</code>: fix panic for deleted taxid using <code>-F/--fill-miss-rank</code>. <a href="https://github.com/shenwei356/taxonkit/issues/55">#55</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.9.0">TaxonKit v0.9.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.9.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.9.0/total.svg" /></a><ul>
<li>new command <code>taxonkit profile2cami</code>: converting metagenomic profile table to CAMI format</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.8.0">TaxonKit v0.8.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.8.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.8.0/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li><strong>accept input of TaxIds via flag <code>-I/--taxid-field</code></strong>.</li>
<li><strong>accept single taxonomy names</strong>.</li>
<li><em>show warning message for TaxIds with the same lineage</em>. <a href="https://github.com/shenwei356/taxonkit/issues/42">#42</a></li>
<li>better flag checking. <a href="https://github.com/shenwei356/taxonkit/issues/40">#40</a></li>
</ul>
</li>
<li><code>taxonkit lca</code>:<ul>
<li>slightly speedup.</li>
</ul>
</li>
<li><code>taxonkit genautocomplete</code>:<ul>
<li>support bash|zsh|fish/powershell</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.2">TaxonKit v0.7.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.7.2/total.svg" /></a><ul>
<li><code>taxonkit lineage</code>:<ul>
<li>new flag <code>-R/--show-lineage-ranks</code> for appending ranks of all levels.</li>
<li>reduce memory occupation and slightly speedup.</li>
</ul>
</li>
<li><code>taxonkit filter</code>:<ul>
<li>flag <code>-E/--equal-to</code> supports multiple values.</li>
<li>new flag <code>-n/--save-predictable-norank</code>: do not discard some special ranks without order when using -L, where rank of the closest higher node is still lower than rank cutoff.</li>
</ul>
</li>
<li><code>taxonkit reformat</code>:<ul>
<li><strong>new placeholder <code>{t}</code> for <code>subspecies/strain</code>, <code>{T}</code> for <code>strain</code></strong>. Thanks @wqssf102 for feedback.</li>
<li>new flag <code>-S/--pseudo-strain</code> for using the node with lowest rank as strain name, only if which rank is lower than "species" and not "subpecies" nor "strain". </li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.1">TaxonKit v0.7.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.7.1/total.svg" /></a><ul>
<li><code>taxonkit filter</code>: <ul>
<li>disable unnecessary stdin check when using flag <code>--list-order</code> or <code>--list-ranks</code>. <a href="https://github.com/shenwei356/taxonkit/issues/36">#36</a></li>
<li>better handling of black list, empty default value: "no rank" and "clade". And you need use <code>-N/--discard-noranks</code> to explicitly filter out "no rank", "clade". <a href="https://github.com/shenwei356/taxonkit/issues/37">#37</a></li>
<li>update help message. Thanks @standage for improve this command! <a href="https://github.com/shenwei356/taxonkit/issues/38">#38</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.0">TaxonKit v0.7.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.7.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.7.0/total.svg" /></a><ul>
<li><code>taxonkit</code>: <strong>2-3X faster taxonomy data loading</strong>.</li>
<li>new command <code>taxonkit filter</code>: <strong>filtering TaxIds by taxonomic rank range</strong>. <a href="https://github.com/shenwei356/taxonkit/issues/32">#32</a></li>
<li>new command <code>taxonkit lca</code>: Computing lowest common ancestor (LCA) for TaxIds.</li>
<li><code>taxonkit reformat</code>:<ul>
<li>new flag <code>-P/--add-prefix</code>: <strong>add prefixes for all ranks</strong>, single prefix for a rank is defined by flag <code>--prefix-X</code>, where <code>X</code> may be <code>k</code>, <code>p</code>, <code>c</code>, <code>o</code>, <code>f</code>, <code>s</code>, <code>S</code>.</li>
<li>new flag <code>-T/--trim</code>: do not fill missing rank lower than current rank.</li>
</ul>
</li>
<li><code>taxonkit list</code>: do not duplicate root node.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.2">TaxonKit v0.6.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.6.2/total.svg" /></a><ul>
<li><code>taxonkit reformat -F</code>: fix taxids of abbreviated lineage containing names shared by different taxids. <a href="https://github.com/shenwei356/taxonkit/issues/35">#35</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.1">TaxonKit v0.6.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.6.1/total.svg" /></a><ul>
<li><code>taxonkit lineage</code>: <ul>
<li>new flag <code>-n/--show-name</code> for appending scientific name.</li>
<li>new flag <code>-L/--no-lineage</code> for hide lineage, this is for fast retrieving names or/and ranks.</li>
</ul>
</li>
<li><code>taxonkit reformat</code>:<ul>
<li>fix flag <code>-F/--fill-miss-rank</code>.</li>
<li>discard order restriction of rank symbols.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.0">TaxonKit v0.6.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.6.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.6.0/total.svg" /></a><ul>
<li><code>taxonkit list</code>:<ul>
<li>check merged and deleted taxids.</li>
<li>fix bug of json output. <a href="https://github.com/shenwei356/taxonkit/issues/30">#30</a></li>
</ul>
</li>
<li><code>taxonkit name2taxid</code>: new flag <code>-s/--sci-name</code> for limiting to searching scientific names. <a href="https://github.com/shenwei356/taxonkit/issues/29">#29</a></li>
<li><code>taxonkit version</code>: make checking update optional</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.5.0">TaxonKit v0.5.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.5.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.5.0/total.svg" /></a><ul>
<li><code>taxonkit</code>: requiring delnodes.dmp and merged.dmp.</li>
<li><code>taxonkit lineage</code>: detect deleted and merged taxids now. <a href="https://github.com/shenwei356/taxonkit/issues/19">#19</a></li>
<li><code>taxonkit list/name2taxid</code>: add short flag <code>-r</code> for <code>--show-rank</code>, <code>-n</code> for <code>--show-name</code>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.3">TaxonKit v0.4.3</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.4.3/total.svg" /></a><ul>
<li><code>taxonkit taxid-changelog</code>: rewrite logic, fix bug and add more change types</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.2">TaxonKit v0.4.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.4.2/total.svg" /></a><ul>
<li><code>taxonkit taxid-changelog</code>: change output of <code>ABSORB</code>, do not merged into one record for changes in different versions.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.1">TaxonKit v0.4.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.4.1/total.svg" /></a><ul>
<li><code>taxonkit taxid-changelog</code>: <ul>
<li>add fields: <code>name</code> and <code>rank</code>.</li>
<li>and fix sorting bug.</li>
<li>detailed lineage change status</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.0">TaxonKit v0.4.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.4.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.4.0/total.svg" /></a><ul>
<li>new command: <code>taxonkit taxid-changelog</code>: for creating taxid changelog from dump archive</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.3.0">TaxonKit v0.3.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.3.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.3.0/total.svg" /></a><ul>
<li>this version is almost the same as v0.2.5</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.5">TaxonKit v0.2.5</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.5/total.svg" /></a><ul>
<li>add global flag: <code>--line-buffered</code> to disable output buffer. <a href="https://github.com/shenwei356/taxonkit/issues/11">#11</a></li>
<li>replace global flags <code>--names-file</code> and <code>--nodes-file</code> with <code>--data-dir</code>, also support environment variable <code>TAXONKIT_DB</code>. <a href="https://github.com/shenwei356/taxonkit/issues/17">#17</a></li>
<li><code>taxonkit reformat</code>: detects lineages containing unofficial taxon name and won't show panic message.</li>
<li><code>taxonkit name2taxid</code>: supports synonyms names. <a href="https://github.com/shenwei356/taxonkit/commit/d4fac1c1138a571957f52eb431ff0d85c03852a8">#9</a></li>
<li><code>taxokit lineage</code>: add flag <code>-r/--show-rank</code> to print rank at another new column.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.4">TaxonKit v0.2.4</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.4/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>:<ul>
<li>more accurate result when using flag <code>-F/--fill-miss-rank</code> to estimate and fill missing rank with original lineage information</li>
<li>supporting escape strings like <code>\t</code>, <code>\n</code>, <a href="https://github.com/shenwei356/taxonkit/issues/5">#5</a></li>
<li>outputting corresponding taxids for reformated lineage. <a href="https://github.com/shenwei356/taxonkit/issues/8">#8</a></li>
</ul>
</li>
<li><code>taxonkit lineage</code>:<ul>
<li>fix bug for taxid <code>1</code> <a href="https://github.com/shenwei356/taxonkit/issues/7">#7</a></li>
<li>add flag <code>-d/--delimiter</code>.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.3">TaxonKit v0.2.3</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.3/total.svg" /></a><ul>
<li>fix bug brought in v0.2.1</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.2">TaxonKit v0.2.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.2/total.svg" /></a><ul>
<li>make verbose information optional <a href="https://github.com/shenwei356/taxonkit/issues/4">#4</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.1">TaxonKit v0.2.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.1/total.svg" /></a><ul>
<li><code>taxonkit list</code>: fix bug of no output for leaf nodes of the taxonomic tree.
<a href="https://github.com/shenwei356/taxonkit/issues/4">#4</a></li>
<li>add new command <code>genautocomplete</code> to generate <strong>shell autocompletion</strong> script!</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.0">TaxonKit v0.2.0</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.2.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.2.0/total.svg" /></a><ul>
<li>add command <code>name2taxid</code> to query taxid by taxon scientific name.</li>
<li><code>lineage</code>, <code>reformat</code>: <strong><em>changed flags and default operations</em></strong>,
 check the <a href="http://bioinf.shenwei.me/taxonkit/usage/">usage</a>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.8">TaxonKit v0.1.8</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.8"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.8/total.svg" /></a><ul>
<li><code>taxonkit lineage</code>, add an extra column of lineage in Taxid.
 <a href="https://github.com/shenwei356/taxonkit/issues/3">#3</a>. e.g.,</li>
<li>fix colorful output in windows.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.7">TaxonKit v0.1.7</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.7"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.7/total.svg" /></a><ul>
<li><code>taxonkit reformat</code>: supports reading stdin from output of <code>taxonkit lineage</code>,
 reformated lineages are appended to input data.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.6">TaxonKit v0.1.6</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.6"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.6/total.svg" /></a><ul>
<li>remove flag <code>-f/--formated-rank</code> from <code>taxonkit lineage</code>,
  using <code>taxonkit reformat</code> can archieve same result.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.5">TaxonKit v0.1.5</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.5/total.svg" /></a><ul>
<li>reorganize code and flags</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.4">TaxonKit v0.1.4</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.4/total.svg" /></a><ul>
<li>add flag <code>--fill</code> for <code>taxonkit reformat</code>, which estimates and fills missing rank with original lineage information</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.3">TaxonKit v0.1.3</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.3/total.svg" /></a><ul>
<li>add command of <code>taxonkit reformat</code> which reformats full lineage to custom format</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.2">TaxonKit v0.1.2</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.2/total.svg" /></a><ul>
<li>add command of <code>taxonkit lineage</code>, users can query lineage of given taxon IDs from file</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.1">TaxonKit v0.1.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1.1/total.svg" /></a><ul>
<li>add feature of <code>taxonkit list</code>, users can choose output in readable JSON
 format by flag <code>--json</code> so the taxonomy tree could be collapse and
 uncollapse in modern text editor.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1">TaxonKit v0.1</a>
<a href="https://github.com/shenwei356/taxonkit/releases/tag/v0.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/taxonkit/v0.1/total.svg" /></a><ul>
<li>first release</li>
</ul>
</li>
</ul>
<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//taxonkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.81fa17fe.min.js"></script>
      
    
  </body>
</html>